// @portinari/portinari-ui/esm5/lib/services/po-dialog/po-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { browserLanguage, poLocaleDefault } from '../../utils/util';
import { PoDialogType } from './po-dialog.enum';
import { PoModalComponent } from '../../components/po-modal/po-modal.component';
/** @type {?} */
export var poDialogAlertLiteralsDefault = {
    en: (/** @type {?} */ ({ ok: 'Ok' })),
    es: (/** @type {?} */ ({ ok: 'Ok' })),
    pt: (/** @type {?} */ ({ ok: 'Ok' })),
    ru: (/** @type {?} */ ({ ok: 'Ок' }))
};
/** @type {?} */
export var poDialogConfirmLiteralsDefault = {
    en: (/** @type {?} */ ({ cancel: 'Cancel', confirm: 'Confirm' })),
    es: (/** @type {?} */ ({ cancel: 'Cancelar', confirm: 'Confirmar' })),
    pt: (/** @type {?} */ ({ cancel: 'Cancelar', confirm: 'Confirmar' })),
    ru: (/** @type {?} */ ({ cancel: 'отменить', confirm: 'подтвердить' }))
};
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que serve como container do po-dialog.service
 */
var PoDialogComponent = /** @class */ (function () {
    function PoDialogComponent() {
        // Objeto primary do poModal
        this.primaryAction = { label: 'ok', action: (/**
             * @return {?}
             */
            function () { }) };
    }
    /**
     * @return {?}
     */
    PoDialogComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.closeSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    PoDialogComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.closeSubscription = this.poModal.onXClosed.subscribe((/**
         * @param {?} close
         * @return {?}
         */
        function (close) { return _this.destroy(); }));
    };
    // Fecha o poModal
    // Fecha o poModal
    /**
     * @return {?}
     */
    PoDialogComponent.prototype.close = 
    // Fecha o poModal
    /**
     * @return {?}
     */
    function () {
        this.poModal.close();
        this.destroy();
    };
    /**
     * @return {?}
     */
    PoDialogComponent.prototype.destroy = /**
     * @return {?}
     */
    function () {
        if (this.componentRef) {
            this.componentRef.destroy();
        }
    };
    // Insere os valores recebidos de this.open para o poModal
    // Insere os valores recebidos de this.open para o poModal
    /**
     * @param {?=} primaryLabel
     * @param {?=} primaryAction
     * @param {?=} secondaryLabel
     * @param {?=} secondaryAction
     * @return {?}
     */
    PoDialogComponent.prototype.configDialog = 
    // Insere os valores recebidos de this.open para o poModal
    /**
     * @param {?=} primaryLabel
     * @param {?=} primaryAction
     * @param {?=} secondaryLabel
     * @param {?=} secondaryAction
     * @return {?}
     */
    function (primaryLabel, primaryAction, secondaryLabel, secondaryAction) {
        var _this = this;
        this.primaryAction = {
            label: primaryLabel,
            action: (/**
             * @return {?}
             */
            function () {
                if (primaryAction) {
                    primaryAction();
                }
                _this.close();
            })
        };
        if (secondaryLabel) {
            this.secondaryAction = {
                label: secondaryLabel,
                action: (/**
                 * @return {?}
                 */
                function () {
                    if (secondaryAction) {
                        secondaryAction();
                    }
                    _this.close();
                })
            };
        }
    };
    // Insere os valores recebidos de po-dialog.service de acordo com o tipo de diálago solicitado
    // Insere os valores recebidos de po-dialog.service de acordo com o tipo de diálago solicitado
    /**
     * @param {?} dialogOptions
     * @param {?} dialogType
     * @param {?=} componentRef
     * @return {?}
     */
    PoDialogComponent.prototype.open = 
    // Insere os valores recebidos de po-dialog.service de acordo com o tipo de diálago solicitado
    /**
     * @param {?} dialogOptions
     * @param {?} dialogType
     * @param {?=} componentRef
     * @return {?}
     */
    function (dialogOptions, dialogType, componentRef) {
        this.title = dialogOptions.title;
        this.message = dialogOptions.message;
        this.componentRef = componentRef;
        this.setDialogLiterals(dialogOptions, dialogType);
        switch (dialogType) {
            case PoDialogType.Confirm: {
                this.configDialog(this.literalsConfirm.confirm, ((/** @type {?} */ (dialogOptions))).confirm, this.literalsConfirm.cancel, ((/** @type {?} */ (dialogOptions))).cancel);
                break;
            }
            case PoDialogType.Alert: {
                this.configDialog(this.literalsAlert.ok, ((/** @type {?} */ (dialogOptions))).ok);
                break;
            }
        }
        this.poModal.open();
    };
    /**
     * @private
     * @param {?} dialogOptions
     * @param {?} dialogType
     * @return {?}
     */
    PoDialogComponent.prototype.setDialogLiterals = /**
     * @private
     * @param {?} dialogOptions
     * @param {?} dialogType
     * @return {?}
     */
    function (dialogOptions, dialogType) {
        /** @type {?} */
        var alertLiterals = poDialogAlertLiteralsDefault;
        /** @type {?} */
        var confirmLiterals = poDialogConfirmLiteralsDefault;
        /** @type {?} */
        var literals = dialogOptions.literals;
        if (dialogType === PoDialogType.Alert) {
            this.literalsAlert = tslib_1.__assign({}, alertLiterals[poLocaleDefault], alertLiterals[browserLanguage()], literals);
        }
        else {
            this.literalsConfirm = tslib_1.__assign({}, confirmLiterals[poLocaleDefault], confirmLiterals[browserLanguage()], literals);
        }
    };
    PoDialogComponent.decorators = [
        { type: Component, args: [{
                    template: "<po-modal\n  [p-title]=\"title\"\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\">\n  {{ message }}\n</po-modal>\n"
                }] }
    ];
    PoDialogComponent.propDecorators = {
        poModal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }]
    };
    return PoDialogComponent;
}());
export { PoDialogComponent };
if (false) {
    /** @type {?} */
    PoDialogComponent.prototype.poModal;
    /** @type {?} */
    PoDialogComponent.prototype.title;
    /** @type {?} */
    PoDialogComponent.prototype.message;
    /** @type {?} */
    PoDialogComponent.prototype.primaryAction;
    /** @type {?} */
    PoDialogComponent.prototype.secondaryAction;
    /** @type {?} */
    PoDialogComponent.prototype.literalsAlert;
    /** @type {?} */
    PoDialogComponent.prototype.literalsConfirm;
    /**
     * @type {?}
     * @private
     */
    PoDialogComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    PoDialogComponent.prototype.closeSubscription;
}
