// @portinari/portinari-ui/esm5/lib/services/po-dialog/po-dialog.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { PoModalModule } from '../../components/po-modal/po-modal.module';
import { PoDialogComponent } from './po-dialog.component';
import { PoDialogService } from './po-dialog.service';
var PoDialogModule = /** @class */ (function () {
    function PoDialogModule() {
    }
    PoDialogModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        PoDialogComponent
                    ],
                    imports: [
                        PoModalModule
                    ],
                    exports: [
                        PoDialogComponent
                    ],
                    entryComponents: [
                        PoDialogComponent
                    ],
                    providers: [
                        PoDialogService
                    ]
                },] }
    ];
    return PoDialogModule;
}());
export { PoDialogModule };
