// @portinari/portinari-ui/esm2015/lib/components/po-menu/services/po-menu-items.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço que implementa a comunicação entre os items do po-menu.
 */
export class PoMenuItemsService {
    constructor() {
        this.subjectParent = new Subject();
        this.subjectChild = new Subject();
    }
    // Envia informações do click do po-menu-item para o po-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    sendToParentMenuClicked(menu) {
        this.subjectChild.next(menu);
    }
    // Recebe do po-menu-item sua informação de click.
    /**
     * @return {?}
     */
    receiveFromChildMenuClicked() {
        return this.subjectChild.asObservable();
    }
    // Envia para os po-menu-item a resposta do processamento de click de um po-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    sendToChildMenuClicked(menu) {
        this.subjectParent.next(menu);
    }
    // Recebe do po-menu as informações processadas do click de um po-menu-item.
    /**
     * @return {?}
     */
    receiveFromParentMenuClicked() {
        return this.subjectParent.asObservable();
    }
}
PoMenuItemsService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoMenuItemsService.prototype.subjectParent;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemsService.prototype.subjectChild;
}
