// @portinari/portinari-ui/esm5/lib/components/po-slide/po-slide-circles/po-slide-circles.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para o conjunto de círculos que faz a passagem de itens do `po-slide`.
 */
var PoSlideCirclesComponent = /** @class */ (function () {
    function PoSlideCirclesComponent() {
        /**
         * Evento emitido ao clicar em um controle.
         */
        this.click = new EventEmitter();
    }
    PoSlideCirclesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-slide-circles',
                    template: "<div class=\"po-slide-circles\">\n  <ng-container *ngFor=\"let item of items; let index = index;\">\n    <div\n      class=\"po-slide-circle\"\n      [class.po-slide-active-circle]=\"index === currentSlideIndex\"\n      (click)=\"click.emit(index)\">\n    </div>\n  </ng-container>\n</div>\n"
                }] }
    ];
    PoSlideCirclesComponent.propDecorators = {
        currentSlideIndex: [{ type: Input, args: ['p-current-slide-index',] }],
        items: [{ type: Input, args: ['p-items',] }],
        click: [{ type: Output, args: ['p-click',] }]
    };
    return PoSlideCirclesComponent;
}());
export { PoSlideCirclesComponent };
if (false) {
    /**
     * Index do slide atual.
     * @type {?}
     */
    PoSlideCirclesComponent.prototype.currentSlideIndex;
    /**
     * Itens do slide.
     * @type {?}
     */
    PoSlideCirclesComponent.prototype.items;
    /**
     * Evento emitido ao clicar em um controle.
     * @type {?}
     */
    PoSlideCirclesComponent.prototype.click;
}
