// @portinari/portinari-ui/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area-overlay/po-upload-drag-drop-area-overlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
var PoUploadDragDropAreaOverlayComponent = /** @class */ (function () {
    function PoUploadDragDropAreaOverlayComponent(renderer) {
        this.renderer = renderer;
        this.areaElement = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PoUploadDragDropAreaOverlayComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.target) {
            this.setPosition(this.target);
        }
        this.areaElement.emit(this.DragDropAreaFixed.nativeElement);
    };
    /**
     * @private
     * @param {?} targetElement
     * @return {?}
     */
    PoUploadDragDropAreaOverlayComponent.prototype.setPosition = /**
     * @private
     * @param {?} targetElement
     * @return {?}
     */
    function (targetElement) {
        var _this = this;
        /** @type {?} */
        var boundingClientRect = targetElement.nativeElement.getBoundingClientRect();
        /** @type {?} */
        var clientRectKeys = ['bottom', 'left', 'height', 'right', 'top', 'width'];
        clientRectKeys.forEach((/**
         * @param {?} clientRectKey
         * @return {?}
         */
        function (clientRectKey) {
            /** @type {?} */
            var clientRectValue = boundingClientRect[clientRectKey];
            _this.renderer.setStyle(_this.DragDropAreaFixed.nativeElement, clientRectKey, clientRectValue + "px");
        }));
    };
    PoUploadDragDropAreaOverlayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-upload-drag-drop-area-overlay',
                    template: "<div class=\"po-overlay-fixed\">\n\n  <po-upload-drag-drop-area #DragDropAreaFixed\n    class=\"po-upload-drag-drop-area-overlay\"\n    [p-directory-compatible]=\"directoryCompatible\"\n    [p-disabled]=\"disabled\"\n    [p-literals]=\"literals\"\n    [p-overlay]=\"true\">\n  </po-upload-drag-drop-area>\n\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PoUploadDragDropAreaOverlayComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    PoUploadDragDropAreaOverlayComponent.propDecorators = {
        DragDropAreaFixed: [{ type: ViewChild, args: ['DragDropAreaFixed', { read: ElementRef, static: true },] }],
        directoryCompatible: [{ type: Input, args: ['p-directory-compatible',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        target: [{ type: Input, args: ['p-target',] }],
        areaElement: [{ type: Output, args: ['p-area-element',] }]
    };
    return PoUploadDragDropAreaOverlayComponent;
}());
export { PoUploadDragDropAreaOverlayComponent };
if (false) {
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.DragDropAreaFixed;
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.directoryCompatible;
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.disabled;
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.literals;
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.target;
    /** @type {?} */
    PoUploadDragDropAreaOverlayComponent.prototype.areaElement;
    /**
     * @type {?}
     * @private
     */
    PoUploadDragDropAreaOverlayComponent.prototype.renderer;
}
