// @portinari/portinari-ui/esm2015/lib/components/po-field/po-upload/po-upload-file.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PoUploadStatus } from './po-upload-status.enum';
// Classe responsável pelo arquivo do PO Upload
export class PoUploadFile {
    /**
     * @param {?} file
     */
    constructor(file) {
        if (file) {
            this.name = file.name;
            this.displayName = `${file.name} - ${this.getFileSize(file.size)}`;
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = PoUploadStatus.None;
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    getExtension(value) {
        if (value) {
            /** @type {?} */
            const extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    }
    /**
     * @private
     * @return {?}
     */
    generateUUID() {
        /**
         * @return {?}
         */
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
            s4() + '-' + s4() + s4() + s4();
    }
    // Retorna o tamanho do arquivo em KBytes.
    /**
     * @private
     * @param {?} size
     * @return {?}
     */
    getFileSize(size) {
        /** @type {?} */
        let kbSize = 0;
        if (size) {
            kbSize = Math.ceil(size / 1024);
        }
        return `${kbSize} KB`;
    }
}
if (false) {
    /** @type {?} */
    PoUploadFile.prototype.name;
    /** @type {?} */
    PoUploadFile.prototype.rawFile;
    /** @type {?} */
    PoUploadFile.prototype.status;
    /** @type {?} */
    PoUploadFile.prototype.uid;
    /** @type {?} */
    PoUploadFile.prototype.extension;
    /** @type {?} */
    PoUploadFile.prototype.size;
    /** @type {?} */
    PoUploadFile.prototype.displayName;
    /** @type {?} */
    PoUploadFile.prototype.percent;
}
