// @portinari/portinari-ui/esm2015/lib/components/po-navbar/po-navbar-items/po-navbar-items.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { PoNavbarItemComponent } from './po-navbar-item/po-navbar-item.component';
export class PoNavbarItemsComponent {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.routeSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subscribeToRoute();
    }
    /**
     * @param {?} item
     * @return {?}
     */
    selectItem(item) {
        this.selectedItem = item;
    }
    /**
     * @private
     * @param {?} urlRouter
     * @return {?}
     */
    checkActiveItemByUrl(urlRouter) {
        this.selectedItem = this.items.find((/**
         * @param {?} item
         * @return {?}
         */
        item => item.link === urlRouter));
    }
    /**
     * @private
     * @return {?}
     */
    checkRouterChildrenFragments() {
        /** @type {?} */
        const childrenPrimary = this.router.parseUrl(this.router.url).root.children['primary'];
        return childrenPrimary ? `/${childrenPrimary.segments.map((/**
         * @param {?} it
         * @return {?}
         */
        it => it.path)).join('/')}` : '';
    }
    /**
     * @private
     * @return {?}
     */
    subscribeToRoute() {
        this.routeSubscription = this.router.events.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        val => {
            if (val instanceof NavigationEnd || val instanceof NavigationCancel) {
                /** @type {?} */
                const urlRouter = this.checkRouterChildrenFragments();
                this.checkActiveItemByUrl(urlRouter);
            }
        }));
    }
}
PoNavbarItemsComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-navbar-items',
                template: "<nav>\n  <ul class=\"po-navbar-items-container\" #navbarItemsContainer>\n    <li *ngFor=\"let item of items\"\n      class=\"po-navbar-item\"\n      [class.po-navbar-item-selected]=\"selectedItem === item\">\n      <po-navbar-item\n        [p-action]=\"item.action\"\n        [p-clickable]=\"selectedItem !== item\"\n        [p-label]=\"item.label\"\n        [p-link]=\"item.link\"\n        (p-click)=\"selectItem(item)\">\n      </po-navbar-item>\n    </li>\n  </ul>\n</nav>\n"
            }] }
];
/** @nocollapse */
PoNavbarItemsComponent.ctorParameters = () => [
    { type: Router }
];
PoNavbarItemsComponent.propDecorators = {
    navbarItemsContainer: [{ type: ViewChild, args: ['navbarItemsContainer', { read: ElementRef, static: true },] }],
    allNavbarItems: [{ type: ViewChildren, args: [PoNavbarItemComponent, { read: ElementRef },] }],
    items: [{ type: Input, args: ['p-items',] }]
};
if (false) {
    /** @type {?} */
    PoNavbarItemsComponent.prototype.selectedItem;
    /**
     * @type {?}
     * @private
     */
    PoNavbarItemsComponent.prototype.routeSubscription;
    /** @type {?} */
    PoNavbarItemsComponent.prototype.navbarItemsContainer;
    /** @type {?} */
    PoNavbarItemsComponent.prototype.allNavbarItems;
    /** @type {?} */
    PoNavbarItemsComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    PoNavbarItemsComponent.prototype.router;
}
