// @portinari/portinari-ui/esm2015/lib/components/po-progress/po-progress-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean, convertToInt } from '../../utils/util';
import { PoProgressStatus } from './enums/po-progress-status.enum';
/** @type {?} */
const poProgressMaxValue = 100;
/** @type {?} */
const poProgressMinValue = 0;
/**
 * \@description
 *
 * Componente de barra de progresso que possibilita exibir visualmente o progresso/carregamento de uma tarefa.
 *
 * Este componente pode ser utilizado no *upload* de arquivos, uma atualização no sistema ou o processamento de uma imagem.
 */
export class PoProgressBaseComponent {
    constructor() {
        this._value = 0;
        /**
         * \@optional
         *
         * \@description
         *
         * Status da barra de progresso que indicará visualmente ao usuário
         * o andamento, por exemplo, se a mesma foi concluída com sucesso.
         *
         * \@default `PoProgressStatus.Default`
         */
        this.status = PoProgressStatus.Default;
        /**
         * \@optional
         *
         * \@description
         *
         * Evento que será disparado ao clicar no ícone de cancelamento ("x") na parte inferior da barra de progresso.
         *
         * Ao ser disparado, a função receberá como parâmetro o status atual da barra de progresso.
         *
         * > Se nenhuma função for passada para o evento ou a barra de progresso estiver com o status `PoProgressStatus.Success`,
         * o ícone de cancelamento não será exibido.
         */
        this.cancel = new EventEmitter();
        /**
         * \@optional
         *
         * \@description
         *
         * Evento que será disparado ao clicar no ícone de tentar novamente na parte inferior da barra de progresso.
         *
         * > o ícone será exibido apenas se informar uma função neste evento e o status da barra de progresso for
         * `PoProgressStatus.Error`.
         */
        this.retry = new EventEmitter();
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Habilita o modo indeterminado na barra de progresso, que mostra uma animação fixa sem um valor estabelecido.
     *
     * Esta opção pode ser utilizada quando não souber quanto tempo levará para que um processo seja concluído.
     *
     * > Caso esta propriedade e a `p-value` seja habilitada, a propriedade `p-value` será ignorada.
     *
     * \@default `false`
     * @param {?} indeterminate
     * @return {?}
     */
    set indeterminate(indeterminate) {
        this._indeterminate = convertToBoolean(indeterminate);
    }
    /**
     * @return {?}
     */
    get indeterminate() {
        return this._indeterminate;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Valor que representará o progresso.
     *
     * > Os valores aceitos são números inteiros de `0` à `100`.
     *
     * \@default `0`
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        /** @type {?} */
        const integerValue = convertToInt(value, poProgressMinValue);
        /** @type {?} */
        const isProgressRangeValue = this.isProgressRangeValue(integerValue);
        this._value = isProgressRangeValue ? integerValue : poProgressMinValue;
    }
    /**
     * @return {?}
     */
    get value() {
        return this._value;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isProgressRangeValue(value) {
        return value >= poProgressMinValue && value <= poProgressMaxValue;
    }
}
PoProgressBaseComponent.propDecorators = {
    indeterminate: [{ type: Input, args: ['p-indeterminate',] }],
    info: [{ type: Input, args: ['p-info',] }],
    infoIcon: [{ type: Input, args: ['p-info-icon',] }],
    status: [{ type: Input, args: ['p-status',] }],
    text: [{ type: Input, args: ['p-text',] }],
    value: [{ type: Input, args: ['p-value',] }],
    cancel: [{ type: Output, args: ['p-cancel',] }],
    retry: [{ type: Output, args: ['p-retry',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoProgressBaseComponent.prototype._indeterminate;
    /**
     * @type {?}
     * @private
     */
    PoProgressBaseComponent.prototype._value;
    /**
     * \@optional
     *
     * \@description
     *
     * Informação adicional que aparecerá abaixo da barra de progresso ao lado direito.
     * @type {?}
     */
    PoProgressBaseComponent.prototype.info;
    /**
     * \@optional
     *
     * \@description
     *
     * Ícone que aparecerá ao lado do texto da propriedade `p-info`.
     *
     * Exemplo: `po-icon-ok`.
     * @type {?}
     */
    PoProgressBaseComponent.prototype.infoIcon;
    /**
     * \@optional
     *
     * \@description
     *
     * Status da barra de progresso que indicará visualmente ao usuário
     * o andamento, por exemplo, se a mesma foi concluída com sucesso.
     *
     * \@default `PoProgressStatus.Default`
     * @type {?}
     */
    PoProgressBaseComponent.prototype.status;
    /**
     * \@optional
     *
     * \@description
     *
     * Texto principal que aparecerá abaixo da barra de progresso no lado esquerdo.
     * @type {?}
     */
    PoProgressBaseComponent.prototype.text;
    /**
     * \@optional
     *
     * \@description
     *
     * Evento que será disparado ao clicar no ícone de cancelamento ("x") na parte inferior da barra de progresso.
     *
     * Ao ser disparado, a função receberá como parâmetro o status atual da barra de progresso.
     *
     * > Se nenhuma função for passada para o evento ou a barra de progresso estiver com o status `PoProgressStatus.Success`,
     * o ícone de cancelamento não será exibido.
     * @type {?}
     */
    PoProgressBaseComponent.prototype.cancel;
    /**
     * \@optional
     *
     * \@description
     *
     * Evento que será disparado ao clicar no ícone de tentar novamente na parte inferior da barra de progresso.
     *
     * > o ícone será exibido apenas se informar uma função neste evento e o status da barra de progresso for
     * `PoProgressStatus.Error`.
     * @type {?}
     */
    PoProgressBaseComponent.prototype.retry;
}
