// @portinari/portinari-ui/esm2015/lib/components/po-field/po-clean/po-clean.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { PoCleanBaseComponent } from './po-clean-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoCleanBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="po-clean-labs" title="Portinari Clean Labs">
 *   <file name="sample-po-clean-labs.component.html"> </file>
 *   <file name="sample-po-clean-labs.component.ts"> </file>
 * </example-private>
 */
export class PoCleanComponent extends PoCleanBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
        this.viewRef = viewRef;
        this.parentComponent = this.viewRef['_view']['component'];
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    setInputValue(value) {
        if (this.inputRef && this.inputRef.nativeElement) {
            this.inputRef.nativeElement.value = value;
        }
    }
    /**
     * @return {?}
     */
    getInputValue() {
        if (this.inputRef && this.inputRef.nativeElement) {
            return this.inputRef.nativeElement.value;
        }
    }
}
PoCleanComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-clean',
                template: "<span *ngIf=\"showIcon()\"\n  class=\"po-icon po-field-icon po-icon-close\"\n  (click)=\"clear()\">\n</span>\n"
            }] }
];
/** @nocollapse */
PoCleanComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoCleanComponent.prototype.viewRef;
}
