// @portinari/portinari-ui/esm2015/lib/components/po-navbar/po-navbar-actions/po-navbar-action/po-navbar-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, isExternalLink, openExternalLink } from '../../../../utils/util';
export class PoNavbarActionComponent {
    /**
     * @param {?} viewContainerRef
     * @param {?} router
     */
    constructor(viewContainerRef, router) {
        this.router = router;
        this.parentRef = viewContainerRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    click() {
        if (this.action) {
            callFunction(this.action, this.parentRef || this);
            return;
        }
        if (this.link) {
            return this.openUrl(this.link);
        }
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    openUrl(url) {
        if (isExternalLink(url)) {
            return openExternalLink(url);
        }
        if (url) {
            return this.router.navigate([url]);
        }
    }
}
PoNavbarActionComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-navbar-action',
                template: "<div \n  tabindex=\"0\"\n  class=\"po-navbar-action-content po-clickable\"\n  (click)=\"click()\">\n\n  <span\n    class=\"po-icon {{ icon }}\"\n    [p-tooltip]=\"tooltip\">\n  </span>\n\n</div>\n"
            }] }
];
/** @nocollapse */
PoNavbarActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Router }
];
PoNavbarActionComponent.propDecorators = {
    action: [{ type: Input, args: ['p-action',] }],
    icon: [{ type: Input, args: ['p-icon',] }],
    label: [{ type: Input, args: ['p-label',] }],
    link: [{ type: Input, args: ['p-link',] }],
    tooltip: [{ type: Input, args: ['p-tooltip-text',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoNavbarActionComponent.prototype.parentRef;
    /** @type {?} */
    PoNavbarActionComponent.prototype.action;
    /** @type {?} */
    PoNavbarActionComponent.prototype.icon;
    /** @type {?} */
    PoNavbarActionComponent.prototype.label;
    /** @type {?} */
    PoNavbarActionComponent.prototype.link;
    /** @type {?} */
    PoNavbarActionComponent.prototype.tooltip;
    /**
     * @type {?}
     * @private
     */
    PoNavbarActionComponent.prototype.router;
}
