// @portinari/portinari-ui/esm5/lib/components/po-field/po-lookup/po-lookup-modal/po-lookup-modal-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { browserLanguage, isTypeof, poLocaleDefault } from '../../../../utils/util';
import { PoModalComponent } from '../../../../components/po-modal/po-modal.component';
import { PoTableColumnSortType } from '../../../po-table';
import { poTableLiteralsDefault } from '../../../po-table/po-table-base.component';
/** @type {?} */
export var poLookupLiteralsDefault = {
    en: (/** @type {?} */ ({
        modalPrimaryActionLabel: 'Select',
        modalSecondaryActionLabel: 'Cancel',
        modalPlaceholder: 'Search',
        modalTitle: 'Select a record',
        modalTableNoColumns: poTableLiteralsDefault.en.noColumns,
        modalTableNoData: poTableLiteralsDefault.en.noData,
        modalTableLoadingData: poTableLiteralsDefault.en.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.en.loadMoreData
    })),
    es: (/** @type {?} */ ({
        modalPrimaryActionLabel: 'Seleccionar',
        modalSecondaryActionLabel: 'Cancelar',
        modalPlaceholder: 'Buscar',
        modalTitle: 'Seleccione un registro',
        modalTableNoColumns: poTableLiteralsDefault.es.noColumns,
        modalTableNoData: poTableLiteralsDefault.es.noData,
        modalTableLoadingData: poTableLiteralsDefault.es.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.es.loadMoreData
    })),
    pt: (/** @type {?} */ ({
        modalPrimaryActionLabel: 'Selecionar',
        modalSecondaryActionLabel: 'Cancelar',
        modalPlaceholder: 'Pesquisar',
        modalTitle: 'Selecione um registro',
        modalTableNoColumns: poTableLiteralsDefault.pt.noColumns,
        modalTableNoData: poTableLiteralsDefault.pt.noData,
        modalTableLoadingData: poTableLiteralsDefault.pt.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.pt.loadMoreData
    })),
    ru: (/** @type {?} */ ({
        modalPrimaryActionLabel: 'выбирать',
        modalSecondaryActionLabel: 'отменить',
        modalPlaceholder: 'поиск',
        modalTitle: 'Выберите запись',
        modalTableNoColumns: poTableLiteralsDefault.ru.noColumns,
        modalTableNoData: poTableLiteralsDefault.ru.noData,
        modalTableLoadingData: poTableLiteralsDefault.ru.loadingData,
        modalTableLoadMoreData: poTableLiteralsDefault.ru.loadMoreData
    }))
};
/**
 * \@docsPrivate
 *
 * Classe base do componente Po Lookup Modal.
 * @abstract
 */
var PoLookupModalBaseComponent = /** @class */ (function () {
    function PoLookupModalBaseComponent() {
        var _this = this;
        this.hasNext = true;
        this.isLoading = false;
        this.page = 1;
        this.pageSize = 10;
        this.primaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                _this.items.forEach((/**
                 * @param {?} element
                 * @return {?}
                 */
                function (element) {
                    if (element['$selected']) {
                        _this.model.emit(element);
                        _this.poModal.close();
                    }
                }));
            }),
            label: this.literals.modalPrimaryActionLabel
        };
        this.searchValue = '';
        this.secondaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                _this.model.emit(null);
                _this.poModal.close();
            }),
            label: this.literals.modalSecondaryActionLabel
        };
        /**
         * Evento utilizado ao selecionar um registro da tabela.
         */
        this.model = new EventEmitter();
    }
    Object.defineProperty(PoLookupModalBaseComponent.prototype, "literals", {
        get: /**
         * @return {?}
         */
        function () {
            return this._literals || poLookupLiteralsDefault[browserLanguage()];
        },
        /** Objeto com as literais usadas no `po-lookup-modal`. */
        set: /**
         * Objeto com as literais usadas no `po-lookup-modal`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poLookupLiteralsDefault[poLocaleDefault], poLookupLiteralsDefault[browserLanguage()], value);
                if (value.modalTitle) {
                    this.title = this.literals.modalTitle;
                }
            }
            else {
                this._literals = poLookupLiteralsDefault[browserLanguage()];
            }
            this.primaryAction.label = this.literals.modalPrimaryActionLabel;
            this.secondaryAction.label = this.literals.modalSecondaryActionLabel;
            this.setTableLiterals();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoLookupModalBaseComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        /** Título da modal. */
        set: /**
         * Título da modal.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = isTypeof(value, 'string') ? value : this.literals.modalTitle;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.filterSubscription) {
            this.filterSubscription.unsubscribe();
        }
        if (this.searchSubscription) {
            this.searchSubscription.unsubscribe();
        }
        if (this.showMoreSubscription) {
            this.showMoreSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initializeData();
        this.setTableLiterals();
    };
    /**
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.search = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.page = 1;
        if (this.searchValue) {
            this.isLoading = true;
            this.searchSubscription = this.getFilteredItems(this.searchValue).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                _this.items = data.items;
                _this.hasNext = data.hasNext;
                _this.isLoading = false;
            }));
        }
        else {
            this.initializeData();
        }
    };
    /**
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.showMoreEvent = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.page++;
        this.isLoading = true;
        this.showMoreSubscription = this.getFilteredItems(this.searchValue).subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            data.items.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                _this.items.push(item);
            }));
            _this.hasNext = data.hasNext;
            _this.isLoading = false;
        }));
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.getFilteredItems = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        var _a = this, page = _a.page, pageSize = _a.pageSize, filterParams = _a.filterParams;
        if (this.filterService.getFilteredItems) {
            /** @type {?} */
            var filteredParams = this.getFilteredParams(filter);
            return this.filterService.getFilteredItems(filteredParams);
        }
        return this.filterService.getFilteredData(filter, page, pageSize, filterParams);
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.getFilteredParams = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        var _a = this, page = _a.page, pageSize = _a.pageSize, filterParams = _a.filterParams, sort = _a.sort;
        /** @type {?} */
        var filteredParams = {};
        /** @type {?} */
        var order = this.getOrderParam(sort);
        /** @type {?} */
        var params = { filter: filter, page: page, pageSize: pageSize, order: order, filterParams: filterParams };
        for (var key in params) {
            if (params.hasOwnProperty(key) && params[key]) {
                filteredParams[key] = params[key];
            }
        }
        return filteredParams;
    };
    /**
     * @private
     * @param {?=} sort
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.getOrderParam = /**
     * @private
     * @param {?=} sort
     * @return {?}
     */
    function (sort) {
        if (sort === void 0) { sort = { type: undefined }; }
        var column = sort.column, type = sort.type;
        if (!column) {
            return;
        }
        if (type === PoTableColumnSortType.Descending) {
            return "-" + column.property;
        }
        return "" + column.property;
    };
    /**
     * @private
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.initializeData = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.isLoading = true;
        this.filterSubscription = this.getFilteredItems('').subscribe((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.items = data.items;
            _this.hasNext = data.hasNext;
            _this.isLoading = false;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.setTableLiterals = /**
     * @private
     * @return {?}
     */
    function () {
        this.tableLiterals = {
            'noColumns': this.literals.modalTableNoColumns,
            'noData': this.literals.modalTableNoData,
            'loadingData': this.literals.modalTableLoadingData,
            'loadMoreData': this.literals.modalTableLoadMoreData,
        };
    };
    PoLookupModalBaseComponent.propDecorators = {
        poModal: [{ type: ViewChild, args: [PoModalComponent, { static: true },] }],
        columns: [{ type: Input, args: ['p-columns',] }],
        items: [{ type: Input, args: ['p-items',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        title: [{ type: Input, args: ['p-title',] }],
        filterService: [{ type: Input, args: ['p-filter-service',] }],
        filterParams: [{ type: Input, args: ['p-filter-params',] }],
        model: [{ type: Output, args: ['p-change-model',] }]
    };
    return PoLookupModalBaseComponent;
}());
export { PoLookupModalBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoLookupModalBaseComponent.prototype._literals;
    /**
     * @type {?}
     * @private
     */
    PoLookupModalBaseComponent.prototype._title;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.hasNext;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.isLoading;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.page;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.pageSize;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.primaryAction;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.searchValue;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.secondaryAction;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.tableLiterals;
    /**
     * @type {?}
     * @protected
     */
    PoLookupModalBaseComponent.prototype.sort;
    /**
     * @type {?}
     * @private
     */
    PoLookupModalBaseComponent.prototype.filterSubscription;
    /**
     * @type {?}
     * @private
     */
    PoLookupModalBaseComponent.prototype.searchSubscription;
    /**
     * @type {?}
     * @private
     */
    PoLookupModalBaseComponent.prototype.showMoreSubscription;
    /** @type {?} */
    PoLookupModalBaseComponent.prototype.poModal;
    /**
     * Lista das colunas da tabela.
     * Essa propriedade deve receber um array de objetos que implementam a interface PoLookupColumn.
     * @type {?}
     */
    PoLookupModalBaseComponent.prototype.columns;
    /**
     * Lista de itens da tabela.
     * @type {?}
     */
    PoLookupModalBaseComponent.prototype.items;
    /**
     * Classe de serviço com a implementação do cliente.
     * @type {?}
     */
    PoLookupModalBaseComponent.prototype.filterService;
    /**
     * Classe de serviço com a implementação do cliente.
     * @type {?}
     */
    PoLookupModalBaseComponent.prototype.filterParams;
    /**
     * Evento utilizado ao selecionar um registro da tabela.
     * @type {?}
     */
    PoLookupModalBaseComponent.prototype.model;
    /**
     * @abstract
     * @return {?}
     */
    PoLookupModalBaseComponent.prototype.openModal = function () { };
}
