// @portinari/portinari-ui/esm5/lib/components/po-field/po-lookup/po-lookup-modal/po-lookup-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, ViewChild } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { PoLookupModalBaseComponent } from '../po-lookup-modal/po-lookup-modal-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoLookupModalBaseComponent
 */
var PoLookupModalComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoLookupModalComponent, _super);
    function PoLookupModalComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.keyUpObservable = null;
        _this.containerHeight = 375;
        _this.tableHeight = 370;
        return _this;
    }
    /**
     * @return {?}
     */
    PoLookupModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.initializeEventInput();
        this.setTableHeight();
    };
    /**
     * @return {?}
     */
    PoLookupModalComponent.prototype.initializeEventInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.keyUpObservable = fromEvent(this.inputSearchEl.nativeElement, 'keyup')
            .pipe(filter((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.validateEnterPressed(e); })), debounceTime(400));
        this.keyUpObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.search();
        }));
    };
    /**
     * @return {?}
     */
    PoLookupModalComponent.prototype.openModal = /**
     * @return {?}
     */
    function () {
        this.poModal.open();
    };
    /**
     * @param {?} sort
     * @return {?}
     */
    PoLookupModalComponent.prototype.sortBy = /**
     * @param {?} sort
     * @return {?}
     */
    function (sort) {
        this.sort = sort;
    };
    /**
     * @private
     * @return {?}
     */
    PoLookupModalComponent.prototype.setTableHeight = /**
     * @private
     * @return {?}
     */
    function () {
        if (window.innerHeight < 615) {
            this.tableHeight -= 50;
            this.containerHeight -= 50;
        }
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    PoLookupModalComponent.prototype.validateEnterPressed = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return e.keyCode === 13;
    };
    PoLookupModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-lookup-modal',
                    template: "<po-modal\n  p-click-out=\"false\"\n  p-hide-close=\"false\"\n  p-size=\"lg\"\n  [p-primary-action]=\"primaryAction\"\n  [p-secondary-action]=\"secondaryAction\"\n  [p-title]=\"title\">\n\n  <po-field-container\n    class=\"po-lookup-header po-md-6 po-pull-right\"\n    [p-optional]=\"false\">\n\n    <div class=\"po-field-container-content\">\n      <input #inpsearch\n        class=\"po-input po-input-icon-right\"\n        name=\"contentSearch\"\n        [(ngModel)]=\"searchValue\"\n        [placeholder]=\"literals.modalPlaceholder\"\n        type=\"text\">\n\n      <div class=\"po-field-icon-container-right\">\n        <span #iconLookup\n          class=\"po-icon po-field-icon po-icon-search\"\n          (click)=\"search()\">\n        </span>\n      </div>\n    </div>\n\n  </po-field-container>\n\n  <div class=\"po-row po-lookup-container-table\" [style.height.px]=\"containerHeight\">\n\n    <po-table\n      class=\"po-md-12\"\n      p-checkbox=\"true\"\n      p-hide-detail=\"true\"\n      p-single-select=\"true\"\n      p-sort=\"true\"\n      [p-columns]=\"columns\"\n      [p-height]=\"tableHeight\"\n      [p-items]=\"items\"\n      [p-literals]=\"tableLiterals\"\n      [p-loading]=\"isLoading\"\n      [p-show-more-disabled]=\"!hasNext\"\n      (p-show-more)=\"showMoreEvent()\"\n      (p-sort-by)=\"sortBy($event)\">\n    </po-table>\n\n  </div>\n\n</po-modal>\n"
                }] }
    ];
    PoLookupModalComponent.propDecorators = {
        inputSearchEl: [{ type: ViewChild, args: ['inpsearch', { static: true },] }]
    };
    return PoLookupModalComponent;
}(PoLookupModalBaseComponent));
export { PoLookupModalComponent };
if (false) {
    /** @type {?} */
    PoLookupModalComponent.prototype.inputSearchEl;
    /** @type {?} */
    PoLookupModalComponent.prototype.keyUpObservable;
    /** @type {?} */
    PoLookupModalComponent.prototype.containerHeight;
    /** @type {?} */
    PoLookupModalComponent.prototype.tableHeight;
}
