// @portinari/portinari-ui/esm5/lib/components/po-breadcrumb/po-breadcrumb-favorite/po-breadcrumb-favorite.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoBreadcrumbFavoriteService } from './po-breadcrumb-favorite.service';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que renderiza o serviço de favoritar do po-breadcrumb.
 */
var PoBreadcrumbFavoriteComponent = /** @class */ (function () {
    function PoBreadcrumbFavoriteComponent(service) {
        this.service = service;
        this.favorite = false;
    }
    /**
     * @return {?}
     */
    PoBreadcrumbFavoriteComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.service.configService(this.favoriteService, this.paramsService, this.itemActive);
        this.getStatusFavorite();
    };
    /**
     * @return {?}
     */
    PoBreadcrumbFavoriteComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.getSubscription) {
            this.getSubscription.unsubscribe();
        }
        if (this.setSubscription) {
            this.setSubscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    PoBreadcrumbFavoriteComponent.prototype.toggleFavoriteAction = /**
     * @return {?}
     */
    function () {
        this.favorite ? this.setStatusFavorite(false) : this.setStatusFavorite(true);
    };
    /**
     * @private
     * @return {?}
     */
    PoBreadcrumbFavoriteComponent.prototype.getStatusFavorite = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.getSubscription = this.service.getFavorite().subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) { return _this.favorite = result.isFavorite; }));
    };
    /**
     * @private
     * @param {?} status
     * @return {?}
     */
    PoBreadcrumbFavoriteComponent.prototype.setStatusFavorite = /**
     * @private
     * @param {?} status
     * @return {?}
     */
    function (status) {
        var _this = this;
        this.setSubscription = this.service.sendStatusFavorite(status).subscribe((/**
         * @param {?} result
         * @return {?}
         */
        function (result) { return _this.favorite = result.isFavorite; }));
    };
    PoBreadcrumbFavoriteComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-breadcrumb-favorite',
                    template: "<div class=\"po-breadcrumb-favorite po-clickable\" (click)=\"toggleFavoriteAction()\">\n\n  <span\n    class=\"po-icon po-icon-star po-breadcrumb-favorite-star po-clickable\"\n    [class.po-breadcrumb-favorite-star-active]=\"favorite\">\n  </span>\n\n  <span class=\"po-hidden-sm po-breadcrumb-favorite-label\" *ngIf=\"favorite\">Desfavoritar</span>\n  <span class=\"po-hidden-sm po-breadcrumb-favorite-label\" *ngIf=\"!favorite\">Favoritar</span>\n</div>\n",
                    providers: [PoBreadcrumbFavoriteService]
                }] }
    ];
    /** @nocollapse */
    PoBreadcrumbFavoriteComponent.ctorParameters = function () { return [
        { type: PoBreadcrumbFavoriteService }
    ]; };
    PoBreadcrumbFavoriteComponent.propDecorators = {
        favoriteService: [{ type: Input, args: ['p-favorite-service',] }],
        itemActive: [{ type: Input, args: ['p-item-active',] }],
        paramsService: [{ type: Input, args: ['p-params-service',] }]
    };
    return PoBreadcrumbFavoriteComponent;
}());
export { PoBreadcrumbFavoriteComponent };
if (false) {
    /** @type {?} */
    PoBreadcrumbFavoriteComponent.prototype.favorite;
    /**
     * @type {?}
     * @private
     */
    PoBreadcrumbFavoriteComponent.prototype.getSubscription;
    /**
     * @type {?}
     * @private
     */
    PoBreadcrumbFavoriteComponent.prototype.setSubscription;
    /** @type {?} */
    PoBreadcrumbFavoriteComponent.prototype.favoriteService;
    /** @type {?} */
    PoBreadcrumbFavoriteComponent.prototype.itemActive;
    /** @type {?} */
    PoBreadcrumbFavoriteComponent.prototype.paramsService;
    /**
     * @type {?}
     * @private
     */
    PoBreadcrumbFavoriteComponent.prototype.service;
}
