// @portinari/portinari-ui/esm5/lib/components/po-widget/po-widget-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, Output, EventEmitter } from '@angular/core';
import { convertToBoolean, isTypeof } from '../../utils/util';
/**
 *
 * \@description
 *
 * O componente `po-widget` é recomendado para exibição de *dashboards*, podendo ser utilizado
 * para incluir vários tipos de conteúdo como: gráficos, tabelas, grids e imagens.
 *
 * Além da exibição de conteúdos, este componente possibilita adicionar ações e um link
 * para ajuda, como também possibilita ser utilizado com ou sem sombra.
 *
 * Para controlar sua largura, é possível utilizar o [Grid System](/guides/grid-system) para um maior
 * controle de seu redimensionamento, assim possibilitando o tratamento para diferentes resoluções.
 * @abstract
 */
var PoWidgetBaseComponent = /** @class */ (function () {
    function PoWidgetBaseComponent() {
        this._disabled = false;
        this._noShadow = false;
        this._primary = false;
        this.containerHeight = 'auto';
        /**
         * Ação que será executada quando o usuário clicar sobre a área total do `po-widget`.
         */
        this.click = new EventEmitter();
        /**
         * Função que será disparada com o valor do `p-disabled` quando esta propriedade for alterada.
         */
        this.onDisabled = new EventEmitter();
        /**
         * Função que será chamada na primeira ação.
         */
        this.primaryAction = new EventEmitter();
        /**
         * Função que será chamada na segunda ação.
         */
        this.secondaryAction = new EventEmitter();
        /**
         * Função chamada ao clicar no ícone de configuração
         */
        this.setting = new EventEmitter();
        /**
         * Função que será chamada ao clicar no título.
         */
        this.titleAction = new EventEmitter();
    }
    Object.defineProperty(PoWidgetBaseComponent.prototype, "background", {
        get: /**
         * @return {?}
         */
        function () {
            return this._background;
        },
        /**
         * @optional
         *
         * @description
         *
         * Aplicação de imagem de fundo.
         * > Se a imagem escolhida intervir na legibilidade do texto contido no `p-widget`,
         * pode-se utilizar a propriedade `p-primary` em conjunto para que os textos fiquem na cor branca.
         *
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Aplicação de imagem de fundo.
         * > Se a imagem escolhida intervir na legibilidade do texto contido no `p-widget`,
         * pode-se utilizar a propriedade `p-primary` em conjunto para que os textos fiquem na cor branca.
         *
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._background = (value && typeof value === 'string') ? value : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita todas as ações do componente.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Desabilita todas as ações do componente.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = convertToBoolean(value);
            this.onDisabled.emit(this.disabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this._height;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define a altura do `po-widget`.
         * A altura mínima para o `po-widget` depende do que será exibido através das propriedades `p-primary-label`,
         * `p-setting`, `p-help` e `p-title`.
         * > Caso não seja informado valor, a propriedade irá assumir o tamanho do conteúdo.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define a altura do `po-widget`.
         * A altura mínima para o `po-widget` depende do que será exibido através das propriedades `p-primary-label`,
         * `p-setting`, `p-help` e `p-title`.
         * > Caso não seja informado valor, a propriedade irá assumir o tamanho do conteúdo.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._height = parseInt((/** @type {?} */ (value)), 10);
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "help", {
        get: /**
         * @return {?}
         */
        function () {
            return this._help;
        },
        /**
         * @optional
         *
         * @description
         *
         * Link de ajuda
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Link de ajuda
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._help = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "noShadow", {
        get: /**
         * @return {?}
         */
        function () {
            return this._noShadow;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita a sombra do `po-widget`.
         *
         * @default `true`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Desabilita a sombra do `po-widget`.
         *
         * \@default `true`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._noShadow = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "primary", {
        get: /**
         * @return {?}
         */
        function () {
            return this._primary;
        },
        /**
         * @optional
         *
         * @description
         *
         * Opção para que o `po-widget` fique em destaque.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Opção para que o `po-widget` fique em destaque.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._primary = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "primaryLabel", {
        get: /**
         * @return {?}
         */
        function () {
            return this._primaryLabel;
        },
        /**
         * @optional
         *
         * @description
         *
         * Descrição da primeira ação.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Descrição da primeira ação.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._primaryLabel = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoWidgetBaseComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        /**
         * @optional
         *
         * @description
         *
         * Título do `po-widget`.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Título do `po-widget`.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._title = isTypeof(value, 'string') ? value : '';
            this.setHeight(this.height);
        },
        enumerable: true,
        configurable: true
    });
    PoWidgetBaseComponent.propDecorators = {
        background: [{ type: Input, args: ['p-background',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        height: [{ type: Input, args: ['p-height',] }],
        help: [{ type: Input, args: ['p-help',] }],
        noShadow: [{ type: Input, args: ['p-no-shadow',] }],
        primary: [{ type: Input, args: ['p-primary',] }],
        primaryLabel: [{ type: Input, args: ['p-primary-label',] }],
        secondaryLabel: [{ type: Input, args: ['p-secondary-label',] }],
        title: [{ type: Input, args: ['p-title',] }],
        click: [{ type: Output, args: ['p-click',] }],
        onDisabled: [{ type: Output, args: ['p-on-disabled',] }],
        primaryAction: [{ type: Output, args: ['p-primary-action',] }],
        secondaryAction: [{ type: Output, args: ['p-secondary-action',] }],
        setting: [{ type: Output, args: ['p-setting',] }],
        titleAction: [{ type: Output, args: ['p-title-action',] }]
    };
    return PoWidgetBaseComponent;
}());
export { PoWidgetBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._background;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._height;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._help;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._noShadow;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._primary;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._primaryLabel;
    /**
     * @type {?}
     * @private
     */
    PoWidgetBaseComponent.prototype._title;
    /** @type {?} */
    PoWidgetBaseComponent.prototype.containerHeight;
    /**
     * Descrição da segunda ação.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.secondaryLabel;
    /**
     * Ação que será executada quando o usuário clicar sobre a área total do `po-widget`.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.click;
    /**
     * Função que será disparada com o valor do `p-disabled` quando esta propriedade for alterada.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.onDisabled;
    /**
     * Função que será chamada na primeira ação.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.primaryAction;
    /**
     * Função que será chamada na segunda ação.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.secondaryAction;
    /**
     * Função chamada ao clicar no ícone de configuração
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.setting;
    /**
     * Função que será chamada ao clicar no título.
     * @type {?}
     */
    PoWidgetBaseComponent.prototype.titleAction;
    /**
     * @abstract
     * @param {?} height
     * @return {?}
     */
    PoWidgetBaseComponent.prototype.setHeight = function (height) { };
}
