// @portinari/portinari-ui/esm2015/lib/components/po-menu-panel/po-menu-panel-item/po-menu-panel-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoMenuPanelItemsService } from '../services/po-menu-panel-items.service';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que implementa cada item do po-menu-panel.
 */
export class PoMenuPanelItemComponent {
    /**
     * @param {?} menuItemsService
     */
    constructor(menuItemsService) {
        this.menuItemsService = menuItemsService;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.itemsSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // subscribe to menu component messages
        this.subscribeMenuClickedFromParent();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    clickMenuItem(event) {
        if (!(event.ctrlKey || event.metaKey)) {
            event.preventDefault();
            // Emmit to parent
            this.menuItemsService.sendToParentMenuClicked(this.menuItemInternal);
        }
    }
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    activateMenu(menu) {
        this.menuItemInternal.isSelected = this.menuItemInternal.id === menu.id;
    }
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    processMenuItem(menu) {
        if (this.menuItemInternal.type === 'internalLink') {
            this.activateMenu(menu.active);
        }
    }
    /**
     * @private
     * @return {?}
     */
    subscribeMenuClickedFromParent() {
        this.itemsSubscription = this.menuItemsService.receiveFromParentMenuClicked().subscribe((/**
         * @param {?} menu
         * @return {?}
         */
        menu => {
            this.processMenuItem(menu);
        }));
    }
}
PoMenuPanelItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-menu-panel-item',
                template: "<!-- menu com link interno -->\n<a *ngIf=\"menuItemInternal.type === 'internalLink'\" [routerLink]=\"menuItemInternal.link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<!-- menu com link externo -->\n<a *ngIf=\"menuItemInternal.type === 'externalLink'\" [href]=\"menuItemInternal.link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<!-- menu sem link -->\n<a *ngIf=\"menuItemInternal.type === 'noLink'\" href=\"javascript:;\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n\n<ng-template #menuItemTemplate>\n  <div class=\"po-menu-panel-item\"\n    p-tooltip-position=\"right\"\n    [class.po-menu-panel-item-selected]=\"menuItemInternal.isSelected\"\n    [p-tooltip]=\"menuItemInternal.label\"\n    (click)=\"clickMenuItem($event);\">\n    <span class=\"po-icon {{menuItemInternal.icon}}\"></span>\n  </div>\n</ng-template>\n"
            }] }
];
/** @nocollapse */
PoMenuPanelItemComponent.ctorParameters = () => [
    { type: PoMenuPanelItemsService }
];
PoMenuPanelItemComponent.propDecorators = {
    menuItemInternal: [{ type: Input, args: ['p-menu-item-internal',] }]
};
if (false) {
    /** @type {?} */
    PoMenuPanelItemComponent.prototype.itemsSubscription;
    /** @type {?} */
    PoMenuPanelItemComponent.prototype.menuItemInternal;
    /**
     * @type {?}
     * @private
     */
    PoMenuPanelItemComponent.prototype.menuItemsService;
}
