// @portinari/portinari-ui/esm2015/lib/components/po-table/po-table-subtitle-footer/po-table-subtitle-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, Renderer2 } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para a criação de um conjunto de legendas.
 */
export class PoTableSubtitleFooterComponent {
    /**
     * @param {?} element
     * @param {?} renderer
     */
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.initializeResizeListener();
        this.debounceResize();
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        if (!this.isVisible && this.getContainerSize() > 0) {
            this.toggleShowCompleteSubtitle();
            this.isVisible = true;
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeResizeListener();
    }
    /**
     * @private
     * @return {?}
     */
    debounceResize() {
        clearTimeout(this.timeoutResize);
        this.timeoutResize = setTimeout((/**
         * @return {?}
         */
        () => {
            this.toggleShowCompleteSubtitle();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    getContainerSize() {
        return this.element.nativeElement.querySelector('.po-table-subtitle-footer-container').offsetWidth;
    }
    /**
     * @private
     * @return {?}
     */
    getItemsSize() {
        /** @type {?} */
        const items = this.element.nativeElement.querySelectorAll('.po-table-subtitle-footer');
        return Array.from(items).map((/**
         * @param {?} item
         * @return {?}
         */
        item => item['offsetWidth'])).reduce((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        (a, b) => a + b), 16);
    }
    /**
     * @private
     * @return {?}
     */
    initializeResizeListener() {
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.debounceResize();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    removeResizeListener() {
        this.resizeListener();
    }
    /**
     * @private
     * @return {?}
     */
    toggleShowCompleteSubtitle() {
        /** @type {?} */
        const containerSize = this.getContainerSize();
        /** @type {?} */
        const itemsSize = this.getItemsSize();
        this.showSubtitle = itemsSize > containerSize;
    }
}
PoTableSubtitleFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-table-subtitle-footer',
                template: "<div class=\"po-table-subtitle-footer-container\">\n  <div *ngFor=\"let subtitle of subtitles\" class=\"po-table-subtitle-footer\">\n    <po-table-subtitle-circle\n      p-hide-title=\"true\"\n      [p-subtitle]=\"subtitle\">\n    </po-table-subtitle-circle>\n    <span class=\"po-table-subtitle-text\">{{ subtitle.label }}</span>\n  </div>\n  <po-table-show-subtitle *ngIf=\"showSubtitle\"\n    [p-literals]=\"literals\"\n    [p-subtitles]=\"subtitles\">\n  </po-table-show-subtitle>\n</div>\n"
            }] }
];
/** @nocollapse */
PoTableSubtitleFooterComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
PoTableSubtitleFooterComponent.propDecorators = {
    literals: [{ type: Input, args: ['p-literals',] }],
    subtitles: [{ type: Input, args: ['p-subtitles',] }]
};
if (false) {
    /** @type {?} */
    PoTableSubtitleFooterComponent.prototype.showSubtitle;
    /**
     * @type {?}
     * @private
     */
    PoTableSubtitleFooterComponent.prototype.isVisible;
    /**
     * @type {?}
     * @private
     */
    PoTableSubtitleFooterComponent.prototype.timeoutResize;
    /**
     * @type {?}
     * @protected
     */
    PoTableSubtitleFooterComponent.prototype.resizeListener;
    /**
     * Propriedade que recebe as literais definidas no `po-table`.
     * @type {?}
     */
    PoTableSubtitleFooterComponent.prototype.literals;
    /**
     * Propriedade que recebe as legendas definidas no `PoTableSubtitleCircleComponent`.
     * @type {?}
     */
    PoTableSubtitleFooterComponent.prototype.subtitles;
    /**
     * @type {?}
     * @private
     */
    PoTableSubtitleFooterComponent.prototype.element;
    /** @type {?} */
    PoTableSubtitleFooterComponent.prototype.renderer;
}
