// @portinari/portinari-ui/esm5/lib/components/po-navbar/po-navbar-item-navigation/po-navbar-item-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var PoNavbarItemNavigationComponent = /** @class */ (function () {
    function PoNavbarItemNavigationComponent() {
        this.click = new EventEmitter();
    }
    PoNavbarItemNavigationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-navbar-item-navigation',
                    template: "<po-navbar-item-navigation-icon class=\"po-navbar-item-navigation-icon\"\n  p-icon=\"left\"\n  [p-disabled]=\"disableLeft\"\n  (p-click)=\"click.emit($event)\">\n</po-navbar-item-navigation-icon>\n\n<po-navbar-item-navigation-icon class=\"po-navbar-item-navigation-icon\"\n  p-icon=\"right\"\n  [p-disabled]=\"disableRight\"\n  (p-click)=\"click.emit($event)\">\n</po-navbar-item-navigation-icon>\n"
                }] }
    ];
    PoNavbarItemNavigationComponent.propDecorators = {
        disableLeft: [{ type: Input, args: ['p-disable-left',] }],
        disableRight: [{ type: Input, args: ['p-disable-right',] }],
        click: [{ type: Output, args: ['p-click',] }]
    };
    return PoNavbarItemNavigationComponent;
}());
export { PoNavbarItemNavigationComponent };
if (false) {
    /** @type {?} */
    PoNavbarItemNavigationComponent.prototype.disableLeft;
    /** @type {?} */
    PoNavbarItemNavigationComponent.prototype.disableRight;
    /** @type {?} */
    PoNavbarItemNavigationComponent.prototype.click;
}
