// @portinari/portinari-ui/esm2015/lib/components/po-field/po-combo/po-combo-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-combo.
 */
export class PoComboFilterService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
        this.fieldLabel = 'label';
        this.fieldValue = 'value';
    }
    /**
     * @return {?}
     */
    get url() { return this._url; }
    /**
     * @param {?} param
     * @param {?=} filterParams
     * @return {?}
     */
    getFilteredData(param, filterParams) {
        /** @type {?} */
        const value = param.value;
        /** @type {?} */
        const params = new HttpParams({
            fromString: `filter=${value}`
        });
        return this.http.get(`${this.url}`, { responseType: 'json', params: params })
            .pipe(map((/**
         * @param {?} response
         * @return {?}
         */
        (response) => this.parseToArrayComboOption(response.items))));
    }
    /**
     * @param {?} value
     * @param {?=} filterParams
     * @return {?}
     */
    getObjectByValue(value, filterParams) {
        return this.http.get(`${this.url}/${value}`).pipe(map((/**
         * @param {?} item
         * @return {?}
         */
        item => this.parseToComboOption(item))));
    }
    /**
     * @param {?} url
     * @param {?} fieldLabel
     * @param {?} fieldValue
     * @return {?}
     */
    configProperties(url, fieldLabel, fieldValue) {
        this._url = url;
        this.fieldLabel = fieldLabel;
        this.fieldValue = fieldValue;
    }
    /**
     * @private
     * @param {?} items
     * @return {?}
     */
    parseToArrayComboOption(items) {
        if (items && items.length > 0) {
            return items.map((/**
             * @param {?} item
             * @return {?}
             */
            item => {
                return this.parseToComboOption(item);
            }));
        }
        return [];
    }
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    parseToComboOption(item) {
        if (item && item[this.fieldValue]) {
            /** @type {?} */
            const label = item[this.fieldLabel];
            /** @type {?} */
            const value = item[this.fieldValue];
            return { label, value };
        }
    }
}
PoComboFilterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PoComboFilterService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoComboFilterService.prototype._url;
    /** @type {?} */
    PoComboFilterService.prototype.fieldLabel;
    /** @type {?} */
    PoComboFilterService.prototype.fieldValue;
    /**
     * @type {?}
     * @private
     */
    PoComboFilterService.prototype.http;
}
