// @portinari/portinari-ui/esm2015/lib/components/po-field/po-multiselect/po-multiselect-item/po-multiselect-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que constrói cada item no dropdown, contendo o checkbox e o label.
 */
export class PoMultiselectItemComponent {
    constructor() {
        /**
         * Esta propriedade indica se o campo está selecionado ou não.
         */
        this.selected = false;
        /**
         * Evento que será disparado toda vez que o usuário marcar ou desmarcar um item.
         */
        this.change = new EventEmitter();
    }
    /**
     * @return {?}
     */
    itemClicked() {
        this.selected = !this.selected;
        this.change.emit(this.selected);
    }
}
PoMultiselectItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-multiselect-item',
                template: "<li [class.po-multiselect-item-selected]=\"selected\" (click)=\"itemClicked()\">\n\n  <a class=\"po-multiselect-item\">\n    <input\n      [checked]=\"selected\"\n      class=\"po-multiselect-checkbox-input\"\n      [class.po-multiselect-checkbox-input-checked]=\"selected\"\n      type=\"checkbox\">\n\n    <label class=\"po-multiselect-checkbox-label po-clickable\">\n      {{ label }}\n    </label>\n  </a>\n</li>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PoMultiselectItemComponent.propDecorators = {
    label: [{ type: Input, args: ['p-label',] }],
    selected: [{ type: Input, args: ['p-selected',] }],
    change: [{ type: Output, args: ['p-change',] }]
};
if (false) {
    /**
     * Rótulo do item.
     * @type {?}
     */
    PoMultiselectItemComponent.prototype.label;
    /**
     * Esta propriedade indica se o campo está selecionado ou não.
     * @type {?}
     */
    PoMultiselectItemComponent.prototype.selected;
    /**
     * Evento que será disparado toda vez que o usuário marcar ou desmarcar um item.
     * @type {?}
     */
    PoMultiselectItemComponent.prototype.change;
}
