// @portinari/portinari-ui/esm5/lib/components/po-menu-panel/services/po-menu-panel-items.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço que implementa a comunicação entre os items do po-menu-panel.
 */
var PoMenuPanelItemsService = /** @class */ (function () {
    function PoMenuPanelItemsService() {
        this.subjectChild = new Subject();
        this.subjectParent = new Subject();
    }
    // Recebe do po-menu-panel-item sua informação de click.
    // Recebe do po-menu-panel-item sua informação de click.
    /**
     * @return {?}
     */
    PoMenuPanelItemsService.prototype.receiveFromChildMenuClicked = 
    // Recebe do po-menu-panel-item sua informação de click.
    /**
     * @return {?}
     */
    function () {
        return this.subjectChild.asObservable();
    };
    // Recebe do po-menu-panel as informações processadas do click de um po-menu-panel-item.
    // Recebe do po-menu-panel as informações processadas do click de um po-menu-panel-item.
    /**
     * @return {?}
     */
    PoMenuPanelItemsService.prototype.receiveFromParentMenuClicked = 
    // Recebe do po-menu-panel as informações processadas do click de um po-menu-panel-item.
    /**
     * @return {?}
     */
    function () {
        return this.subjectParent.asObservable();
    };
    // Envia informações do click do po-menu-panel-item para o po-menu
    // Envia informações do click do po-menu-panel-item para o po-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    PoMenuPanelItemsService.prototype.sendToParentMenuClicked = 
    // Envia informações do click do po-menu-panel-item para o po-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectChild.next(menu);
    };
    // Envia para os po-menu-panel-item a resposta do processamento de click de um po-menu-panel-item.
    // Envia para os po-menu-panel-item a resposta do processamento de click de um po-menu-panel-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    PoMenuPanelItemsService.prototype.sendToChildMenuClicked = 
    // Envia para os po-menu-panel-item a resposta do processamento de click de um po-menu-panel-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectParent.next(menu);
    };
    PoMenuPanelItemsService.decorators = [
        { type: Injectable }
    ];
    return PoMenuPanelItemsService;
}());
export { PoMenuPanelItemsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoMenuPanelItemsService.prototype.subjectChild;
    /**
     * @type {?}
     * @private
     */
    PoMenuPanelItemsService.prototype.subjectParent;
}
