// @portinari/portinari-ui/esm5/lib/components/po-menu/services/po-menu-items.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço que implementa a comunicação entre os items do po-menu.
 */
var PoMenuItemsService = /** @class */ (function () {
    function PoMenuItemsService() {
        this.subjectParent = new Subject();
        this.subjectChild = new Subject();
    }
    // Envia informações do click do po-menu-item para o po-menu
    // Envia informações do click do po-menu-item para o po-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    PoMenuItemsService.prototype.sendToParentMenuClicked = 
    // Envia informações do click do po-menu-item para o po-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectChild.next(menu);
    };
    // Recebe do po-menu-item sua informação de click.
    // Recebe do po-menu-item sua informação de click.
    /**
     * @return {?}
     */
    PoMenuItemsService.prototype.receiveFromChildMenuClicked = 
    // Recebe do po-menu-item sua informação de click.
    /**
     * @return {?}
     */
    function () {
        return this.subjectChild.asObservable();
    };
    // Envia para os po-menu-item a resposta do processamento de click de um po-menu-item.
    // Envia para os po-menu-item a resposta do processamento de click de um po-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    PoMenuItemsService.prototype.sendToChildMenuClicked = 
    // Envia para os po-menu-item a resposta do processamento de click de um po-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectParent.next(menu);
    };
    // Recebe do po-menu as informações processadas do click de um po-menu-item.
    // Recebe do po-menu as informações processadas do click de um po-menu-item.
    /**
     * @return {?}
     */
    PoMenuItemsService.prototype.receiveFromParentMenuClicked = 
    // Recebe do po-menu as informações processadas do click de um po-menu-item.
    /**
     * @return {?}
     */
    function () {
        return this.subjectParent.asObservable();
    };
    PoMenuItemsService.decorators = [
        { type: Injectable }
    ];
    return PoMenuItemsService;
}());
export { PoMenuItemsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoMenuItemsService.prototype.subjectParent;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemsService.prototype.subjectChild;
}
