// @portinari/portinari-ui/esm5/lib/components/po-modal/po-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ElementRef, Renderer2, ViewChild } from '@angular/core';
import { v4 as uuid } from 'uuid';
import { PoModalBaseComponent } from './po-modal-base.component';
import { PoModalService } from './po-modal-service';
/**
 * \@docsExtends PoModalBaseComponent
 *
 * \@example
 *
 * <example name="po-modal-basic" title="Portinari Modal Basic">
 *  <file name="sample-po-modal-basic/sample-po-modal-basic.component.html"> </file>
 *  <file name="sample-po-modal-basic/sample-po-modal-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-modal-labs" title="Portinari Modal Labs">
 *  <file name="sample-po-modal-labs/sample-po-modal-labs.component.html"> </file>
 *  <file name="sample-po-modal-labs/sample-po-modal-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-modal-fruits-salad" title="Portinari Modal - Fruits Salad">
 *  <file name="sample-po-modal-fruits-salad/sample-po-modal-fruits-salad.component.html"> </file>
 *  <file name="sample-po-modal-fruits-salad/sample-po-modal-fruits-salad.component.ts"> </file>
 * </example>
 */
var PoModalComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoModalComponent, _super);
    function PoModalComponent(poModalService, renderer, changeDetector) {
        var _this = _super.call(this) || this;
        _this.poModalService = poModalService;
        _this.renderer = renderer;
        _this.changeDetector = changeDetector;
        _this.focusableElements = 'input, select, textarea, button:not([disabled]), a';
        _this.id = uuid();
        return _this;
    }
    /**
     * @param {?=} xClosed
     * @return {?}
     */
    PoModalComponent.prototype.close = /**
     * @param {?=} xClosed
     * @return {?}
     */
    function (xClosed) {
        if (xClosed === void 0) { xClosed = false; }
        this.poModalService.modalActive = undefined;
        _super.prototype.close.call(this, xClosed);
        this.removeEventListeners();
        if (this.sourceElement) {
            this.sourceElement.focus();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PoModalComponent.prototype.closeModalOnEscapeKey = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.hideClose) {
            event.preventDefault();
            event.stopPropagation();
            this.close();
        }
    };
    /**
     * @return {?}
     */
    PoModalComponent.prototype.getPrimaryActionButtonType = /**
     * @return {?}
     */
    function () {
        return this.primaryAction.danger ? 'danger' : 'primary';
    };
    /**
     * @return {?}
     */
    PoModalComponent.prototype.getSecondaryActionButtonType = /**
     * @return {?}
     */
    function () {
        return this.secondaryAction && this.secondaryAction.danger && !this.primaryAction.danger ? 'danger' : 'default';
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PoModalComponent.prototype.onClickOut = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.clickOut && !this.modalContent.nativeElement.contains(event.target)) {
            this.close();
        }
    };
    /**
     * @return {?}
     */
    PoModalComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.sourceElement = document.activeElement;
        _super.prototype.open.call(this);
        this.handleFocus();
    };
    /**
     * @private
     * @return {?}
     */
    PoModalComponent.prototype.handleFocus = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.poModalService.modalActive = this.id;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.modalContent) {
                _this.initFocus();
                document.addEventListener('focus', _this.focusFunction, true);
            }
        }));
    };
    /**
     * @private
     * @return {?}
     */
    PoModalComponent.prototype.initFocus = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.focusFunction = (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.poModalService.modalActive = _this.poModalService.modalActive || _this.id;
            /** @type {?} */
            var modalElement = _this.modalContent.nativeElement;
            if (!modalElement.contains(event.target) && _this.poModalService.modalActive === _this.id) {
                event.stopPropagation();
                _this.firstElement.focus();
            }
        });
        this.setFirstElement();
        if (this.hideClose) {
            this.firstElement.focus();
        }
        else {
            /** @type {?} */
            var firstFieldElement = this.modalContent.nativeElement.querySelectorAll(this.focusableElements)[1] ||
                this.modalContent.nativeElement;
            firstFieldElement.focus();
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoModalComponent.prototype.removeEventListeners = /**
     * @private
     * @return {?}
     */
    function () {
        document.removeEventListener('focus', this.focusFunction, true);
    };
    /**
     * @private
     * @return {?}
     */
    PoModalComponent.prototype.setFirstElement = /**
     * @private
     * @return {?}
     */
    function () {
        this.firstElement = this.modalContent.nativeElement.querySelector(this.focusableElements) || this.modalContent.nativeElement;
    };
    PoModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-modal',
                    template: "<div *ngIf=\"!isHidden\"\n  class=\"po-modal\"\n  tabindex=\"0\"\n  (keydown.esc)=\"closeModalOnEscapeKey($event)\">\n\n  <div class=\"po-modal-overlay\">\n    <div class=\"po-modal-container po-pb-2 po-pt-2\" (mousedown)=\"onClickOut($event)\">\n\n      <div class=\"po-modal-vertical-align\">\n        <div #modalContent\n          class=\"po-modal-content po-modal-{{ size }}\"\n          tabindex=\"-1\">\n\n          <div class=\"po-modal-header\">\n            <div class=\"po-modal-title po-text-ellipsis\">\n              {{ title }}\n            </div>\n\n            <a *ngIf=\"!hideClose\"\n              class=\"po-modal-header-close-button\"\n              tabindex=\"0\"\n              (click)=\"close(true)\">\n              <span class=\"po-icon po-icon-close\"></span>\n            </a>\n          </div>\n\n          <div class=\"po-modal-body\">\n            <ng-content></ng-content>\n          </div>\n\n          <div class=\"po-modal-footer\">\n            <po-button *ngIf=\"secondaryAction\"\n              [p-disabled]=\"secondaryAction.disabled\"\n              [p-label]=\"secondaryAction.label\"\n              [p-loading]=\"secondaryAction.loading\"\n              [p-type]=\"getSecondaryActionButtonType()\"\n              (p-click)=\"secondaryAction.action()\">\n            </po-button>\n\n            <po-button\n              class=\"po-button-modal-first-action\"\n              [p-disabled]=\"primaryAction.disabled\"\n              [p-label]=\"primaryAction.label\"\n              [p-loading]=\"primaryAction.loading\"\n              [p-type]=\"getPrimaryActionButtonType()\"\n              (p-click)=\"primaryAction.action()\">\n            </po-button>\n          </div>\n\n        </div>\n      </div>\n\n    </div>\n  </div>\n</div>\n\n"
                }] }
    ];
    /** @nocollapse */
    PoModalComponent.ctorParameters = function () { return [
        { type: PoModalService },
        { type: Renderer2 },
        { type: ChangeDetectorRef }
    ]; };
    PoModalComponent.propDecorators = {
        modalContent: [{ type: ViewChild, args: ['modalContent', { read: ElementRef, static: false },] }]
    };
    return PoModalComponent;
}(PoModalBaseComponent));
export { PoModalComponent };
if (false) {
    /** @type {?} */
    PoModalComponent.prototype.modalContent;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.firstElement;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.focusFunction;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.focusableElements;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.id;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.sourceElement;
    /** @type {?} */
    PoModalComponent.prototype.onResizeListener;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.poModalService;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    PoModalComponent.prototype.changeDetector;
}
