// @portinari/portinari-ui/esm5/lib/interceptors/interceptors.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { PoHttpInterceptorModule } from './po-http-interceptor/po-http-interceptor.module';
import { PoHttpRequestModule } from './po-http-request/po-http-request.module';
var PoInterceptorsModule = /** @class */ (function () {
    function PoInterceptorsModule() {
    }
    PoInterceptorsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        PoHttpInterceptorModule,
                        PoHttpRequestModule
                    ],
                    exports: [
                        PoHttpInterceptorModule,
                        PoHttpRequestModule
                    ],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return PoInterceptorsModule;
}());
export { PoInterceptorsModule };
