// @portinari/portinari-ui/esm2015/lib/components/po-field/validators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} required
 * @param {?} disabled
 * @param {?} value
 * @return {?}
 */
export function requiredFailed(required, disabled, value) {
    /** @type {?} */
    const valid = ((typeof value === 'string' && value) || (typeof value === 'object' && value && value.length)
        || (typeof value === 'number' && (value || value === 0)) || (typeof value === 'boolean' && value));
    return (required && !disabled && !valid);
}
/**
 * @param {?} maxlength
 * @param {?} value
 * @return {?}
 */
export function maxlengpoailed(maxlength, value) {
    /** @type {?} */
    const validMaxlength = maxlength || maxlength === 0;
    /** @type {?} */
    const validValue = (value || value === 0) && value.toString();
    return validMaxlength && validValue && validValue.length > Number(maxlength);
}
/**
 * @param {?} minlength
 * @param {?} value
 * @return {?}
 */
export function minlengpoailed(minlength, value) {
    /** @type {?} */
    const validMinlength = minlength || minlength === 0;
    /** @type {?} */
    const validValue = (value || value === 0) && value.toString();
    return validMinlength && validValue && validValue.length < Number(minlength);
}
/**
 * @param {?} pattern
 * @param {?} value
 * @return {?}
 */
export function patternFailed(pattern, value) {
    /** @type {?} */
    let reg;
    try {
        reg = new RegExp(pattern);
    }
    catch (e) {
        return true;
    }
    return (pattern && value && !reg.test(value));
}
/**
 * @param {?} min
 * @param {?} value
 * @return {?}
 */
export function minFailed(min, value) {
    /** @type {?} */
    const validValue = value || value === 0;
    /** @type {?} */
    const validMin = min || min === 0;
    return validValue && validMin && value < min;
}
/**
 * @param {?} max
 * @param {?} value
 * @return {?}
 */
export function maxFailed(max, value) {
    /** @type {?} */
    const validValue = value || value === 0;
    /** @type {?} */
    const validMax = max || max === 0;
    return validValue && validMax && value > max;
}
/**
 * @param {?} value
 * @return {?}
 */
export function dateFailed(value) {
    return (value && isNaN(Date.parse(value)));
}
