// @portinari/portinari-ui/esm5/lib/components/po-stepper/po-step/po-step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input } from '@angular/core';
import { uuid } from '../../../utils/util';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
/**
 * \@description
 *
 * O componente `po-step` é utilizado para envolver e renderizar o conteúdo de um passo (*step*) do `po-stepper`, por exemplo:
 *
 * ```
 * <po-stepper>
 *
 *    <po-step p-label="Endereço">
 *
 *        <!-- Conteúdo referente ao endereço -->
 *
 *    </po-step>
 *
 *    <po-step p-label="Pagamento">
 *
 *        <!-- Conteúdo referente ao pagamento -->
 *
 *    </po-step>
 *
 *  </po-stepper>
 * ```
 *
 * A renderização do conteúdo envolvido na tela e o controle dos status são feitos automaticamente. No qual, o primeiro
 * `po-step` encontrado será colocado como ativo, o próximo fica com o status *default* e os demais ficam
 * desabilitados (*disabled*).
 *
 * Ao clicar no `po-step` que está com o status *default*, o que está ativo ficará com o
 * status de concluído (*done*) e o próximo que estava desabilitado ficará com o status *default* e o restante permanecerá
 * com o status desabilitado.
 *
 * > Ao utilizar o `po-step`, o componente `po-stepper` funcionará de forma sequencial, ou seja, não será possível
 * pular para outro `po-step` que esteja com o status igual a desabilitado (*disabled*).
 *
 * Acesse a [documentação do `po-stepper`](/documentation/po-stepper) para ter mais informações sobre o seu funcionamento
 * e exemplos de uso.
 */
var PoStepComponent = /** @class */ (function () {
    function PoStepComponent(elementRef) {
        this.elementRef = elementRef;
        this._status = PoStepperStatus.Disabled;
        // ID do step
        this.id = uuid();
    }
    Object.defineProperty(PoStepComponent.prototype, "status", {
        get: /**
         * @return {?}
         */
        function () {
            return this._status;
        },
        // Estado de exibição do `po-step`.
        set: 
        // Estado de exibição do `po-step`.
        /**
         * @param {?} status
         * @return {?}
         */
        function (status) {
            this._status = status;
            this.setDisplayOnActiveOrError();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoStepComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.setDisplayOnActiveOrError();
    };
    /**
     * @protected
     * @return {?}
     */
    PoStepComponent.prototype.setDisplayOnActiveOrError = /**
     * @protected
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.style.display =
            (this.status === PoStepperStatus.Active || this.status === PoStepperStatus.Error) ? '' : 'none';
    };
    PoStepComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-step',
                    template: "<ng-content></ng-content>\n"
                }] }
    ];
    /** @nocollapse */
    PoStepComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    PoStepComponent.propDecorators = {
        canActiveNextStep: [{ type: Input, args: ['p-can-active-next-step',] }],
        label: [{ type: Input, args: ['p-label',] }],
        status: [{ type: Input, args: ['p-status',] }]
    };
    return PoStepComponent;
}());
export { PoStepComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoStepComponent.prototype._status;
    /** @type {?} */
    PoStepComponent.prototype.id;
    /**
     * \@optional
     *
     * \@description
     *
     * Função chamada quando o próximo *step* for clicado ou quando o método `PoStepperComponent.next()` for chamado.
     * Ao retornar `true` define que esse *step* ficará ativo e o atual como concluído (*done*).
     *
     * Ao ser disparada, a mesma receberá por parâmetro o `PoStepComponent` atual.
     *
     * O contexto da função que será chamada, será o mesmo que o do `PoStepComponent`, então para poder alterar
     * para o contexto do componente que o está utilizando, pode ser utilizado a propriedade `bind` do Javascript.
     * Por exemplo, para a função `validate`:
     *
     * ```
     * <po-step p-label="Step 1" [p-can-active-next-step]="validate.bind(this)">
     * ...
     * </po-step>
     * ```
     * @type {?}
     */
    PoStepComponent.prototype.canActiveNextStep;
    /**
     * Título que será exibido descrevendo o passo (*step*).
     * @type {?}
     */
    PoStepComponent.prototype.label;
    /**
     * @type {?}
     * @private
     */
    PoStepComponent.prototype.elementRef;
}
