// @portinari/portinari-ui/esm5/lib/components/po-list-view/po-list-view.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { PoButtonModule } from '../po-button/po-button.module';
import { PoPopupModule } from '../po-popup/po-popup.module';
import { PoListViewComponent } from './po-list-view.component';
import { PoListViewContentTemplateDirective } from './po-list-view-content-template/po-list-view-content-template.directive';
import { PoListViewDetailTemplateDirective } from './po-list-view-detail-template/po-list-view-detail-template.directive';
/**
 * \@description
 *
 * Módulo do componente `po-list-view`.
 *
 * > Para o correto funcionamento do componente `po-list-view`, deve ser importado o módulo `BrowserAnimationsModule` no
 * > módulo principal da sua aplicação.
 *
 * Módulo da aplicação:
 * ```
 * import { BrowserAnimationsModule } from '\@angular/platform-browser/animations';
 * import { PoModule } from '\@portinari/portinari-ui';
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     BrowserModule,
 *     BrowserAnimationsModule,
 *     ...
 *     PoModule
 *   ],
 *   declarations: [
 *     AppComponent,
 *     ...
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
var PoListViewModule = /** @class */ (function () {
    function PoListViewModule() {
    }
    PoListViewModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        PoButtonModule,
                        PoPopupModule
                    ],
                    declarations: [
                        PoListViewComponent,
                        PoListViewContentTemplateDirective,
                        PoListViewDetailTemplateDirective
                    ],
                    exports: [
                        PoListViewComponent,
                        PoListViewContentTemplateDirective,
                        PoListViewDetailTemplateDirective
                    ],
                    providers: [],
                    schemas: []
                },] }
    ];
    return PoListViewModule;
}());
export { PoListViewModule };
