// @portinari/portinari-ui/esm2015/lib/components/po-field/po-upload/po-upload-file-restrictions/po-upload-file-restrictions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { browserLanguage, formatBytes, poLocaleDefault } from '../../../../utils/util';
import { poUploadLiteralsDefault } from '../po-upload-base.component';
export class PoUploadFileRestrictionsComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set allowedExtensions(value) {
        this._allowedExtensions = this.formatAllowedExtensions((/** @type {?} */ (value)));
    }
    /**
     * @return {?}
     */
    get allowedExtensions() {
        return this._allowedExtensions;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maxFileSize(value) {
        this._maxFileSize = formatBytes((/** @type {?} */ (value)));
    }
    /**
     * @return {?}
     */
    get maxFileSize() {
        return this._maxFileSize;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set minFileSize(value) {
        this._minFileSize = formatBytes((/** @type {?} */ (value)));
    }
    /**
     * @return {?}
     */
    get minFileSize() {
        return this._minFileSize;
    }
    /**
     * @return {?}
     */
    get language() {
        return browserLanguage();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setLiterals();
    }
    /**
     * @private
     * @param {?} allowedExtensions
     * @return {?}
     */
    formatAllowedExtensions(allowedExtensions) {
        /** @type {?} */
        const conjunction = { 'pt': 'e', 'en': 'and', 'es': 'y' };
        return allowedExtensions ? allowedExtensions
            .join(', ')
            .toUpperCase()
            .replace(/,(?=[^,]*$)/, ` ${conjunction[this.language]}`) : undefined;
    }
    /**
     * @private
     * @return {?}
     */
    setLiterals() {
        this.literals = Object.assign({}, poUploadLiteralsDefault[poLocaleDefault], poUploadLiteralsDefault[this.language]);
        this.changeDetector.detectChanges();
    }
}
PoUploadFileRestrictionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-upload-file-restrictions',
                template: "<p class=\"po-font-text-small\" *ngIf=\"maxFiles > 1\">\n  {{ literals?.numberOfFilesAllowed | poI18n:maxFiles }}\n</p>\n\n<p class=\"po-font-text-small\" *ngIf=\"allowedExtensions\">\n  {{ literals?.allowedFormats | poI18n:allowedExtensions }}\n</p>\n\n<p class=\"po-font-text-small\" *ngIf=\"minFileSize || maxFileSize\">\n  <span>{{ literals?.allowedSizes }}</span>\n\n  <span *ngIf=\"minFileSize && maxFileSize\">\n    {{ literals?.allowedFileSizeRange | poI18n:[minFileSize, maxFileSize] }}\n  </span>\n\n  <span *ngIf=\"minFileSize && !maxFileSize\">\n    {{ literals?.minFileSizeAllowed | poI18n:minFileSize }}\n  </span>\n\n  <span *ngIf=\"maxFileSize && !minFileSize\">\n    {{ literals?.maxFileSizeAllowed | poI18n:maxFileSize }}\n  </span>\n</p>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PoUploadFileRestrictionsComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PoUploadFileRestrictionsComponent.propDecorators = {
    allowedExtensions: [{ type: Input, args: ['p-allowed-extensions',] }],
    maxFiles: [{ type: Input, args: ['p-max-files',] }],
    maxFileSize: [{ type: Input, args: ['p-max-file-size',] }],
    minFileSize: [{ type: Input, args: ['p-min-file-size',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoUploadFileRestrictionsComponent.prototype._allowedExtensions;
    /**
     * @type {?}
     * @private
     */
    PoUploadFileRestrictionsComponent.prototype._maxFileSize;
    /**
     * @type {?}
     * @private
     */
    PoUploadFileRestrictionsComponent.prototype._minFileSize;
    /** @type {?} */
    PoUploadFileRestrictionsComponent.prototype.literals;
    /** @type {?} */
    PoUploadFileRestrictionsComponent.prototype.maxFiles;
    /**
     * @type {?}
     * @private
     */
    PoUploadFileRestrictionsComponent.prototype.changeDetector;
}
