// @portinari/portinari-ui/esm2015/lib/components/po-button-group/po-button-group-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
import { PoButtonGroupToggle } from './po-button-group-toggle.enum';
/** @type {?} */
const PO_TOGGLE_TYPE_DEFAULT = 'none';
/**
 * \@description
 *
 * O componente `po-button-group` é formado por um conjunto de botões distribuídos horizontalmente.
 * Cada botão do grupo é tratado de forma individual, recebendo assim um rótulo, uma ação bem como se deverá estar habilitado ou não.
 *
 * Este componente além de servir como um agrupador de botões para ação, também permite que sejam utilizados
 * para seleções multiplas e únicas.
 *
 * O grupo de botões deve ser utilizado para organizar as ações de maneira uniforme e transmitir a ideia de que os botões fazem
 * parte de um mesmo contexto.
 *
 * #### Boas práticas
 *
 * - Evite usar o `po-button-group` com apenas 1 ação, para isso utilize o `po-button`.
 * - Procure utilizar no máximo 3 ações para cada `po-button-group`.
 *
 * > As recomendações do `po-button` também valem para o `po-button-group`.
 */
export class PoButtonGroupBaseComponent {
    constructor() {
        this._small = false;
        this._toggle = PO_TOGGLE_TYPE_DEFAULT;
        /**
         * Lista de botões.
         */
        this.buttons = [];
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Torna o grupo de botões com tamanho minificado.
     *
     * \@default `false`
     * @param {?} value
     * @return {?}
     */
    set small(value) {
        this._small = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get small() {
        return this._small;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Define o modo de seleção de botões.
     *
     * > Veja os valores válidos no *enum* `PoMultiselectFilterMode`.
     *
     * \@default `none`
     * @param {?} value
     * @return {?}
     */
    set toggle(value) {
        this._toggle = ((/** @type {?} */ (Object))).values(PoButtonGroupToggle).includes(value) ? value : PO_TOGGLE_TYPE_DEFAULT;
        this.checkSelecteds(this._toggle);
    }
    /**
     * @return {?}
     */
    get toggle() {
        return this._toggle;
    }
    /**
     * @param {?} buttonClicked
     * @param {?} buttonIndex
     * @return {?}
     */
    onButtonClick(buttonClicked, buttonIndex) {
        if (this.toggle === PoButtonGroupToggle.Single) {
            this.buttons.forEach((/**
             * @param {?} button
             * @param {?} index
             * @return {?}
             */
            (button, index) => button.selected = index === buttonIndex ? !buttonClicked.selected : false));
        }
        else if (this.toggle === PoButtonGroupToggle.Multiple) {
            buttonClicked.selected = !buttonClicked.selected;
        }
    }
    /**
     * @private
     * @param {?} toggleMode
     * @return {?}
     */
    checkSelecteds(toggleMode) {
        if (toggleMode === PoButtonGroupToggle.None) {
            this.deselectAllButtons();
        }
        else if (toggleMode === PoButtonGroupToggle.Single) {
            /** @type {?} */
            const hasMoreOneSelected = this.buttons.filter((/**
             * @param {?} button
             * @return {?}
             */
            button => button.selected)).length > 1;
            if (hasMoreOneSelected) {
                this.deselectAllButtons();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    deselectAllButtons() {
        this.buttons.forEach((/**
         * @param {?} button
         * @return {?}
         */
        button => button.selected = false));
    }
}
PoButtonGroupBaseComponent.propDecorators = {
    buttons: [{ type: Input, args: ['p-buttons',] }],
    small: [{ type: Input, args: ['p-small',] }],
    toggle: [{ type: Input, args: ['p-toggle',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoButtonGroupBaseComponent.prototype._small;
    /**
     * @type {?}
     * @private
     */
    PoButtonGroupBaseComponent.prototype._toggle;
    /**
     * Lista de botões.
     * @type {?}
     */
    PoButtonGroupBaseComponent.prototype.buttons;
}
