// @portinari/portinari-ui/esm5/lib/components/po-avatar/po-avatar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { PoAvatarBaseComponent } from './po-avatar-base.component';
/**
 * \@docsExtends PoAvatarBaseComponent
 *
 * \@example
 *
 * <example name="po-avatar-basic" title="Portinari Avatar Basic" >
 *  <file name="sample-po-avatar-basic/sample-po-avatar-basic.component.html"> </file>
 *  <file name="sample-po-avatar-basic/sample-po-avatar-basic.component.ts"> </file>
 *  <file name="sample-po-avatar-basic/sample-po-avatar-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-avatar-basic/sample-po-avatar-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-avatar-labs" title="Portinari Avatar Labs" >
 *  <file name="sample-po-avatar-labs/sample-po-avatar-labs.component.html"> </file>
 *  <file name="sample-po-avatar-labs/sample-po-avatar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-avatar-business-card" title="Portinari Avatar - Business Card" >
 *  <file name="sample-po-avatar-business-card/sample-po-avatar-business-card.component.html"> </file>
 *  <file name="sample-po-avatar-business-card/sample-po-avatar-business-card.component.ts"> </file>
 * </example>
 */
var PoAvatarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoAvatarComponent, _super);
    function PoAvatarComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    PoAvatarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.src) {
            this.src = undefined;
        }
    };
    /**
     * @return {?}
     */
    PoAvatarComponent.prototype.onError = /**
     * @return {?}
     */
    function () {
        this.src = undefined;
    };
    PoAvatarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-avatar',
                    template: "<div class=\"po-avatar po-avatar-{{ size }}\">\n  <ng-container *ngIf=\"src; then sourceImage; else defaultIcon\"></ng-container>\n</div>\n\n<ng-template #sourceImage>\n  <img class=\"po-avatar-image\" [src]=\"src\" alt=\"\" (error)=\"onError()\">\n</ng-template>\n\n<ng-template #defaultIcon>\n  <span class=\"po-icon po-icon-user po-avatar-default-icon\"></span>\n</ng-template>\n"
                }] }
    ];
    return PoAvatarComponent;
}(PoAvatarBaseComponent));
export { PoAvatarComponent };
