// @portinari/portinari-ui/esm5/lib/components/po-page/po-page-detail/po-page-detail-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, ViewChild } from '@angular/core';
import { browserLanguage, poLocaleDefault } from './../../../utils/util';
import { PoPageContentComponent } from '../po-page-content/po-page-content.component';
/** @type {?} */
export var poPageDetailLiteralsDefault = {
    en: (/** @type {?} */ ({
        back: 'Back',
        edit: 'Edit',
        remove: 'Remove'
    })),
    es: (/** @type {?} */ ({
        back: 'Volver',
        edit: 'Editar',
        remove: 'Eliminar'
    })),
    pt: (/** @type {?} */ ({
        back: 'Voltar',
        edit: 'Editar',
        remove: 'Remover'
    })),
    ru: (/** @type {?} */ ({
        back: 'возвращение',
        edit: 'редактировать',
        remove: 'удаление'
    }))
};
/**
 * \@description
 *
 * O componente **po-page-detail** é utilizado como container principal para a tela de detalhamento de um registro.
 * Por padrão possui 3 ações, cada ação na tela executa uma função no componente que está utilizando o po-page-detail,
 * são elas:
 *  - Voltar (função: back);
 *  - Editar (função: edit);
 *  - Remover (função: remove);
 *
 * Caso não estiver implementado alguma função, listado anteriormente, o mesmo não será apresentado.
 */
var PoPageDetailBaseComponent = /** @class */ (function () {
    function PoPageDetailBaseComponent() {
    }
    Object.defineProperty(PoPageDetailBaseComponent.prototype, "literals", {
        get: /**
         * @return {?}
         */
        function () {
            return this._literals || poPageDetailLiteralsDefault[browserLanguage()];
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com as literais usadas no `po-page-detail`.
         *
         * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
         *
         * ```
         *  const customLiterals: PoPageDetailLiterals = {
         *    edit: 'Edição',
         *    remove: 'Exclusão',
         *    back: 'Menu'
         *  };
         * ```
         *
         * Ou passando apenas as literais que deseja customizar:
         *
         * ```
         *  const customLiterals: PoPageDetailLiterals = {
         *    remove: 'Excluir registro permanentemente'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-page-detail
         *   [p-literals]="customLiterals">
         * </po-page-detail>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Objeto com as literais usadas no `po-page-detail`.
         *
         * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
         *
         * ```
         *  const customLiterals: PoPageDetailLiterals = {
         *    edit: 'Edição',
         *    remove: 'Exclusão',
         *    back: 'Menu'
         *  };
         * ```
         *
         * Ou passando apenas as literais que deseja customizar:
         *
         * ```
         *  const customLiterals: PoPageDetailLiterals = {
         *    remove: 'Excluir registro permanentemente'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <po-page-detail
         *   [p-literals]="customLiterals">
         * </po-page-detail>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, poPageDetailLiteralsDefault[poLocaleDefault], poPageDetailLiteralsDefault[browserLanguage()], value);
            }
            else {
                this._literals = poPageDetailLiteralsDefault[browserLanguage()];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoPageDetailBaseComponent.prototype, "title", {
        get: /**
         * @return {?}
         */
        function () {
            return this._title;
        },
        /** Título da página. */
        set: /**
         * Título da página.
         * @param {?} title
         * @return {?}
         */
        function (title) {
            this._title = title;
            this.poPageContent.recalculateHeaderSize();
        },
        enumerable: true,
        configurable: true
    });
    PoPageDetailBaseComponent.propDecorators = {
        poPageContent: [{ type: ViewChild, args: [PoPageContentComponent, { static: true },] }],
        breadcrumb: [{ type: Input, args: ['p-breadcrumb',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        title: [{ type: Input, args: ['p-title',] }]
    };
    return PoPageDetailBaseComponent;
}());
export { PoPageDetailBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageDetailBaseComponent.prototype._literals;
    /**
     * @type {?}
     * @private
     */
    PoPageDetailBaseComponent.prototype._title;
    /** @type {?} */
    PoPageDetailBaseComponent.prototype.poPageContent;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    PoPageDetailBaseComponent.prototype.breadcrumb;
}
