// @portinari/portinari-ui/esm5/lib/po.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { PoComponentsModule } from './components/components.module';
import { PoDirectivesModule } from './directives/directives.module';
import { PoGuardsModule } from './guards/guards.module';
import { PoInterceptorsModule } from './interceptors/interceptors.module';
import { PoPipesModule } from './pipes/pipes.module';
import { PoServicesModule } from './services/services.module';
import { PoNotificationService } from './services/po-notification/po-notification.service';
var PoModule = /** @class */ (function () {
    function PoModule() {
    }
    PoModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        PoComponentsModule,
                        PoDirectivesModule,
                        PoGuardsModule,
                        PoInterceptorsModule,
                        PoPipesModule,
                        PoServicesModule
                    ],
                    exports: [
                        PoComponentsModule,
                        PoDirectivesModule,
                        PoInterceptorsModule,
                        PoGuardsModule,
                        PoPipesModule,
                        PoServicesModule
                    ],
                    providers: [PoNotificationService],
                    bootstrap: []
                },] }
    ];
    return PoModule;
}());
export { PoModule };
