// @portinari/portinari-ui/esm5/lib/directives/po-tooltip/po-tooltip.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, HostListener, Renderer2 } from '@angular/core';
import { PoControlPositionService } from './../../services/po-control-position/po-control-position.service';
import { PoTooltipBaseDirective } from './po-tooltip-base.directive';
/**
 * \@docsExtends PoTooltipBaseDirective
 *
 * \@example
 *
 * <example name="po-tooltip-basic" title="Portinari Tooltip Basic" >
 *  <file name="sample-po-tooltip-basic/sample-po-tooltip-basic.component.html"> </file>
 *  <file name="sample-po-tooltip-basic/sample-po-tooltip-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-tooltip-labs" title="Portinari Tooltip Labs" >
 *  <file name="sample-po-tooltip-labs/sample-po-tooltip-labs.component.html"> </file>
 *  <file name="sample-po-tooltip-labs/sample-po-tooltip-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-tooltip-new-user" title="Portinari Tooltip - New User" >
 *  <file name="sample-po-tooltip-new-user/sample-po-tooltip-new-user.component.html"> </file>
 *  <file name="sample-po-tooltip-new-user/sample-po-tooltip-new-user.component.ts"> </file>
 * </example>
 *
 */
var PoTooltipDirective = /** @class */ (function (_super) {
    tslib_1.__extends(PoTooltipDirective, _super);
    function PoTooltipDirective(elementRef, renderer, poControlPosition) {
        var _this = _super.call(this) || this;
        _this.elementRef = elementRef;
        _this.renderer = renderer;
        _this.poControlPosition = poControlPosition;
        _this.tooltipOffset = 8;
        return _this;
    }
    /**
     * @return {?}
     */
    PoTooltipDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initScrollEventListenerFunction();
    };
    /**
     * @return {?}
     */
    PoTooltipDirective.prototype.onMouseEnter = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.tooltip) {
                _this.tooltipContent ? _this.showTooltip() : _this.createTooltip();
                _this.removeArrow(_this.arrowDirection);
                _this.poControlPosition.adjustPosition(_this.tooltipPosition);
                _this.arrowDirection = _this.poControlPosition.getArrowDirection();
                _this.addArrow(_this.arrowDirection);
                _this.lastTooltipText = _this.tooltip;
            }
        }));
    };
    /**
     * @return {?}
     */
    PoTooltipDirective.prototype.onMouseLeave = /**
     * @return {?}
     */
    function () {
        this.hideTooltip();
    };
    /**
     * @private
     * @param {?} arrowDirection
     * @return {?}
     */
    PoTooltipDirective.prototype.addArrow = /**
     * @private
     * @param {?} arrowDirection
     * @return {?}
     */
    function (arrowDirection) {
        this.renderer.addClass(this.divArrow, "po-arrow-" + arrowDirection);
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.addScrollEventListener = /**
     * @private
     * @return {?}
     */
    function () {
        window.addEventListener('scroll', this.eventListenerFunction, true);
    };
    // Monta a estrutura do tooltip
    // Monta a estrutura do tooltip
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.createTooltip = 
    // Monta a estrutura do tooltip
    /**
     * @private
     * @return {?}
     */
    function () {
        this.tooltipContent = this.renderer.createElement('div');
        this.renderer.addClass(this.tooltipContent, 'po-tooltip');
        this.divArrow = this.renderer.createElement('div');
        this.renderer.addClass(this.divArrow, 'po-tooltip-arrow');
        this.divContent = this.renderer.createElement('div');
        this.renderer.addClass(this.divContent, 'po-tooltip-content');
        this.textContent = this.renderer.createText(this.tooltip);
        this.renderer.appendChild(this.divContent, this.textContent);
        this.renderer.appendChild(this.tooltipContent, this.divArrow);
        this.renderer.appendChild(this.tooltipContent, this.divContent);
        this.renderer.appendChild(this.elementRef.nativeElement, this.tooltipContent);
        this.poControlPosition.setElements(this.tooltipContent, this.tooltipOffset, this.elementRef);
        this.addScrollEventListener();
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.initScrollEventListenerFunction = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventListenerFunction = (/**
         * @return {?}
         */
        function () {
            if (!_this.isHidden) {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.poControlPosition.adjustPosition(_this.tooltipPosition);
                }));
            }
        });
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.hideTooltip = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.tooltipContent) {
            this.renderer.addClass(this.tooltipContent, 'po-invisible');
            this.isHidden = true;
            this.removeScrollEventListener();
        }
    };
    /**
     * @private
     * @param {?} arrowDirection
     * @return {?}
     */
    PoTooltipDirective.prototype.removeArrow = /**
     * @private
     * @param {?} arrowDirection
     * @return {?}
     */
    function (arrowDirection) {
        if (this.elementRef.nativeElement.querySelector(".po-arrow-" + arrowDirection)) {
            this.renderer.removeClass(this.divArrow, "po-arrow-" + arrowDirection);
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.removeScrollEventListener = /**
     * @private
     * @return {?}
     */
    function () {
        window.removeEventListener('scroll', this.eventListenerFunction, true);
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.showTooltip = /**
     * @private
     * @return {?}
     */
    function () {
        this.renderer.removeClass(this.tooltipContent, 'po-invisible');
        this.updateTextContent();
        this.isHidden = false;
        this.addScrollEventListener();
    };
    /**
     * @private
     * @return {?}
     */
    PoTooltipDirective.prototype.updateTextContent = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.lastTooltipText !== this.tooltip) {
            this.renderer.removeChild(this.divContent, this.textContent);
            this.textContent = this.renderer.createText(this.tooltip);
            this.renderer.appendChild(this.divContent, this.textContent);
        }
    };
    PoTooltipDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[p-tooltip]',
                    providers: [PoControlPositionService]
                },] }
    ];
    /** @nocollapse */
    PoTooltipDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: PoControlPositionService }
    ]; };
    PoTooltipDirective.propDecorators = {
        onMouseEnter: [{ type: HostListener, args: ['mouseenter',] }],
        onMouseLeave: [{ type: HostListener, args: ['mouseleave',] }]
    };
    return PoTooltipDirective;
}(PoTooltipBaseDirective));
export { PoTooltipDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.arrowDirection;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.divArrow;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.divContent;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.isHidden;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.lastTooltipText;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.textContent;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.tooltipContent;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.tooltipOffset;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.eventListenerFunction;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    PoTooltipDirective.prototype.poControlPosition;
}
