// @portinari/portinari-ui/esm2015/lib/components/po-navbar/po-navbar-items/po-navbar-item/po-navbar-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { isExternalLink } from '../../../../utils/util';
export class PoNavbarItemComponent {
    constructor() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get type() {
        return isExternalLink(this.link) ? 'externalLink' : 'internalLink';
    }
    /**
     * @param {?=} label
     * @param {?=} link
     * @return {?}
     */
    itemClick(label, link) {
        if (this.action) {
            this.action({ label, link });
        }
        this.click.emit();
    }
}
PoNavbarItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-navbar-item',
                template: "<ng-container [ngSwitch]=\"type\">\n\n  <a *ngSwitchCase=\"'externalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [href]=\"link\" \n    target=\"_blank\"\n    (click)=\"itemClick()\">{{ label }}</a>\n\n  <a *ngSwitchCase=\"'internalLink'\"\n    class=\"po-navbar-item-link\"\n    [class.po-clickable]=\"clickable\"\n    [routerLink]=\"link\"\n    (click)=\"itemClick(label, link)\">{{ label }}</a>\n\n</ng-container>\n"
            }] }
];
PoNavbarItemComponent.propDecorators = {
    action: [{ type: Input, args: ['p-action',] }],
    clickable: [{ type: Input, args: ['p-clickable',] }],
    label: [{ type: Input, args: ['p-label',] }],
    link: [{ type: Input, args: ['p-link',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
if (false) {
    /** @type {?} */
    PoNavbarItemComponent.prototype.action;
    /** @type {?} */
    PoNavbarItemComponent.prototype.clickable;
    /** @type {?} */
    PoNavbarItemComponent.prototype.label;
    /** @type {?} */
    PoNavbarItemComponent.prototype.link;
    /** @type {?} */
    PoNavbarItemComponent.prototype.click;
}
