// @portinari/portinari-ui/esm2015/lib/components/po-tabs/po-tab-dropdown/po-tab-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { PoPopoverComponent } from '../../po-popover/po-popover.component';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para agrupamento de componentes `po-tab-button` que os rendereiza em uma um caixa de diálogo
 * no formato de lista.
 */
export class PoTabDropdownComponent {
    constructor() {
        // Lista de abas
        this.tabs = [];
        // Evento que será emitido ao ativar uma aba
        this.activated = new EventEmitter();
        // Evento que será emitido a aba for desabilitada ou ocultada
        this.changeState = new EventEmitter();
        // Evento de click
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    getContainerClass() {
        return this.small ? 'po-tab-dropdown-container-sm' : 'po-tab-dropdown-container';
    }
    /**
     * @return {?}
     */
    togglePopover() {
        this.popover.isHidden ? this.popover.open() : this.popover.close();
    }
}
PoTabDropdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-tab-dropdown',
                template: "<div #tabDrop\n  class=\"po-tab-dropdown-content po-tab-button-md\"\n  [attr.tabindex]=\"0\"\n  [class.po-tab-button-sm]=\"small\"\n  (keyup.enter)=\"togglePopover()\">\n  <div class=\"po-tab-button-label\">\n    {{ label }}\n    <span class=\"po-icon\" [ngClass]=\"(popover.isHidden ? 'po-icon-arrow-down' : 'po-icon-arrow-up')\">\n    </span>\n  </div>\n</div>\n\n<po-popover #popover\n  p-hide-arrow\n  p-position=\"bottom\"\n  [p-target]=\"tabDropdown\">\n\n  <div [ngClass]=\"getContainerClass()\">\n    <po-tab-button *ngFor=\"let tab of tabs\"\n      class=\"po-tab-dropdown-button\"\n      [class.po-tab-dropdown-button-active]=\"tab.active\"\n      [p-active]=\"tab.active\"\n      [p-disabled]=\"tab.disabled\"\n      [p-hide]=\"tab.hide\"\n      [p-id]=\"tab.id\"\n      [p-label]=\"tab.label\"\n      [p-small]=\"small\"\n      (p-activated)=\"activated.emit(tab)\"\n      (p-change-state)=\"changeState.emit(tab)\"\n      (p-click)=\"popover.close();click.emit(tab);\">\n    </po-tab-button>\n  </div>\n\n</po-popover>\n"
            }] }
];
PoTabDropdownComponent.propDecorators = {
    popover: [{ type: ViewChild, args: ['popover', { static: true },] }],
    tabDropdown: [{ type: ViewChild, args: ['tabDrop', { read: ElementRef, static: true },] }],
    label: [{ type: Input, args: ['p-label',] }],
    small: [{ type: Input, args: ['p-small',] }],
    tabs: [{ type: Input, args: ['p-tabs',] }],
    activated: [{ type: Output, args: ['p-activated',] }],
    changeState: [{ type: Output, args: ['p-change-state',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
if (false) {
    /** @type {?} */
    PoTabDropdownComponent.prototype.popover;
    /** @type {?} */
    PoTabDropdownComponent.prototype.tabDropdown;
    /** @type {?} */
    PoTabDropdownComponent.prototype.label;
    /** @type {?} */
    PoTabDropdownComponent.prototype.small;
    /** @type {?} */
    PoTabDropdownComponent.prototype.tabs;
    /** @type {?} */
    PoTabDropdownComponent.prototype.activated;
    /** @type {?} */
    PoTabDropdownComponent.prototype.changeState;
    /** @type {?} */
    PoTabDropdownComponent.prototype.click;
}
