// @portinari/portinari-ui/esm2015/lib/components/po-breadcrumb/po-breadcrumb-item/po-breadcrumb-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que renderiza cada item do po-breadcrumb.
 */
export class PoBreadcrumbItemComponent {
    constructor() {
        // Especifica se item é o link ativo.
        this.itemActive = false;
    }
}
PoBreadcrumbItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-breadcrumb-item',
                template: "<li *ngIf=\"!itemActive\"\n  [class.po-breadcrumb-item]=\"link || action\"\n  [class.po-breadcrumb-item-unclickable]=\"!link && !action\">\n\n  <a *ngIf=\"link\" [routerLink]=\"link\">\n    <label class=\"po-breadcrumb-label\">{{ label }}</label>\n    <div class=\"po-breadcrumb-arrow\"></div>\n  </a>\n\n  <a *ngIf=\"!link && action\" (click)=\"action(label)\">\n    <label class=\"po-breadcrumb-label\">{{ label }}</label>\n    <div class=\"po-breadcrumb-arrow\"></div>\n  </a>\n\n  <a *ngIf=\"!link && !action\">\n    <label class=\"po-breadcrumb-label\">{{ label }}</label>\n    <div class=\"po-breadcrumb-arrow\"></div>\n  </a>\n</li>\n\n<li *ngIf=\"itemActive\" class=\"po-breadcrumb-item-unclickable\">\n  <label class=\"po-breadcrumb-label\">{{ label }}</label>\n</li>\n"
            }] }
];
PoBreadcrumbItemComponent.propDecorators = {
    action: [{ type: Input, args: ['p-action',] }],
    label: [{ type: Input, args: ['p-label',] }],
    link: [{ type: Input, args: ['p-link',] }],
    itemActive: [{ type: Input, args: ['p-item-active',] }]
};
if (false) {
    /** @type {?} */
    PoBreadcrumbItemComponent.prototype.action;
    /** @type {?} */
    PoBreadcrumbItemComponent.prototype.label;
    /** @type {?} */
    PoBreadcrumbItemComponent.prototype.link;
    /** @type {?} */
    PoBreadcrumbItemComponent.prototype.itemActive;
}
