// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-list/po-page-list-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { poLocaleDefault } from './../../../utils/util';
import { PoPageDefaultBaseComponent } from '../po-page-default/po-page-default-base.component';
/** @type {?} */
export const poPageListLiteralsDefault = {
    en: (/** @type {?} */ ({
        otherActions: 'Other actions'
    })),
    es: (/** @type {?} */ ({
        otherActions: 'Otras acciones'
    })),
    pt: (/** @type {?} */ ({
        otherActions: 'Outras ações'
    })),
    ru: (/** @type {?} */ ({
        otherActions: 'Другие действия'
    }))
};
/**
 * \@description
 *
 * O componente `po-page-list` é utilizado como o container principal para as telas de listagem de dados,
 * podendo ser apresentado como lista ou tabela.
 *
 * Este componente possibilita realizar filtro dos dados, no qual permite que seja atribuido uma função que será executada no momento
 * da filtragem. Este comportamento pode ser acionado tanto ao *click* do ícone [po-icon-search](/guides/icons)
 * quanto ao pressionar da tecla *ENTER* quando o foco estiver no campo de pesquisa.
 *
 * Para facilitar a manipulação e visualização dos filtros aplicados, é possível também utilizar o componente
 * [`po-disclaimer-group`](/documentation/po-disclaimer-group).
 * @abstract
 */
export class PoPageListBaseComponent extends PoPageDefaultBaseComponent {
    /**
     * @param {?} languageService
     */
    constructor(languageService) {
        super();
        this.language = languageService.getShortLanguage();
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto que implementa as propriedades da interface `PoDisclaimerGroup`.
     * @param {?} value
     * @return {?}
     */
    set disclaimerGroup(value) {
        if (!value) {
            value = (/** @type {?} */ ({}));
        }
        this._disclaimerGroup = value;
    }
    /**
     * @return {?}
     */
    get disclaimerGroup() {
        return this._disclaimerGroup;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto com as literais usadas no `po-page-list`.
     *
     * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
     *
     * ```
     *  const customLiterals: PoPageListLiterals = {
     *    otherActions: 'Mais ações'
     *  };
     * ```
     *
     * Ou passando apenas as literais que deseja customizar:
     *
     * ```
     *  const customLiterals: PoPageListLiterals = {
     *    otherActions: 'Ações da página'
     *  };
     * ```
     *
     * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
     *
     * ```
     * <po-page-list
     *   [p-literals]="customLiterals">
     * </po-page-list>
     * ```
     *
     * > O valor padrão será traduzido de acordo com o idioma configurado no [`PoI18nService`](/documentation/po-i18n) ou *browser*.
     * @param {?} value
     * @return {?}
     */
    set literals(value) {
        if (value instanceof Object && !(value instanceof Array)) {
            this._literals = Object.assign({}, poPageListLiteralsDefault[poLocaleDefault], poPageListLiteralsDefault[this.language], value);
        }
        else {
            this._literals = poPageListLiteralsDefault[this.language];
        }
    }
    /**
     * @return {?}
     */
    get literals() {
        return this._literals || poPageListLiteralsDefault[this.language];
    }
}
PoPageListBaseComponent.propDecorators = {
    breadcrumb: [{ type: Input, args: ['p-breadcrumb',] }],
    disclaimerGroup: [{ type: Input, args: ['p-disclaimer-group',] }],
    filter: [{ type: Input, args: ['p-filter',] }],
    literals: [{ type: Input, args: ['p-literals',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageListBaseComponent.prototype._disclaimerGroup;
    /**
     * @type {?}
     * @private
     */
    PoPageListBaseComponent.prototype._literals;
    /**
     * @type {?}
     * @protected
     */
    PoPageListBaseComponent.prototype.language;
    /**
     * @type {?}
     * @protected
     */
    PoPageListBaseComponent.prototype.resizeListener;
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto que implementa as propriedades da interface `PoBreadcrumb`.
     * @type {?}
     */
    PoPageListBaseComponent.prototype.breadcrumb;
    /**
     * \@description
     *
     * Objeto que implementa as propriedades da interface `PoPageFilter`.
     * @type {?}
     */
    PoPageListBaseComponent.prototype.filter;
}
