// @portinari/portinari-ui/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { convertToInt } from '../../../../utils/util';
import { PoUploadDragDropAreaComponent } from './po-upload-drag-drop-area/po-upload-drag-drop-area.component';
/** @type {?} */
var PoUploadDragDropHeightDefault = 320;
/** @type {?} */
var PoUploadDragDropHeightMin = 160;
var PoUploadDragDropComponent = /** @class */ (function () {
    function PoUploadDragDropComponent(changeDetector) {
        this.changeDetector = changeDetector;
        this._dragDropHeight = PoUploadDragDropHeightDefault;
        this.isDragOver = false;
        this.fileChange = new EventEmitter();
        this.selectFiles = new EventEmitter();
    }
    Object.defineProperty(PoUploadDragDropComponent.prototype, "dragDropHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dragDropHeight;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var dragDropHeight = convertToInt(value, PoUploadDragDropHeightDefault);
            this._dragDropHeight = (dragDropHeight < PoUploadDragDropHeightMin) ? PoUploadDragDropHeightMin : dragDropHeight;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoUploadDragDropComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.dragDropAreaComponent.focus();
    };
    /**
     * @param {?} element
     * @return {?}
     */
    PoUploadDragDropComponent.prototype.onAreaElement = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        this.areaElement = element;
        // necessário para não ocorrer o erro HasBeenChecked
        this.changeDetector.detectChanges();
    };
    /**
     * @return {?}
     */
    PoUploadDragDropComponent.prototype.onDragLeave = /**
     * @return {?}
     */
    function () {
        this.isDragOver = false;
    };
    /**
     * @return {?}
     */
    PoUploadDragDropComponent.prototype.onDragOver = /**
     * @return {?}
     */
    function () {
        this.isDragOver = true;
    };
    /**
     * @param {?} files
     * @return {?}
     */
    PoUploadDragDropComponent.prototype.onFileChange = /**
     * @param {?} files
     * @return {?}
     */
    function (files) {
        this.isDragOver = false;
        this.fileChange.emit(files);
    };
    PoUploadDragDropComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-upload-drag-drop',
                    template: "<po-upload-drag-drop-area-overlay #dragDropOverlay *ngIf=\"isDragOver\"\r\n  [p-directory-compatible]=\"directoryCompatible\"\r\n  [p-disabled]=\"disabled\"\r\n  [p-literals]=\"literals\"\r\n  [p-target]=\"dragDropAreaComponent.elementRef\"\r\n  (p-area-element)=\"onAreaElement($event)\">\r\n</po-upload-drag-drop-area-overlay>\r\n\r\n<po-upload-drag-drop-area\r\n  p-upload-drag-drop\r\n  [p-area-element]=\"areaElement\"\r\n  [p-directory-compatible]=\"directoryCompatible\"\r\n  [p-disabled]=\"disabled\"\r\n  [p-height]=\"dragDropHeight\"\r\n  [p-literals]=\"literals\"\r\n  [p-overlay]=\"false\"\r\n  (p-drag-leave)=\"onDragLeave()\"\r\n  (p-drag-over)=\"onDragOver()\"\r\n  (p-file-change)=\"onFileChange($event)\"\r\n  (p-select-files)=\"selectFiles.emit()\">\r\n</po-upload-drag-drop-area>\r\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PoUploadDragDropComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    PoUploadDragDropComponent.propDecorators = {
        dragDropOverlayElement: [{ type: ViewChild, args: ['dragDropOverlay', { read: ElementRef, static: false },] }],
        dragDropAreaComponent: [{ type: ViewChild, args: [PoUploadDragDropAreaComponent, { static: true },] }],
        directoryCompatible: [{ type: Input, args: ['p-directory-compatible',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        dragDropHeight: [{ type: Input, args: ['p-drag-drop-height',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        fileChange: [{ type: Output, args: ['p-file-change',] }],
        selectFiles: [{ type: Output, args: ['p-select-files',] }]
    };
    return PoUploadDragDropComponent;
}());
export { PoUploadDragDropComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoUploadDragDropComponent.prototype._dragDropHeight;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.areaElement;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.isDragOver;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.dragDropOverlayElement;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.dragDropAreaComponent;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.directoryCompatible;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.disabled;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.literals;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.fileChange;
    /** @type {?} */
    PoUploadDragDropComponent.prototype.selectFiles;
    /**
     * @type {?}
     * @private
     */
    PoUploadDragDropComponent.prototype.changeDetector;
}
