// @portinari/portinari-ui/esm2015/lib/components/po-field/po-multiselect/po-multiselect-dropdown/po-multiselect-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, HostListener, Input, Output, ViewChild } from '@angular/core';
import { PoMultiselectSearchComponent } from './../po-multiselect-search/po-multiselect-search.component';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que construíra o dropdown, contendo o campo de pesquisa e os itens para seleção.
 */
export class PoMultiselectDropdownComponent {
    constructor() {
        this.scrollTop = 0;
        this.show = false;
        this.haveOptions = false;
        /**
         * Propriedade que indica se o campo de pesquisa deverá ser escondido.
         */
        this.hideSearch = false;
        /**
         * Propriedade que recebe a lista de opções selecionadas.
         */
        this.selectedValues = [];
        /**
         * Propriedade que recebe a lista de opções que deverão ser criadas no dropdown.
         */
        this.options = [];
        /**
         * Evento disparado a cada tecla digitada na pesquisa.
         */
        this.changeSearch = new EventEmitter();
        /**
         * Evento disparado a cada alteração na lista das opções selecionadas.
         */
        this.change = new EventEmitter();
        /**
         * Evento disparado quando for detectada uma ação que necessite fechar o dropdown.
         * Por exemplo, no caso de ser teclado TAB dentro do dropdown, então é disparado este evento
         * para notificar o componente principal que deve fechar o dropdown.
         */
        this.closeDropdown = new EventEmitter();
    }
    /**
     * @param {?} index
     * @return {?}
     */
    scrollTo(index) {
        this.scrollTop = (index <= 2) ? 0 : (index * 44) - 88;
    }
    /**
     * @param {?} option
     * @return {?}
     */
    isSelectedItem(option) {
        return this.selectedValues.some((/**
         * @param {?} selectedItem
         * @return {?}
         */
        selectedItem => selectedItem === option.value));
    }
    /**
     * @param {?} check
     * @param {?} option
     * @return {?}
     */
    clickItem(check, option) {
        this.updateSelectedValues(check, option);
        if (!this.hideSearch) {
            this.searchElement.setFocus();
        }
    }
    /**
     * @param {?} checked
     * @param {?} option
     * @return {?}
     */
    updateSelectedValues(checked, option) {
        if (checked) {
            this.selectedValues.push(option.value);
        }
        else {
            /** @type {?} */
            const indexSelectedValues = this.selectedValues.indexOf(option.value);
            this.selectedValues.splice(indexSelectedValues, 1);
        }
        this.change.emit(this.selectedValues);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    callChangeSearch(event) {
        this.changeSearch.emit(event);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeydown(event) {
        if (event.keyCode === 9) {
            this.closeDropdown.emit();
        }
    }
    /**
     * @param {?} toOpen
     * @return {?}
     */
    controlVisibility(toOpen) {
        this.show = toOpen;
        this.checkInitialOptions();
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (toOpen && this.searchElement && !this.hideSearch) {
                this.searchElement.setFocus();
                this.searchElement.clean();
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    checkInitialOptions() {
        if (this.options.length) {
            this.haveOptions = true;
        }
    }
}
PoMultiselectDropdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-multiselect-dropdown',
                template: "<div #container class=\"po-multiselect-container\" [hidden]=\"!show\">\n\n  <po-multiselect-search #searchElement\n    *ngIf=\"!hideSearch && haveOptions\"\n    [p-literals]=\"literals\"\n    [p-placeholder]=\"placeholderSearch\"\n    (p-change)=\"callChangeSearch($event)\">\n  </po-multiselect-search>\n\n  <ul class=\"po-multiselect-items-container\" [scrollTop]=\"scrollTop\" #ulElement>\n    <div *ngIf=\"!options.length\" class=\"po-multiselect-container-no-data po-text-center\">\n      <span> {{ literals.noData }}</span>\n    </div>\n\n    <po-multiselect-item *ngFor=\"let option of options\"\n      [p-label]=\"option.label\"\n      [p-selected]=\"isSelectedItem(option)\"\n      (p-change)=\"clickItem($event, option)\">\n    </po-multiselect-item>\n  </ul>\n</div>\n"
            }] }
];
PoMultiselectDropdownComponent.propDecorators = {
    hideSearch: [{ type: Input, args: ['p-hide-search',] }],
    literals: [{ type: Input, args: ['p-literals',] }],
    placeholderSearch: [{ type: Input, args: ['p-placeholder-search',] }],
    selectedValues: [{ type: Input, args: ['p-selected-values',] }],
    options: [{ type: Input, args: ['p-options',] }],
    changeSearch: [{ type: Output, args: ['p-change-search',] }],
    change: [{ type: Output, args: ['p-change',] }],
    closeDropdown: [{ type: Output, args: ['p-close-dropdown',] }],
    container: [{ type: ViewChild, args: ['container', { read: ElementRef, static: true },] }],
    ulElement: [{ type: ViewChild, args: ['ulElement', { read: ElementRef, static: true },] }],
    searchElement: [{ type: ViewChild, args: ['searchElement', { static: false },] }],
    onKeydown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.scrollTop;
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.show;
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.haveOptions;
    /**
     * Propriedade que indica se o campo de pesquisa deverá ser escondido.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.hideSearch;
    /**
     * Propriedade que que recebe as literais definidas no componente `po-multiselect`.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.literals;
    /**
     * Placeholder do campo de pesquisa.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.placeholderSearch;
    /**
     * Propriedade que recebe a lista de opções selecionadas.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.selectedValues;
    /**
     * Propriedade que recebe a lista de opções que deverão ser criadas no dropdown.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.options;
    /**
     * Evento disparado a cada tecla digitada na pesquisa.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.changeSearch;
    /**
     * Evento disparado a cada alteração na lista das opções selecionadas.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.change;
    /**
     * Evento disparado quando for detectada uma ação que necessite fechar o dropdown.
     * Por exemplo, no caso de ser teclado TAB dentro do dropdown, então é disparado este evento
     * para notificar o componente principal que deve fechar o dropdown.
     * @type {?}
     */
    PoMultiselectDropdownComponent.prototype.closeDropdown;
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.container;
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.ulElement;
    /** @type {?} */
    PoMultiselectDropdownComponent.prototype.searchElement;
}
