// @portinari/portinari-ui/esm5/lib/components/po-stepper/po-stepper-circle/po-stepper-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
/** @type {?} */
var poMediumStepSize = 32;
/** @type {?} */
var poLargeStepSize = 48;
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
var PoStepperCircleComponent = /** @class */ (function () {
    function PoStepperCircleComponent() {
    }
    Object.defineProperty(PoStepperCircleComponent.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === PoStepperStatus.Active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isDefault", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === PoStepperStatus.Default;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isDisabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === PoStepperStatus.Disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isDone", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === PoStepperStatus.Done;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isError", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === PoStepperStatus.Error;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isLargeStep", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size >= poLargeStepSize;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoStepperCircleComponent.prototype, "isMediumStep", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size >= poMediumStepSize && !this.isLargeStep;
        },
        enumerable: true,
        configurable: true
    });
    PoStepperCircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-stepper-circle',
                    template: "<div class=\"po-stepper-circle\"\n  [class.po-stepper-circle-with-icon]=\"icons\"\n  [style.height.px]=\"size\"\n  [style.width.px]=\"size\"\n  [tabindex]=\"isDisabled ? -1 : 0\">\n\n  <span class=\"po-stepper-circle-content\"\n    [class.po-icon]=\"icons\"\n    [class.po-icon-exclamation]=\"icons && isError\"\n    [class.po-icon-info]=\"icons && (isActive || isDefault || isDisabled)\"\n    [class.po-icon-ok]=\"icons && isDone\"\n    [class.po-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.po-stepper-circle-content-md]=\"isMediumStep\">\n\n    {{ !icons ? content: '' }}\n  </span>\n</div>\n"
                }] }
    ];
    PoStepperCircleComponent.propDecorators = {
        content: [{ type: Input, args: ['p-content',] }],
        icons: [{ type: Input, args: ['p-icons',] }],
        size: [{ type: Input, args: ['p-size',] }],
        status: [{ type: Input, args: ['p-status',] }]
    };
    return PoStepperCircleComponent;
}());
export { PoStepperCircleComponent };
if (false) {
    /** @type {?} */
    PoStepperCircleComponent.prototype.content;
    /** @type {?} */
    PoStepperCircleComponent.prototype.icons;
    /** @type {?} */
    PoStepperCircleComponent.prototype.size;
    /** @type {?} */
    PoStepperCircleComponent.prototype.status;
}
