// @portinari/portinari-ui/esm2015/lib/components/po-table/po-table-subtitle-circle/po-table-subtitle-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoColorPaletteService } from './../../../services/po-color-palette/po-color-palette.service';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para a criação da representação da legenda, em formato de círculo.
 */
export class PoTableSubtitleCircleComponent {
    /**
     * @param {?} poColorPaletteService
     */
    constructor(poColorPaletteService) {
        this.poColorPaletteService = poColorPaletteService;
        /**
         * Esconde a tooltip.
         */
        this.hideTitle = false;
    }
    /**
     * Objeto com os dados da legenda.
     * @param {?} subtitle
     * @return {?}
     */
    set subtitle(subtitle) {
        if (subtitle) {
            subtitle.color = this.poColorPaletteService.getColor(subtitle);
        }
        this._subtitle = subtitle;
    }
    /**
     * @return {?}
     */
    get subtitle() {
        return this._subtitle;
    }
}
PoTableSubtitleCircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-table-subtitle-circle',
                template: "<span\n  [title]=\"hideTitle ? '' : subtitle?.label\"\n  class=\"po-table-subtitle-circle po-{{ subtitle?.color }}\">\n  {{ subtitle?.content }}\n</span>\n"
            }] }
];
/** @nocollapse */
PoTableSubtitleCircleComponent.ctorParameters = () => [
    { type: PoColorPaletteService }
];
PoTableSubtitleCircleComponent.propDecorators = {
    subtitle: [{ type: Input, args: ['p-subtitle',] }],
    hideTitle: [{ type: Input, args: ['p-hide-title',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTableSubtitleCircleComponent.prototype._subtitle;
    /**
     * Esconde a tooltip.
     * @type {?}
     */
    PoTableSubtitleCircleComponent.prototype.hideTitle;
    /**
     * @type {?}
     * @private
     */
    PoTableSubtitleCircleComponent.prototype.poColorPaletteService;
}
