// @portinari/portinari-ui/esm2015/lib/components/po-toolbar/po-toolbar-actions/po-toolbar-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { isTypeof } from '../../../utils/util';
import { PoControlPositionService } from '../../../services/po-control-position/po-control-position.service';
/** @type {?} */
const poToolbarActionsIconDefault = 'po-icon-more';
/**
 * \@docsPrivate
 *
 * \@usedBy PoToolbarComponent
 *
 * \@description
 *
 * O componente `po-toolbar-actions` tem como objetivo receber uma lista de ações e um ícone que podem ser personalizados.
 */
export class PoToolbarActionsComponent {
    constructor() {
        this._actionsIcon = poToolbarActionsIconDefault;
    }
    /**
     * Define o ícone das ações.
     * @param {?} icon
     * @return {?}
     */
    set actionsIcon(icon) {
        this._actionsIcon = isTypeof(icon, 'string') ? icon : poToolbarActionsIconDefault;
    }
    /**
     * @return {?}
     */
    get actionsIcon() {
        return this._actionsIcon;
    }
}
PoToolbarActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-toolbar-actions',
                template: "<div class=\"po-toolbar-actions po-clickable\" (click)=\"popup.toggle()\">\n  <span #toolbarActions class=\"po-icon {{ actionsIcon }} po-toolbar-icon\"></span>\n</div>\n\n<po-popup #popup\n  [p-actions]=\"actions\"\n  [p-target]=\"toolbarActions\">\n</po-popup>\n",
                providers: [PoControlPositionService]
            }] }
];
PoToolbarActionsComponent.propDecorators = {
    actions: [{ type: Input, args: ['p-actions',] }],
    actionsIcon: [{ type: Input, args: ['p-actions-icon',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoToolbarActionsComponent.prototype._actionsIcon;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    PoToolbarActionsComponent.prototype.actions;
}
