// @portinari/portinari-ui/esm5/lib/components/po-field/po-upload/po-upload-drag-drop/po-upload-drag-drop-area/po-upload-drag-drop-area.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
var PoUploadDragDropAreaComponent = /** @class */ (function () {
    function PoUploadDragDropAreaComponent(elementRef) {
        this.elementRef = elementRef;
        this.selectFiles = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PoUploadDragDropAreaComponent.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.selectFilesLinkElement.nativeElement.focus();
    };
    PoUploadDragDropAreaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-upload-drag-drop-area',
                    template: "<div\r\n  class=\"po-upload-drag-drop-area\"\r\n  [ngClass]=\"{'po-upload-drag-drop-area-disabled': disabled}\"\r\n  [style.height.px]=\"height\">\r\n\r\n  <div class=\"po-upload-drag-drop-area-container\">\r\n    <ng-container *ngIf=\"overlay && !disabled; then overlayTemplate; else defaultTemplate\">\r\n    </ng-container>\r\n  </div>\r\n\r\n</div>\r\n\r\n<ng-template #overlayTemplate>\r\n  <span class=\"po-upload-drag-drop-area-overlay-icon po-icon po-icon-upload-cloud\"></span>\r\n\r\n  <div class=\"po-upload-drag-drop-area-overlay-label\">{{ directoryCompatible ? literals?.dropFoldersHere : literals?.dropFilesHere }}</div>\r\n</ng-template>\r\n\r\n<ng-template #defaultTemplate>\r\n  <span class=\"po-upload-drag-drop-area-icon po-icon po-icon-upload-cloud\"></span>\r\n\r\n  <div class=\"po-upload-drag-drop-area-label\">{{ directoryCompatible ? literals?.dragFoldersHere : literals?.dragFilesHere }}</div>\r\n\r\n  <button #selectFilesLink\r\n    class=\"po-upload-drag-drop-area-select-files\"\r\n    [disabled]=\"disabled\"\r\n    [ngClass]=\"{'po-clickable': !disabled}\"\r\n    (click)=\"selectFiles.emit()\">{{ directoryCompatible ? literals?.selectFolderOnComputer : literals?.selectFilesOnComputer }}</button>\r\n</ng-template>\r\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PoUploadDragDropAreaComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    PoUploadDragDropAreaComponent.propDecorators = {
        directoryCompatible: [{ type: Input, args: ['p-directory-compatible',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        height: [{ type: Input, args: ['p-height',] }],
        literals: [{ type: Input, args: ['p-literals',] }],
        overlay: [{ type: Input, args: ['p-overlay',] }],
        selectFiles: [{ type: Output, args: ['p-select-files',] }],
        selectFilesLinkElement: [{ type: ViewChild, args: ['selectFilesLink', { static: false },] }]
    };
    return PoUploadDragDropAreaComponent;
}());
export { PoUploadDragDropAreaComponent };
if (false) {
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.directoryCompatible;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.disabled;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.height;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.literals;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.overlay;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.selectFiles;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.selectFilesLinkElement;
    /** @type {?} */
    PoUploadDragDropAreaComponent.prototype.elementRef;
}
