// @portinari/portinari-ui/esm2015/lib/components/po-accordion/po-accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, QueryList } from '@angular/core';
import { PoAccordionBaseComponent } from './po-accordion-base.component';
import { PoAccordionItemComponent } from './po-accordion-item/po-accordion-item.component';
import { PoAccordionService } from './services/po-accordion.service';
/**
 * \@docsExtends PoAccordionBaseComponent
 *
 * \@example
 *
 * <example name="po-accordion-basic" title="Portinari Accordion Basic" >
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.html"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-labs" title="Portinari Accordion Labs" >
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.html"> </file>
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-faq" title="Portinari Accordion - FAQs" >
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.html"> </file>
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.ts"> </file>
 * </example>
 */
export class PoAccordionComponent extends PoAccordionBaseComponent {
    /**
     * @param {?} accordionService
     */
    constructor(accordionService) {
        super();
        this.accordionService = accordionService;
        this.receiveFromChildAccordionSubscription();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.accordionServiceSubscription.unsubscribe();
    }
    /**
     * @param {?} event
     * @param {?} poAccordionItem
     * @return {?}
     */
    headerToggle(event, poAccordionItem) {
        poAccordionItem.expanded = event;
        this.toggle(poAccordionItem);
    }
    /**
     * @private
     * @return {?}
     */
    receiveFromChildAccordionSubscription() {
        this.accordionServiceSubscription = this.accordionService.receiveFromChildAccordionClicked()
            .subscribe((/**
         * @param {?} poAccordionItem
         * @return {?}
         */
        poAccordionItem => this.toggle(poAccordionItem)));
    }
    /**
     * @private
     * @param {?} poAccordionItem
     * @return {?}
     */
    toggle(poAccordionItem) {
        /** @type {?} */
        const isCurrentAccordionCollapsed = !poAccordionItem.expanded;
        if (isCurrentAccordionCollapsed) {
            this.expandedActiveAccordionItem = null;
            return;
        }
        if (this.expandedActiveAccordionItem) {
            this.expandedActiveAccordionItem.collapse();
        }
        this.expandedActiveAccordionItem = poAccordionItem;
    }
}
PoAccordionComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-accordion',
                template: "<div class=\"po-accordion po-container po-container-no-padding\">\n\n  <section *ngFor=\"let poAccordionItem of poAccordionItems\"\n    class=\"po-accordion-item\"\n    [class.po-accordion-item-active]=\"poAccordionItem.expanded\">\n\n    <po-accordion-item-header\n      [p-expanded]=\"poAccordionItem.expanded\"\n      [p-label]=\"poAccordionItem.label\"\n      (p-toggle)=\"headerToggle($event, poAccordionItem)\">\n    </po-accordion-item-header>\n\n    <po-accordion-item-body [p-expanded]=\"poAccordionItem.expanded\">\n      <ng-container *ngTemplateOutlet=\"poAccordionItem.templateRef\"></ng-container>\n    </po-accordion-item-body>\n  </section>\n\n</div>\n",
                providers: [PoAccordionService]
            }] }
];
/** @nocollapse */
PoAccordionComponent.ctorParameters = () => [
    { type: PoAccordionService }
];
PoAccordionComponent.propDecorators = {
    poAccordionItems: [{ type: ContentChildren, args: [PoAccordionItemComponent,] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.accordionServiceSubscription;
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.expandedActiveAccordionItem;
    /** @type {?} */
    PoAccordionComponent.prototype.poAccordionItems;
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.accordionService;
}
