// @portinari/portinari-ui/esm2015/lib/components/po-progress/po-progress-bar/po-progress-bar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
export class PoProgressBarComponent {
    /**
     * @return {?}
     */
    get valueScale() {
        return `${this.value / 100}`;
    }
}
PoProgressBarComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-progress-bar',
                template: "<div [class.po-progress-bar-indeterminate]=\"indeterminate\">\n  <div class=\"po-progress-bar-element po-progress-bar-primary\" [style.transform]=\"'scaleX('+ valueScale +')'\"></div>\n  <div class=\"po-progress-bar-element po-progress-bar-secondary\"></div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PoProgressBarComponent.propDecorators = {
    indeterminate: [{ type: Input, args: ['p-indeterminate',] }],
    value: [{ type: Input, args: ['p-value',] }]
};
if (false) {
    /** @type {?} */
    PoProgressBarComponent.prototype.indeterminate;
    /** @type {?} */
    PoProgressBarComponent.prototype.value;
}
