// @portinari/portinari-ui/esm2015/lib/components/po-tabs/po-tab/po-tab-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean, uuid } from '../../../utils/util';
/**
 * \@description
 *
 * O componente `po-tab` renderiza uma aba na qual envolve um conteúdo HTML.
 *
 * Com este componente é possível atribuir um rótulo para auxiliar na identificação do conteúdo, ativar para que o mesmo seja exibido,
 * desabilitar para impossibilitar o acesso, bem como ocultar para indisponibilizar a aba.
 *
 * > Para controlar a navegação entre diversas abas, utilizar o componente [`po-tabs`](/documentation/po-tabs).
 * @abstract
 */
export class PoTabBaseComponent {
    constructor() {
        this._active = false;
        this._disabled = false;
        this._hide = false;
        // ID da aba
        this.id = uuid();
        /**
         * Método disparado ao clicar na aba.
         */
        this.click = new EventEmitter();
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Ativa a aba exibindo seu conteúdo.
     *
     * > Sugere-se utilizar na aba de conteúdo inicial.
     *
     * \@default `false`
     * @param {?} active
     * @return {?}
     */
    set active(active) {
        this._active = convertToBoolean(active);
        this.setDisplayOnActive();
    }
    /**
     * @return {?}
     */
    get active() {
        return this._active;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Desabilita a aba.
     *
     * \@default `false`
     * @param {?} disabled
     * @return {?}
     */
    set disabled(disabled) {
        this._disabled = convertToBoolean(disabled);
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Oculta a aba.
     *
     * \@default `false`
     * @param {?} hide
     * @return {?}
     */
    set hide(hide) {
        this._hide = convertToBoolean(hide);
    }
    /**
     * @return {?}
     */
    get hide() {
        return this._hide;
    }
}
PoTabBaseComponent.propDecorators = {
    active: [{ type: Input, args: ['p-active',] }],
    disabled: [{ type: Input, args: ['p-disabled',] }],
    hide: [{ type: Input, args: ['p-hide',] }],
    label: [{ type: Input, args: ['p-label',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._hide;
    /** @type {?} */
    PoTabBaseComponent.prototype.id;
    /**
     * Rótulo da aba.
     * @type {?}
     */
    PoTabBaseComponent.prototype.label;
    /**
     * Método disparado ao clicar na aba.
     * @type {?}
     */
    PoTabBaseComponent.prototype.click;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    PoTabBaseComponent.prototype.setDisplayOnActive = function () { };
}
