// @portinari/portinari-ui/esm2015/lib/components/po-tabs/po-tab/po-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import { PoTabBaseComponent } from './po-tab-base.component';
/**
 * \@docsExtends PoTabBaseComponent
 */
export class PoTabComponent extends PoTabBaseComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super();
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.setDisplayOnActive();
    }
    /**
     * @protected
     * @return {?}
     */
    setDisplayOnActive() {
        this.elementRef.nativeElement.style.display = this.active ? '' : 'none';
    }
}
PoTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-tab',
                template: "<ng-content></ng-content>\n"
            }] }
];
/** @nocollapse */
PoTabComponent.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTabComponent.prototype.elementRef;
}
