// @portinari/portinari-ui/esm5/lib/components/po-table/po-table.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ContentChild, ElementRef, IterableDiffers, QueryList, Renderer2, ViewChild, ViewChildren, ViewContainerRef } from '@angular/core';
import { DecimalPipe } from '@angular/common';
import { Router } from '@angular/router';
import { capitalizeFirstLetter, convertToBoolean } from '../../utils/util';
import { PoDateService } from '../../services/po-date/po-date.service';
import { PoPopupComponent } from '../po-popup/po-popup.component';
import { PoTableBaseComponent } from './po-table-base.component';
import { PoTableRowTemplateDirective } from './po-table-row-template/po-table-row-template.directive';
/**
 * \@docsExtends PoTableBaseComponent
 *
 * \@example
 *
 * <example name="po-table-basic" title="Portinari Table Basic">
 *  <file name="sample-po-table-basic/sample-po-table-basic.component.ts"> </file>
 *  <file name="sample-po-table-basic/sample-po-table-basic.component.html"> </file>
 * </example>
 *
 * <example name="po-table-labs" title="Portinari Table Labs">
 *  <file name="sample-po-table-labs/sample-po-table-labs.component.ts"> </file>
 *  <file name="sample-po-table-labs/sample-po-table-labs.component.html"> </file>
 *  <file name="sample-po-table-labs/sample-po-table-labs.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-table-labs/sample-po-table-labs.component.po.ts"> </file>
 *  <file name="sample-po-table-labs/sample-po-table-labs.service.ts"> </file>
 * </example>
 *
 * <example name="po-table-transport" title="Portinari Table - Transport">
 *  <file name="sample-po-table-transport/sample-po-table-transport.component.ts"> </file>
 *  <file name="sample-po-table-transport/sample-po-table-transport.component.html"> </file>
 *  <file name="sample-po-table-transport/sample-po-table-transport.service.ts"> </file>
 * </example>
 *
 * <example name="po-table-airfare" title="Portinari Table - Airfare">
 *  <file name="sample-po-table-airfare/sample-po-table-airfare.component.ts"> </file>
 *  <file name="sample-po-table-airfare/sample-po-table-airfare.component.html"> </file>
 *  <file name="sample-po-table-airfare/sample-po-table-airfare.service.ts"> </file>
 * </example>
 *
 * <example name="po-table-components" title="Portinari Table - Po Field Components">
 *  <file name="sample-po-table-components/sample-po-table-components.component.ts"> </file>
 *  <file name="sample-po-table-components/sample-po-table-components.component.html"> </file>
 *  <file name="sample-po-table-components/sample-po-table-components.service.ts"> </file>
 * </example>
 */
var PoTableComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoTableComponent, _super);
    function PoTableComponent(poDate, differs, viewRef, renderer, changeDetector, decimalPipe, router) {
        var _this = _super.call(this, poDate) || this;
        _this.changeDetector = changeDetector;
        _this.decimalPipe = decimalPipe;
        _this.router = router;
        _this.tableOpacity = 0;
        _this.initialized = false;
        _this.visibleElement = false;
        _this.parentRef = viewRef['_view']['component'];
        _this.differ = differs.find([]).create(null);
        // TODO: #5550 ao remover este listener, no portal, quando as colunas forem fixas não sofrem
        // alteração de largura, pois o ngDoCheck não é executado.
        _this.clickListener = renderer.listen('document', 'click', (/**
         * @return {?}
         */
        function () { }));
        _this.resizeListener = renderer.listen('window', 'resize', (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.debounceResize();
        }));
        return _this;
    }
    Object.defineProperty(PoTableComponent.prototype, "columnManagerTarget", {
        get: /**
         * @return {?}
         */
        function () {
            return this._columnManagerTarget;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._columnManagerTarget = value;
            this.changeDetector.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "detailHideSelect", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var masterDetail = this.getColumnMasterDetail();
            return masterDetail && masterDetail.detail ? masterDetail.detail.hideSelect : false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "displayColumnManagerCell", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.visibleActions.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "firstAction", {
        get: /**
         * @return {?}
         */
        function () {
            return this.visibleActions && this.visibleActions[0];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasCheckboxColumn", {
        get: /**
         * @return {?}
         */
        function () {
            return this.checkbox && this.hasItems() && this.hasMainColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasFooter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.hasItems() && this.hasVisibleSubtitleColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasMainColumns", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.mainColumns.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasMasterDetailColumn", {
        get: /**
         * @return {?}
         */
        function () {
            return this.hasMainColumns &&
                this.hasItems() && !this.hideDetail && (this.getColumnMasterDetail() !== undefined || this.hasRowTemplate);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasRowTemplate", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.tableRowTemplate;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasValidColumns", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.validColumns.length;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "hasVisibleSubtitleColumns", {
        get: /**
         * @return {?}
         */
        function () {
            return this.getSubtitleColumns().some((/**
             * @param {?} column
             * @return {?}
             */
            function (column) { return column.visible !== false; }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "isSingleAction", {
        get: /**
         * @return {?}
         */
        function () {
            return this.visibleActions.length === 1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "mainColumns", {
        // Colunas que são inseridas no <head> da tabela
        get: 
        // Colunas que são inseridas no <head> da tabela
        /**
         * @return {?}
         */
        function () {
            return this.validColumns.filter((/**
             * @param {?} col
             * @return {?}
             */
            function (col) { return col.visible !== false; }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "validColumns", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var typesValid = ['string', 'number', 'boolean', 'date', 'time', 'dateTime', 'currency', 'subtitle', 'link', 'label', 'icon'];
            return this.columns.filter((/**
             * @param {?} col
             * @return {?}
             */
            function (col) { return !col.type || typesValid.includes(col.type); }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTableComponent.prototype, "visibleActions", {
        get: /**
         * @return {?}
         */
        function () {
            return this.actions && this.actions.filter((/**
             * @param {?} action
             * @return {?}
             */
            function (action) { return action.visible !== false; }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoTableComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.initialized = true;
        if (!this.container) {
            this.hideContainer();
        }
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        this.checkChangesItems();
        this.verifyCalculateHeightTableContainer();
        // Permite que os cabeçalhos sejam calculados na primeira vez que o componente torna-se visível,
        // evitando com isso, problemas com Tabs ou Divs que iniciem escondidas.
        if (this.tableWrapperElement.nativeElement.offsetWidth && !this.visibleElement && this.initialized) {
            this.debounceResize();
            this.visibleElement = true;
        }
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeListeners();
    };
    /**
     * @param {?} row
     * @param {?} column
     * @return {?}
     */
    PoTableComponent.prototype.checkDisabled = /**
     * @param {?} row
     * @param {?} column
     * @return {?}
     */
    function (row, column) {
        return column.disabled ? column.disabled(row) : false;
    };
    /**
     * @param {?} row
     * @param {?} tableAction
     * @return {?}
     */
    PoTableComponent.prototype.executeTableAction = /**
     * @param {?} row
     * @param {?} tableAction
     * @return {?}
     */
    function (row, tableAction) {
        if (!row.disabled && !this.validateTableAction(row, tableAction)) {
            tableAction.action.call(this.parentRef, row);
            this.toggleRowAction(row);
        }
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.columnCountForMasterDetail = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var columnManager = 1;
        return (this.mainColumns.length + 1) + (this.actions.length > 0 ? 1 : 0) + (this.checkbox ? 1 : 0) + columnManager;
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.columnCount = /**
     * @return {?}
     */
    function () {
        return (this.mainColumns.length +
            (this.actions.length > 0 ? 1 : 0) +
            (this.checkbox ? 1 : 0) +
            (!this.hideDetail && this.getColumnMasterDetail() !== undefined ? 1 : 0));
    };
    /**
     * @param {?} value
     * @param {?} format
     * @return {?}
     */
    PoTableComponent.prototype.formatNumber = /**
     * @param {?} value
     * @param {?} format
     * @return {?}
     */
    function (value, format) {
        if (!format) {
            return value;
        }
        return this.decimalPipe.transform(value, format);
    };
    /**
     * @param {?} row
     * @param {?} subtitleColumn
     * @return {?}
     */
    PoTableComponent.prototype.getSubtitleColumn = /**
     * @param {?} row
     * @param {?} subtitleColumn
     * @return {?}
     */
    function (row, subtitleColumn) {
        return subtitleColumn.subtitles.find((/**
         * @param {?} subtitleItem
         * @return {?}
         */
        function (subtitleItem) { return row[subtitleColumn.property] === subtitleItem.value; }));
    };
    /**
     * @param {?} rowValue
     * @param {?} columnBoolean
     * @return {?}
     */
    PoTableComponent.prototype.getBooleanLabel = /**
     * @param {?} rowValue
     * @param {?} columnBoolean
     * @return {?}
     */
    function (rowValue, columnBoolean) {
        if (rowValue || rowValue === false || rowValue === 0) {
            rowValue = convertToBoolean(rowValue);
            if (columnBoolean.boolean) {
                return rowValue ? columnBoolean.boolean.trueLabel || 'Sim' : columnBoolean.boolean.falseLabel || 'Não';
            }
            else {
                return rowValue ? 'Sim' : 'Não';
            }
        }
        return rowValue;
    };
    /**
     * @param {?} row
     * @param {?} column
     * @return {?}
     */
    PoTableComponent.prototype.getColumnIcons = /**
     * @param {?} row
     * @param {?} column
     * @return {?}
     */
    function (row, column) {
        if (column.icons && column.icons.length) {
            return column.icons;
        }
        /** @type {?} */
        var rowIcons = row[column.property];
        if (rowIcons && rowIcons.length) {
            return this.getColumnIconsFromProperty(rowIcons, column);
        }
    };
    /**
     * @param {?} row
     * @param {?} columnLabel
     * @return {?}
     */
    PoTableComponent.prototype.getColumnLabel = /**
     * @param {?} row
     * @param {?} columnLabel
     * @return {?}
     */
    function (row, columnLabel) {
        return columnLabel.labels.find((/**
         * @param {?} labelItem
         * @return {?}
         */
        function (labelItem) { return row[columnLabel.property] === labelItem.value; }));
    };
    /**
     * @param {?} column
     * @return {?}
     */
    PoTableComponent.prototype.getColumnTitleLabel = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        return column.label || capitalizeFirstLetter(column.property);
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.verifyWidthColumnsPixels = /**
     * @return {?}
     */
    function () {
        return this.hasMainColumns ? this.mainColumns.every((/**
         * @param {?} column
         * @return {?}
         */
        function (column) { return column.width && column.width.includes('px'); })) : false;
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.calculateWidthHeaders = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            if (_this.height) {
                _this.headersTable.forEach((/**
                 * @param {?} header
                 * @return {?}
                 */
                function (header) {
                    /** @type {?} */
                    var divHeader = header.nativeElement.querySelector('.po-table-header-fixed-inner');
                    if (divHeader) {
                        divHeader.style.width = header.nativeElement.offsetWidth + "px";
                    }
                }));
            }
        }));
    };
    /**
     * @param {?} row
     * @return {?}
     */
    PoTableComponent.prototype.containsMasterDetail = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        return row[this.getNameColumnDetail()] && row[this.getNameColumnDetail()].length;
    };
    /**
     * @param {?} row
     * @param {?} index
     * @return {?}
     */
    PoTableComponent.prototype.isShowRowTemplate = /**
     * @param {?} row
     * @param {?} index
     * @return {?}
     */
    function (row, index) {
        if (this.tableRowTemplate && this.tableRowTemplate.poTableRowTemplateShow) {
            return this.tableRowTemplate.poTableRowTemplateShow(row, index);
        }
        return true;
    };
    /**
     * @param {?} row
     * @return {?}
     */
    PoTableComponent.prototype.isShowMasterDetail = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        return !this.hideDetail &&
            this.getNameColumnDetail() &&
            row.$showDetail &&
            this.containsMasterDetail(row) &&
            !this.hasRowTemplate;
    };
    /**
     * @param {?} columns
     * @return {?}
     */
    PoTableComponent.prototype.onVisibleColumnsChange = /**
     * @param {?} columns
     * @return {?}
     */
    function (columns) {
        this.columns = columns;
        this.changeDetector.detectChanges();
    };
    /**
     * @param {?} event
     * @param {?=} column
     * @param {?=} row
     * @return {?}
     */
    PoTableComponent.prototype.tooltipMouseEnter = /**
     * @param {?} event
     * @param {?=} column
     * @param {?=} row
     * @return {?}
     */
    function (event, column, row) {
        this.tooltipText = undefined;
        if (this.hideTextOverflow &&
            event.target.offsetWidth < event.target.scrollWidth &&
            event.target.innerText.trim()) {
            return this.tooltipText = event.target.innerText;
        }
        if (column) {
            this.checkingIfColumnHasTooltip(column, row);
        }
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.tooltipMouseLeave = /**
     * @return {?}
     */
    function () {
        this.tooltipText = undefined;
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.verifyChangeHeightInFooter = /**
     * @return {?}
     */
    function () {
        return this.footerHeight !== this.getHeightTableFooter();
    };
    /**
     * @return {?}
     */
    PoTableComponent.prototype.verifyCalculateHeightTableContainer = /**
     * @return {?}
     */
    function () {
        if (this.height && this.verifyChangeHeightInFooter()) {
            this.footerHeight = this.getHeightTableFooter();
            this.calculateHeightTableContainer(this.height);
        }
    };
    /**
     * @param {?} height
     * @return {?}
     */
    PoTableComponent.prototype.calculateHeightTableContainer = /**
     * @param {?} height
     * @return {?}
     */
    function (height) {
        /** @type {?} */
        var value = parseFloat(height);
        this.heightTableContainer = value ? (value - this.getHeightTableFooter()) : undefined;
        this.setTableOpacity(1);
        this.changeDetector.detectChanges();
    };
    /**
     * @param {?} row
     * @param {?} targetRef
     * @return {?}
     */
    PoTableComponent.prototype.togglePopup = /**
     * @param {?} row
     * @param {?} targetRef
     * @return {?}
     */
    function (row, targetRef) {
        this.popupTarget = targetRef;
        this.changeDetector.detectChanges();
        this.poPopupComponent.toggle(row);
    };
    /**
     * @param {?} row
     * @param {?} tableAction
     * @return {?}
     */
    PoTableComponent.prototype.validateTableAction = /**
     * @param {?} row
     * @param {?} tableAction
     * @return {?}
     */
    function (row, tableAction) {
        if (typeof tableAction.disabled === 'function') {
            return tableAction.disabled.call(this.parentRef, row);
        }
        else {
            return tableAction.disabled;
        }
    };
    /**
     * @protected
     * @param {?} container
     * @return {?}
     */
    PoTableComponent.prototype.showContainer = /**
     * @protected
     * @param {?} container
     * @return {?}
     */
    function (container) {
        /** @type {?} */
        var containerClassList = this.tableContainerElement.nativeElement.firstChild.classList;
        containerClassList.add('po-container');
        container === 'border' ? containerClassList.add('po-container-no-shadow') : containerClassList.remove('po-container-no-shadow');
    };
    /**
     * @private
     * @return {?}
     */
    PoTableComponent.prototype.checkChangesItems = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var changesItems = this.differ.diff(this.items);
        if (changesItems && this.selectAll) {
            this.selectAll = null;
        }
        if (changesItems && !this.hasColumns && this.hasItems()) {
            this.columns = this.getDefaultColumns(this.items[0]);
        }
    };
    /**
     * @private
     * @param {?} column
     * @param {?} row
     * @return {?}
     */
    PoTableComponent.prototype.checkingIfColumnHasTooltip = /**
     * @private
     * @param {?} column
     * @param {?} row
     * @return {?}
     */
    function (column, row) {
        if (column.type === 'link' && column.tooltip && !this.checkDisabled(row, column)) {
            return this.tooltipText = column.tooltip;
        }
        if (column.type === 'label') {
            /** @type {?} */
            var columnLabel = this.getColumnLabel(row, column);
            return this.tooltipText = columnLabel.tooltip;
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoTableComponent.prototype.getHeightTableFooter = /**
     * @private
     * @return {?}
     */
    function () {
        return this.tableFooterElement ? this.tableFooterElement.nativeElement.offsetHeight : 0;
    };
    /**
     * @private
     * @return {?}
     */
    PoTableComponent.prototype.hideContainer = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var containerClassList = this.tableContainerElement.nativeElement.firstChild.classList;
        containerClassList.remove('po-container');
    };
    /**
     * @private
     * @return {?}
     */
    PoTableComponent.prototype.debounceResize = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        clearTimeout(this.timeoutResize);
        this.timeoutResize = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.calculateWidthHeaders();
            // show the table
            _this.setTableOpacity(1);
        }));
    };
    /**
     * @private
     * @param {?} rowIcons
     * @param {?} column
     * @return {?}
     */
    PoTableComponent.prototype.getColumnIconsFromProperty = /**
     * @private
     * @param {?} rowIcons
     * @param {?} column
     * @return {?}
     */
    function (rowIcons, column) {
        /** @type {?} */
        var icons = [];
        var action = column.action, color = column.color, disabled = column.disabled;
        /** @type {?} */
        var icon = { action: action, color: color, disabled: disabled };
        rowIcons.forEach((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return icons.push(typeof value === 'string' ? tslib_1.__assign({}, icon, { value: value }) : value); }));
        return icons;
    };
    /**
     * @private
     * @return {?}
     */
    PoTableComponent.prototype.removeListeners = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.resizeListener) {
            this.resizeListener();
        }
        if (this.clickListener) {
            this.clickListener();
        }
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    PoTableComponent.prototype.setTableOpacity = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.tableOpacity = value;
    };
    PoTableComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-table',
                    template: "<po-container #tableContainer p-no-padding>\n  <div [class.po-table-container-relative]=\"loading\">\n    <div *ngIf=\"loading\" class=\"po-table-overlay\">\n      <po-loading class=\"po-table-overlay-content\" [p-text]=\"literals.loadingData\"></po-loading>\n    </div>\n\n    <div class=\"po-table-main-container\">\n      <div #tableWrapper\n        class=\"po-table-wrapper\"\n        [class.po-table-header-fixed-columns-pixels]=\"verifyWidthColumnsPixels()\"\n        [style.opacity]=\"tableOpacity\">\n\n        <div class=\"po-table-container\" *ngIf=\"height\" [style.height.px]=\"heightTableContainer\">\n          <div class=\"po-table-header-fixed po-table-header\"></div>\n          <div class=\"po-table-container-fixed-inner\">\n            <ng-container *ngTemplateOutlet=\"tableTemplate\"></ng-container>\n          </div>\n        </div>\n\n        <div *ngIf=\"!height\">\n          <ng-container *ngTemplateOutlet=\"tableTemplate\"></ng-container>\n        </div>\n\n      </div>\n    </div>\n  </div>\n\n  <div class=\"po-table-footer\" *ngIf=\"hasFooter\">\n    <div *ngFor=\"let column of getSubtitleColumns()\">\n      <po-table-subtitle-footer [p-literals]=\"literals\" [p-subtitles]=\"column.subtitles\"></po-table-subtitle-footer>\n    </div>\n  </div>\n</po-container>\n\n<div class=\"po-row po-table-footer-show-more\" [class.po-invisible]=\"showMore.observers.length === 0\" #tableFooter>\n  <po-button\n    class=\"po-offset-xl-4 po-offset-lg-4 po-offset-md-3 po-lg-4 po-md-6\"\n    [p-disabled]=\"showMoreDisabled\"\n    [p-label]=\"literals.loadMoreData\"\n    (p-click)=\"onShowMore()\">\n  </po-button>\n</div>\n\n<ng-template #tableTemplate>\n  <table class=\"po-table\"\n    [class.po-table-striped]=\"striped\"\n    [class.po-table-layout-fixed]=\"hideTextOverflow\">\n    <thead>\n      <tr [class.po-table-header]=\"!height\">\n\n        <th *ngIf=\"hasCheckboxColumn\" class=\"po-table-column-checkbox\">\n          <div [class.po-table-header-fixed-inner]=\"height\">\n            <input *ngIf=\"!hideSelectAll\" type=\"checkbox\" class=\"po-table-checkbox\"\n            [class.po-table-checkbox-checked]=\"selectAll\"\n            [class.po-table-checkbox-indeterminate]=\"selectAll === null\">\n            <label *ngIf=\"!hideSelectAll\" class=\"po-table-checkbox-label po-clickable\" (click)=\"selectAllRows()\"></label>\n          </div>\n        </th>\n\n        <th *ngIf=\"hasMasterDetailColumn\"\n          class=\"po-table-header-column po-table-header-master-detail\">\n        </th>\n\n        <th *ngIf=\"!hasMainColumns\" #noColumnsHeader class=\"po-table-header-column po-text-center\">\n          <ng-container *ngIf=\"height; then noColumnsWithHeight; else noColumnsWithoutHeight\"> </ng-container>\n\n          <ng-template #noColumnsWithHeight>\n            <div class=\"po-table-header-fixed-inner\" [style.width.px]=\"noColumnsHeader.offsetWidth\">\n              {{ hasValidColumns ? literals.noVisibleColumn : literals.noColumns }}\n            </div>\n          </ng-template>\n\n          <ng-template #noColumnsWithoutHeight>\n            {{ hasValidColumns ? literals.noVisibleColumn : literals.noColumns }}\n          </ng-template>\n\n        </th>\n\n        <th *ngFor=\"let column of mainColumns; let i = index\" #headersTable\n          class=\"po-table-header-ellipsis\"\n          [style.width]=\"column.width\" [style.max-width]=\"column.width\" [style.min-width]=\"column.width\"\n          [class.po-clickable]=\"sort\"\n          [class.po-table-column-right]= \"column.type === 'currency' || column.type === 'number'\"\n          [class.po-table-header-subtitle]=\"column.type === 'subtitle'\"\n          (click)=\"sortColumn(column)\">\n            <div *ngIf=\"height\" class=\"po-table-header-fixed-inner\">\n              <ng-container *ngTemplateOutlet=\"contentHeaderTemplate\"></ng-container>\n            </div>\n            <div *ngIf=\"!height\">\n              <ng-container *ngTemplateOutlet=\"contentHeaderTemplate\"></ng-container>\n            </div>\n            <ng-template #contentHeaderTemplate>\n              <span *ngIf=\"sort\"\n                [class.po-table-header-icon-unselected]= \"sortedColumn?.property !== column\"\n                [class.po-table-header-icon-descending]= \"sortedColumn?.property === column && sortedColumn.ascending\"\n                [class.po-table-header-icon-ascending]= \"sortedColumn?.property === column && !sortedColumn.ascending\">\n              </span>\n              <span\n                class=\"po-table-header-ellipsis po-table-header-block\"\n                [p-tooltip]=\"tooltipText\"\n                (mouseenter)=\"tooltipMouseEnter($event)\"\n                (mouseleave)=\"tooltipMouseLeave()\">\n                {{ getColumnTitleLabel(column) }}\n              </span>\n            </ng-template>\n        </th>\n\n        <th #columnManager *ngIf=\"hasValidColumns\"\n          [class.po-table-header-column-manager]=\"!isSingleAction\"\n          [class.po-table-header-column-manager-border]=\"!height && container\"\n          [class.po-table-header-single-action]=\"isSingleAction\">\n          <div\n            [class.po-table-header-column-manager-border]=\"height && container\"\n            [class.po-table-header-column-manager-fixed-inner]=\"height\"\n            [style.width.px]=\"height && visibleActions.length ? columnManager.offsetWidth : undefined\">\n            <button #columnManagerTarget\n              class=\"po-table-header-column-manager-button po-icon po-icon-settings po-clickable\"\n              p-tooltip-position=\"left\"\n              [p-tooltip]=\"literals.columnsManager\">\n            </button>\n          </div>\n        </th>\n\n      </tr>\n    </thead>\n\n    <tbody class=\"po-table-group-row\" *ngIf=\"!hasItems() || !hasMainColumns\">\n      <tr class=\"po-table-row\">\n        <td [colSpan]=\"columnCount()\" class=\"po-table-no-data po-text-center\">\n          <span> {{ literals.noData }} </span>\n        </td>\n      </tr>\n    </tbody>\n\n    <ng-container *ngIf=\"hasMainColumns\">\n      <tbody class=\"po-table-group-row\" *ngFor=\"let row of items, let rowIndex = index;\">\n        <tr class=\"po-table-row\" [class.po-table-row-active]=\"row.$selected || row.$selected === null && checkbox\">\n          <td *ngIf=\"checkbox\" class=\"po-table-column po-table-column-checkbox\">\n            <ng-container *ngIf=\"singleSelect; then inputRadio; else inputCheckbox\"> </ng-container>\n\n            <ng-template #inputRadio>\n              <input type=\"radio\" class=\"po-radio-group-input\" [class.po-radio-group-input-checked]=\"row.$selected\">\n              <label class=\"po-radio-group-label po-clickable\" (click)=\"checkbox ? selectRow(row) : 'javascript:;'\"></label>\n            </ng-template>\n\n            <ng-template #inputCheckbox>\n              <input type=\"checkbox\" class=\"po-table-checkbox\" [class.po-table-checkbox-checked]=\"row.$selected\">\n              <label class=\"po-table-checkbox-label po-clickable\" (click)=\"checkbox ? selectRow(row) : 'javascript:;'\"></label>\n            </ng-template>\n\n          </td>\n          <td *ngIf=\"(getColumnMasterDetail() !== undefined) && !hideDetail || hasRowTemplate\"\n            class=\"po-table-column-detail-toggle\"\n            (click)=\"row.$showDetail = !row.$showDetail\">\n            <span *ngIf=\"(containsMasterDetail(row) && !hasRowTemplate) || isShowRowTemplate(row, rowIndex) && hasRowTemplate\"\n              class=\"po-icon po-clickable\"\n              [class.po-icon-arrow-up]=\"row.$showDetail\"\n              [class.po-icon-arrow-down]=\"!row.$showDetail\">\n            </span>\n          </td>\n\n          <td *ngFor=\"let column of mainColumns, let columnIndex = index;\"\n            [style.width]=\"column.width\" [style.max-width]=\"column.width\" [style.min-width]=\"column.width\"\n            [class.po-table-column]=\"column.type !== 'icon'\"\n            [class.po-table-column-right]=\"column.type == 'currency' || column.type == 'number'\"\n            [class.po-table-column-center]=\"column.type === 'subtitle'\"\n            [class.po-table-column-icons]=\"column.type === 'icon'\"\n            [ngClass]=\"getClassColor(row, column)\"\n            (click)=\"checkbox ? selectRow(row) : 'javascript:;'\">\n\n            <div class=\"po-table-column-cell\"\n              [class.po-table-body-ellipsis]=\"hideTextOverflow\"\n              [ngSwitch]=\"column.type\"\n              [p-tooltip]=\"tooltipText\"\n              (mouseenter)=\"tooltipMouseEnter($event, column, row)\"\n              (mouseleave)=\"tooltipMouseLeave()\">\n              <span *ngSwitchCase=\"'boolean'\">{{ getBooleanLabel(row[column.property], column) }}</span>\n              <span *ngSwitchCase=\"'currency'\">{{ row[column.property] | currency:column.format:'symbol':'1.2-2' }}</span>\n              <span *ngSwitchCase=\"'date'\" >{{ row[column.property] | date: column.format || 'dd/MM/yyyy' }}</span>\n              <span *ngSwitchCase=\"'time'\" >{{ row[column.property] | po_time: column.format || 'HH:mm:ss.ffffff' }}</span>\n              <span *ngSwitchCase=\"'dateTime'\" >{{ row[column.property] | date: column.format || 'dd/MM/yyyy HH:mm:ss' }}</span>\n              <span *ngSwitchCase=\"'number'\">{{ formatNumber(row[column.property], column.format) }}</span>\n              <po-table-column-link *ngSwitchCase=\"'link'\"\n                [p-action]=\"column.action\"\n                [p-disabled]=\"checkDisabled(row, column)\"\n                [p-link]=\"row[column.link]\"\n                [p-row]=\"row\"\n                [p-value]=\"row[column.property]\">\n              </po-table-column-link>\n\n              <po-table-column-icon *ngSwitchCase=\"'icon'\"\n                [p-icons]=\"getColumnIcons(row, column)\"\n                [p-row]=\"row\">\n              </po-table-column-icon>\n\n              <span *ngSwitchCase=\"'subtitle'\">\n                <po-table-subtitle-circle [p-subtitle]=\"getSubtitleColumn(row, column)\"></po-table-subtitle-circle>\n              </span>\n              <span *ngSwitchCase=\"'label'\">\n                <po-table-column-label [p-value]=\"getColumnLabel(row, column)\"></po-table-column-label>\n              </span>\n              <span *ngSwitchDefault>{{ row[column.property] }}</span>\n            </div>\n          </td>\n          <td *ngIf=\"isSingleAction\" class=\"po-table-column po-table-column-single-action\">\n            <div *ngIf=\"firstAction.visible !== false\"\n              class=\"po-table-single-action po-clickable\"\n              [class.po-table-action-disabled]= \"firstAction.disabled ? validateTableAction(row, firstAction) : false\"\n              (click)=\"executeTableAction(row, firstAction)\">\n\n              <span *ngIf=\"firstAction.icon\" class=\"po-icon {{ firstAction.icon }}\"></span>\n              {{ firstAction.label }}\n            </div>\n          </td>\n          <td *ngIf=\"visibleActions.length > 1\" class=\"po-table-column-actions\">\n            <span #popupTarget class=\"po-icon po-icon-more po-clickable\" (click)=\"togglePopup(row, popupTarget)\">\n            </span>\n          </td>\n          <!-- Column Manager -->\n          <td *ngIf=\"displayColumnManagerCell\" class=\"po-table-column\"></td>\n        </tr>\n\n        <tr *ngIf=\"hasMainColumns && hasRowTemplate && row.$showDetail && isShowRowTemplate(row, rowIndex)\">\n          <td class=\"po-table-row-template-container\" [colSpan]=\"columnCountForMasterDetail()\">\n\n            <ng-template\n              [ngTemplateOutlet]=\"tableRowTemplate.templateRef\"\n              [ngTemplateOutletContext]=\"{ $implicit: row, rowIndex: rowIndex }\">\n            </ng-template>\n\n          </td>\n        </tr>\n\n        <tr *ngIf=\"hasMainColumns && isShowMasterDetail(row)\">\n          <td class=\"po-table-column-detail\" [colSpan]=\"columnCountForMasterDetail()\">\n\n            <po-table-detail\n              [p-checkbox]=\"checkbox && !detailHideSelect\"\n              [p-detail]=\"getColumnMasterDetail().detail\"\n              [p-items]=\"row[getNameColumnDetail()]\"\n              (p-select-row)=\"selectDetailRow($event)\">\n            </po-table-detail>\n\n          </td>\n        </tr>\n      </tbody>\n    </ng-container>\n  </table>\n</ng-template>\n\n<po-popup #popup\n  [p-actions]=\"actions\"\n  [p-target]=\"popupTarget\">\n</po-popup>\n\n<po-table-column-manager\n  [p-columns]=\"columns\"\n  [p-max-columns]=\"maxColumns\"\n  [p-target]=\"columnManagerTarget\"\n  (p-visible-columns-change)=\"onVisibleColumnsChange($event)\">\n</po-table-column-manager>\n",
                    providers: [PoDateService]
                }] }
    ];
    /** @nocollapse */
    PoTableComponent.ctorParameters = function () { return [
        { type: PoDateService },
        { type: IterableDiffers },
        { type: ViewContainerRef },
        { type: Renderer2 },
        { type: ChangeDetectorRef },
        { type: DecimalPipe },
        { type: Router }
    ]; };
    PoTableComponent.propDecorators = {
        tableRowTemplate: [{ type: ContentChild, args: [PoTableRowTemplateDirective, { static: true },] }],
        poPopupComponent: [{ type: ViewChild, args: ['popup', { static: false },] }],
        columnManagerTarget: [{ type: ViewChild, args: ['columnManagerTarget', { static: false },] }],
        tableContainerElement: [{ type: ViewChild, args: ['tableContainer', { read: ElementRef, static: true },] }],
        tableFooterElement: [{ type: ViewChild, args: ['tableFooter', { read: ElementRef, static: true },] }],
        tableWrapperElement: [{ type: ViewChild, args: ['tableWrapper', { read: ElementRef, static: true },] }],
        actionsIconElement: [{ type: ViewChildren, args: ['actionsIconElement', { read: ElementRef },] }],
        actionsElement: [{ type: ViewChildren, args: ['actionsElement', { read: ElementRef },] }],
        headersTable: [{ type: ViewChildren, args: ['headersTable',] }]
    };
    return PoTableComponent;
}(PoTableBaseComponent));
export { PoTableComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype._columnManagerTarget;
    /** @type {?} */
    PoTableComponent.prototype.heightTableContainer;
    /** @type {?} */
    PoTableComponent.prototype.parentRef;
    /** @type {?} */
    PoTableComponent.prototype.popupTarget;
    /** @type {?} */
    PoTableComponent.prototype.tableOpacity;
    /** @type {?} */
    PoTableComponent.prototype.tooltipText;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.differ;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.footerHeight;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.initialized;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.timeoutResize;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.visibleElement;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.clickListener;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.resizeListener;
    /** @type {?} */
    PoTableComponent.prototype.tableRowTemplate;
    /** @type {?} */
    PoTableComponent.prototype.poPopupComponent;
    /** @type {?} */
    PoTableComponent.prototype.tableContainerElement;
    /** @type {?} */
    PoTableComponent.prototype.tableFooterElement;
    /** @type {?} */
    PoTableComponent.prototype.tableWrapperElement;
    /** @type {?} */
    PoTableComponent.prototype.actionsIconElement;
    /** @type {?} */
    PoTableComponent.prototype.actionsElement;
    /** @type {?} */
    PoTableComponent.prototype.headersTable;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.changeDetector;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.decimalPipe;
    /**
     * @type {?}
     * @private
     */
    PoTableComponent.prototype.router;
}
