// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-default/po-page-default-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, ViewChild } from '@angular/core';
import { PoPageContentComponent } from '../po-page-content/po-page-content.component';
/**
 * \@description
 *
 * O componente `po-page-default` é utilizado como o container principal para as telas sem um template definido.
 * @abstract
 */
export class PoPageDefaultBaseComponent {
    constructor() {
        this._actions = [];
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Nesta propriedade deve ser definido um array de objetos que implementam a interface `PoPageAction`.
     * @param {?} actions
     * @return {?}
     */
    set actions(actions) {
        this._actions = Array.isArray(actions) ? actions : [];
        this.setDropdownActions();
    }
    /**
     * @return {?}
     */
    get actions() {
        return this._actions;
    }
    /**
     * Título da página.
     * @param {?} title
     * @return {?}
     */
    set title(title) {
        this._title = title;
        this.poPageContent.recalculateHeaderSize();
    }
    /**
     * @return {?}
     */
    get title() {
        return this._title;
    }
}
PoPageDefaultBaseComponent.propDecorators = {
    poPageContent: [{ type: ViewChild, args: [PoPageContentComponent, { static: true },] }],
    actions: [{ type: Input, args: ['p-actions',] }],
    breadcrumb: [{ type: Input, args: ['p-breadcrumb',] }],
    title: [{ type: Input, args: ['p-title',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoPageDefaultBaseComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    PoPageDefaultBaseComponent.prototype._title;
    /** @type {?} */
    PoPageDefaultBaseComponent.prototype.poPageContent;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    PoPageDefaultBaseComponent.prototype.breadcrumb;
    /**
     * @abstract
     * @return {?}
     */
    PoPageDefaultBaseComponent.prototype.setDropdownActions = function () { };
}
