// @portinari/portinari-ui/esm2015/lib/components/po-button/po-button-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
/** @type {?} */
const PO_BUTTON_TYPES = ['default', 'primary', 'danger', 'link'];
/** @type {?} */
const PO_BUTTON_TYPE_DEFAULT = 'default';
/**
 * \@description
 *
 * O `po-button` permite que o usuário execute ações predefinidas pelo desenvolvedor.
 *
 * Através dos tipos, é possível identificar a importância de cada ação, sendo ela primária (`primary`) ou até mesmo uma
 * ação irreversível (`danger`), como a exclusão de um registro.
 *
 * #### Boas práticas
 *
 * - Evite `labels` extensos que quebram o layout do `po-button`, use `labels` diretos, curtos e intuitivos.
 * - Utilize apenas um `po-button` configurado como `primary` por página.
 * - Para ações irreversíveis use sempre o tipo `danger`.
 */
export class PoButtonBaseComponent {
    constructor() {
        this._disabled = false;
        this._loading = false;
        this._small = false;
        this._type = 'default';
        /**
         * Ação que será executada quando o usuário clicar sobre o `po-button`.
         */
        this.click = new EventEmitter();
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Exibe um ícone de carregamento à esquerda do _label_ do botão.
     *
     * > Quando esta propriedade estiver habilitada, desabilitará o botão.
     *
     * \@default `false`
     * @param {?} value
     * @return {?}
     */
    set loading(value) {
        this._loading = convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get loading() {
        return this._loading;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Deixa o botão menor.
     *
     * \@default `false`
     * @param {?} value
     * @return {?}
     */
    set small(value) {
        this._small = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get small() {
        return this._small;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Define o estilo do `po-button`.
     *
     * Valore válidos:
     *  - `default`: estilo padrão do `po-button`.
     *  - `primary`: deixa o `po-button` com destaque, deve ser usado para ações primárias.
     *  - `danger`: deve ser usado para ações que o usuário precisa ter cuidado ao executa-lá.
     *  - `link`: o `po-button` recebe o estilo de um link.
     *
     * \@default `default`
     * @param {?} value
     * @return {?}
     */
    set type(value) {
        this._type = PO_BUTTON_TYPES.includes(value) ? value : PO_BUTTON_TYPE_DEFAULT;
    }
    /**
     * @return {?}
     */
    get type() {
        return this._type;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Desabilita o `po-button` e não permite que o usuário interaja com o mesmo.
     *
     * \@default `false`
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
}
PoButtonBaseComponent.propDecorators = {
    label: [{ type: Input, args: ['p-label',] }],
    loading: [{ type: Input, args: ['p-loading',] }],
    icon: [{ type: Input, args: ['p-icon',] }],
    small: [{ type: Input, args: ['p-small',] }],
    type: [{ type: Input, args: ['p-type',] }],
    disabled: [{ type: Input, args: ['p-disabled',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._loading;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._small;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._type;
    /**
     * \@optional
     *
     * \@description
     *
     * Label do botão.
     * @type {?}
     */
    PoButtonBaseComponent.prototype.label;
    /**
     * Ícone exibido ao lado esquerdo do label do botão.
     *
     * É possível usar qualquer uma dos ícones da [Biblioteca de ícones](/guides/icons).
     * @type {?}
     */
    PoButtonBaseComponent.prototype.icon;
    /**
     * Ação que será executada quando o usuário clicar sobre o `po-button`.
     * @type {?}
     */
    PoButtonBaseComponent.prototype.click;
}
