// @portinari/portinari-ui/esm5/lib/components/po-field/po-clean/po-clean-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, EventEmitter, Input, Output } from '@angular/core';
/**
 * \@docsPrivate
 *
 * Este componente é de uso interno utilizado por componentes de entrada de dados com o objetivo de resetar as informações do model.
 *
 * Por padrão limpa o valor do campo e executa o método onChangePropagate, caso tenha a necessidade de tratar a função de limpar o campo,
 * deve implementar a interface PoClean.
 * @abstract
 */
var PoCleanBaseComponent = /** @class */ (function () {
    function PoCleanBaseComponent() {
        /**
         * Valor que será atribuído ao campo quando for clicado no po-clean.
         */
        this.defaultValue = '';
        /**
         * Evento disparado quando executada ação do po-clean.
         * Este evento deve ser usado para avisar para o componente que está usando o po-clean, que o botão foi disparado,
         * e provavelmente será preciso emitir o evento para atualizar o model.
         */
        this.changeEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PoCleanBaseComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.setInputValue(this.defaultValue);
        this.changeEvent.emit(this.defaultValue);
    };
    /**
     * @return {?}
     */
    PoCleanBaseComponent.prototype.showIcon = /**
     * @return {?}
     */
    function () {
        return this.defaultValue !== this.getInputValue() && this.hasCleanAttr() && !this.isDisabled() && !this.isReadonly();
    };
    // Este método verifica se o componente pai possui a propriedade clean diferente de vazio,
    // ou seja, se o po-clean deve ser usado.
    // Este método verifica se o componente pai possui a propriedade clean diferente de vazio,
    // ou seja, se o po-clean deve ser usado.
    /**
     * @private
     * @return {?}
     */
    PoCleanBaseComponent.prototype.hasCleanAttr = 
    // Este método verifica se o componente pai possui a propriedade clean diferente de vazio,
    // ou seja, se o po-clean deve ser usado.
    /**
     * @private
     * @return {?}
     */
    function () {
        return this.parentComponent.clean;
    };
    // Este método verifica se o componente pai está desabilitado.
    // Este método verifica se o componente pai está desabilitado.
    /**
     * @private
     * @return {?}
     */
    PoCleanBaseComponent.prototype.isDisabled = 
    // Este método verifica se o componente pai está desabilitado.
    /**
     * @private
     * @return {?}
     */
    function () {
        return this.parentComponent.disabled;
    };
    // Este método verifica se o componente pai está somente leitura.
    // Este método verifica se o componente pai está somente leitura.
    /**
     * @private
     * @return {?}
     */
    PoCleanBaseComponent.prototype.isReadonly = 
    // Este método verifica se o componente pai está somente leitura.
    /**
     * @private
     * @return {?}
     */
    function () {
        return this.parentComponent.readonly;
    };
    PoCleanBaseComponent.propDecorators = {
        inputRef: [{ type: Input, args: ['p-element-ref',] }],
        defaultValue: [{ type: Input, args: ['p-default-value',] }],
        changeEvent: [{ type: Output, args: ['p-change-event',] }]
    };
    return PoCleanBaseComponent;
}());
export { PoCleanBaseComponent };
if (false) {
    /**
     * Nesta propriedade deve-se informar o elementRef do campo de entrada que utilizará o po-clean.
     * @type {?}
     */
    PoCleanBaseComponent.prototype.inputRef;
    /**
     * Valor que será atribuído ao campo quando for clicado no po-clean.
     * @type {?}
     */
    PoCleanBaseComponent.prototype.defaultValue;
    /**
     * Evento disparado quando executada ação do po-clean.
     * Este evento deve ser usado para avisar para o componente que está usando o po-clean, que o botão foi disparado,
     * e provavelmente será preciso emitir o evento para atualizar o model.
     * @type {?}
     */
    PoCleanBaseComponent.prototype.changeEvent;
    /**
     * @type {?}
     * @protected
     */
    PoCleanBaseComponent.prototype.parentComponent;
    /**
     * @abstract
     * @param {?} value
     * @return {?}
     */
    PoCleanBaseComponent.prototype.setInputValue = function (value) { };
    /**
     * @abstract
     * @return {?}
     */
    PoCleanBaseComponent.prototype.getInputValue = function () { };
}
