// @portinari/portinari-ui/esm5/lib/components/po-menu/po-menu-filter/po-menu-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que apresenta o campo de pesquisa no po-menu.
 */
var PoMenuFilterComponent = /** @class */ (function () {
    function PoMenuFilterComponent() {
        // Variável necessária para o po-clean identificar que deve ser criado.
        this.clean = true;
        this.filter = new EventEmitter();
    }
    /**
     * @param {?} search
     * @return {?}
     */
    PoMenuFilterComponent.prototype.filterItems = /**
     * @param {?} search
     * @return {?}
     */
    function (search) {
        this.filter.emit(search);
    };
    PoMenuFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-menu-filter',
                    template: "<div class=\"po-menu-filter-container\">\n\n  <input #inputFilter\n    type=\"text\"\n    class=\"po-menu-filter\"\n    placeholder=\"Pesquisar\"\n    (keyup)=\"filterItems(inputFilter.value)\">\n\n  <div class=\"po-menu-filter-search-icon-container\">\n    <span *ngIf=\"!loading\" class=\"po-icon po-menu-filter-icon po-icon-search\"></span>\n    <po-loading-icon *ngIf=\"loading\" ></po-loading-icon>\n  </div>\n  <div class=\"po-menu-filter-close-icon-container\">\n    <po-clean [p-element-ref]=\"inputFilterElement\" (p-change-event)=\"filterItems(inputFilter.value)\"></po-clean>\n  </div>\n\n</div>\n"
                }] }
    ];
    PoMenuFilterComponent.propDecorators = {
        loading: [{ type: Input, args: ['p-loading',] }],
        inputFilterElement: [{ type: ViewChild, args: ['inputFilter', { read: ElementRef, static: true },] }],
        filter: [{ type: Output, args: ['p-filter',] }]
    };
    return PoMenuFilterComponent;
}());
export { PoMenuFilterComponent };
if (false) {
    /** @type {?} */
    PoMenuFilterComponent.prototype.clean;
    /** @type {?} */
    PoMenuFilterComponent.prototype.loading;
    /** @type {?} */
    PoMenuFilterComponent.prototype.inputFilterElement;
    /** @type {?} */
    PoMenuFilterComponent.prototype.filter;
}
