// @portinari/portinari-ui/esm5/lib/components/po-chart/po-chart-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToInt, isTypeof } from '../../utils/util';
import { PoChartType } from './enums/po-chart-type.enum';
/** @type {?} */
var poChartDefaultHeight = 400;
/** @type {?} */
var poChartMinHeight = 200;
/** @type {?} */
var poChartTypeDefault = PoChartType.Pie;
/**
 * \@description
 *
 * O `po-chart` é um componente para renderização de dados através de gráficos, com isso facilitando a compreensão e tornando a
 * visualização destes dados mais agradável.
 *
 * Este componente também possibilita a definição das seguintes propriedades:
 *  - altura
 *  - series
 *  - tipo
 *  - título
 *
 * Além das definições de propriedades, também é possível definir uma ação que será executada ao clicar em determinado elemento do gráfico
 * e outra que será executada ao passar o *mouse* sobre o elemento.
 *
 * #### Boas práticas
 *
 * - Para que o gráfico não fique ilegível e incompreensível, evite uma quantia excessiva de séries.
 *
 * @abstract
 */
var PoChartBaseComponent = /** @class */ (function () {
    function PoChartBaseComponent() {
        this._height = poChartDefaultHeight;
        this._type = poChartTypeDefault;
        this.poChartType = PoChartType;
        /**
         * Evento executado quando o usuário clicar sobre um elemento do gráfico.
         *
         * > Será passado por parâmetro um objeto contendo a categoria e valor da série.
         */
        this.seriesClick = new EventEmitter();
        /**
         * Evento executado quando o usuário passar o *mouse* sobre um elemento do gráfico.
         *
         * > Será passado por parâmetro um objeto contendo a categoria e valor da série.
         */
        this.seriesHover = new EventEmitter();
    }
    Object.defineProperty(PoChartBaseComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this._height;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define a altura do gráfico.
         *
         * > O valor mínimo que pode ser informado é 200.
         *
         * @default `400px`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define a altura do gráfico.
         *
         * > O valor mínimo que pode ser informado é 200.
         *
         * \@default `400px`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var intValue = convertToInt(value);
            /** @type {?} */
            var height;
            if (isTypeof(value, 'number')) {
                height = intValue <= poChartMinHeight ? poChartMinHeight : intValue;
            }
            else {
                height = poChartDefaultHeight;
            }
            this._height = height;
            this.rebuildComponent();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoChartBaseComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this._type;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define o tipo de gráfico.
         *
         * > Veja os valores válidos no *enum* `PoChartType`.
         *
         * @default `PoChartType.Pie`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define o tipo de gráfico.
         *
         * > Veja os valores válidos no *enum* `PoChartType`.
         *
         * \@default `PoChartType.Pie`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._type = ((/** @type {?} */ (Object))).values(PoChartType).includes(value) ? value : poChartTypeDefault;
            this.rebuildComponent();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    PoChartBaseComponent.prototype.onSeriesClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.seriesClick.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PoChartBaseComponent.prototype.onSeriesHover = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.seriesHover.emit(event);
    };
    PoChartBaseComponent.propDecorators = {
        height: [{ type: Input, args: ['p-height',] }],
        series: [{ type: Input, args: ['p-series',] }],
        title: [{ type: Input, args: ['p-title',] }],
        type: [{ type: Input, args: ['p-type',] }],
        seriesClick: [{ type: Output, args: ['p-series-click',] }],
        seriesHover: [{ type: Output, args: ['p-series-hover',] }]
    };
    return PoChartBaseComponent;
}());
export { PoChartBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoChartBaseComponent.prototype._height;
    /**
     * @type {?}
     * @private
     */
    PoChartBaseComponent.prototype._type;
    /** @type {?} */
    PoChartBaseComponent.prototype.poChartType;
    /**
     * \@description
     *
     * Define os elementos do gráfico que serão criados dinamicamente.
     *
     * > A coleção de objetos deve implementar alguma das interfaces abaixo:
     * - `PoDonutChartSeries`
     * - `PoPieChartSeries`
     * @type {?}
     */
    PoChartBaseComponent.prototype.series;
    /**
     * Define o título do gráfico.
     * @type {?}
     */
    PoChartBaseComponent.prototype.title;
    /**
     * Evento executado quando o usuário clicar sobre um elemento do gráfico.
     *
     * > Será passado por parâmetro um objeto contendo a categoria e valor da série.
     * @type {?}
     */
    PoChartBaseComponent.prototype.seriesClick;
    /**
     * Evento executado quando o usuário passar o *mouse* sobre um elemento do gráfico.
     *
     * > Será passado por parâmetro um objeto contendo a categoria e valor da série.
     * @type {?}
     */
    PoChartBaseComponent.prototype.seriesHover;
    /**
     * @abstract
     * @return {?}
     */
    PoChartBaseComponent.prototype.rebuildComponent = function () { };
}
