// @portinari/portinari-ui/esm5/lib/components/po-slide/po-slide-item/po-slide-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { isIEOrEdge } from '../../../utils/util';
import { PoSlideContentTemplateDirective } from '../directives/po-slide-content-template.directive';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para cada item do po-slide.
 */
var PoSlideItemComponent = /** @class */ (function () {
    function PoSlideItemComponent() {
        this.isIEOrEdge = isIEOrEdge();
    }
    /**
     * @return {?}
     */
    PoSlideItemComponent.prototype.setLinkType = /**
     * @return {?}
     */
    function () {
        if (!this.template && this.link) {
            return this.link.startsWith('http') ? 'externalLink' : 'internalLink';
        }
        return 'noLink';
    };
    PoSlideItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-slide-item',
                    template: "<div #slideItem\n  class=\"po-slide-item\"\n  [class.po-slide-item-background-image]=\"!template && isIEOrEdge\"\n  [style.background-image]=\"!template && isIEOrEdge ? 'url(' + image + ')' : ''\"\n  [ngSwitch]=\"setLinkType()\">\n\n  <!-- slide com link interno -->\n  <a\n    *ngSwitchCase=\"'internalLink'\" \n    class=\"po-slide-item-link\" \n    [routerLink]=\"link\">\n    <ng-container \n      [ngTemplateOutlet]=\"slideItemImage\" \n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image }}\">\n    </ng-container>\n  </a>\n  <!-- slide com link externo -->\n  <a \n    *ngSwitchCase=\"'externalLink'\"\n    class=\"po-slide-item-link\" \n    [href]=\"link\">\n    <ng-container \n      [ngTemplateOutlet]=\"slideItemImage\" \n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image }}\">\n    </ng-container>\n  </a>\n  <!-- slide sem link -->\n  <a \n    *ngSwitchCase=\"'noLink'\"\n    [ngClass]=\"action ? 'po-slide-item-link' : 'po-slide-item-no-link'\"\n    (click)=\"action ? action(data) : undefined\">\n    <ng-container \n      [ngTemplateOutlet]=\"template ? slideItemTemplate : slideItemImage\"\n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image, data: data }}\">\n    </ng-container>\n  </a>\n\n</div>\n\n<ng-template #slideItemTemplate let-item='item'; let-index='index'>\n  <div\n    *ngIf=\"template\"\n    class=\"po-slide-item-content\">\n    <ng-template\n      [ngTemplateOutlet]=\"template.templateRef\"\n      [ngTemplateOutletContext]=\"{ $implicit: item.data, index: index }\">\n    </ng-template>\n  </div>\n</ng-template>\n\n<ng-template #slideItemImage let-item='item'>\n  <img\n    *ngIf=\"item.image && !isIEOrEdge\"\n    class=\"po-slide-image\"\n    [alt]=\"item.alt\"\n    [src]=\"item.image\"\n    [style.height.px]=\"imageHeight\">\n</ng-template>\n"
                }] }
    ];
    PoSlideItemComponent.propDecorators = {
        itemElement: [{ type: ViewChild, args: ['slideItem', { static: true },] }],
        action: [{ type: Input, args: ['p-action',] }],
        alt: [{ type: Input, args: ['p-alt',] }],
        data: [{ type: Input, args: ['p-data',] }],
        image: [{ type: Input, args: ['p-image',] }],
        imageHeight: [{ type: Input, args: ['p-image-height',] }],
        template: [{ type: Input, args: ['p-template',] }],
        link: [{ type: Input, args: ['p-link',] }]
    };
    return PoSlideItemComponent;
}());
export { PoSlideItemComponent };
if (false) {
    /** @type {?} */
    PoSlideItemComponent.prototype.itemElement;
    /** @type {?} */
    PoSlideItemComponent.prototype.isIEOrEdge;
    /**
     * Ação executada ao clicar em uma imagem
     * @type {?}
     */
    PoSlideItemComponent.prototype.action;
    /**
     * Texto alternativo quando a imagem não é encontrada
     * @type {?}
     */
    PoSlideItemComponent.prototype.alt;
    /**
     * Dados para o template customizado
     * @type {?}
     */
    PoSlideItemComponent.prototype.data;
    /**
     * Caminho da imagem
     * @type {?}
     */
    PoSlideItemComponent.prototype.image;
    /**
     * Altura da imagem
     * @type {?}
     */
    PoSlideItemComponent.prototype.imageHeight;
    /**
     * Template customizado
     * @type {?}
     */
    PoSlideItemComponent.prototype.template;
    /**
     * Link executado ao clicar em uma imagem
     * @type {?}
     */
    PoSlideItemComponent.prototype.link;
}
