// @portinari/portinari-ui/esm5/lib/components/po-dropdown/po-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Renderer2, ViewChild } from '@angular/core';
import { isKeyCodeEnter } from './../../utils/util';
import { PoDropdownBaseComponent } from './po-dropdown-base.component';
/**
 * \@docsExtends PoDropdownBaseComponent
 *
 * \@example
 *
 * <example name="po-dropdown-basic" title="Portinari Dropdown Basic" >
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.html"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-dropdown-basic/sample-po-dropdown-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-labs" title="Portinari Dropdown Labs" >
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.html"> </file>
 *  <file name="sample-po-dropdown-labs/sample-po-dropdown-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-dropdown-social-network" title="Portinari Dropdown - Social Network" >
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.html"> </file>
 *  <file name="sample-po-dropdown-social-network/sample-po-dropdown-social-network.component.ts"> </file>
 * </example>
 */
var PoDropdownComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoDropdownComponent, _super);
    function PoDropdownComponent(renderer) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.onScroll = (/**
         * @return {?}
         */
        function () {
            if (_this.open) {
                _this.hideDropdown();
            }
        });
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    PoDropdownComponent.prototype.onKeyDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (isKeyCodeEnter(event)) {
            this.toggleDropdown();
        }
    };
    /**
     * @return {?}
     */
    PoDropdownComponent.prototype.toggleDropdown = /**
     * @return {?}
     */
    function () {
        this.dropdownRef && !this.open && !this.disabled ? this.showDropdown() : this.hideDropdown();
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PoDropdownComponent.prototype.checkClickArea = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.dropdownRef && this.dropdownRef.nativeElement.contains(event.target);
    };
    /**
     * @private
     * @return {?}
     */
    PoDropdownComponent.prototype.hideDropdown = /**
     * @private
     * @return {?}
     */
    function () {
        this.icon = 'po-icon-arrow-down';
        this.removeListeners();
        this.popupRef.close();
        this.open = false;
    };
    /**
     * @private
     * @return {?}
     */
    PoDropdownComponent.prototype.initializeListeners = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.clickoutListener = this.renderer.listen('document', 'click', (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.wasClickedOnDropdown(event);
        }));
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @return {?}
         */
        function () {
            _this.hideDropdown();
        }));
        window.addEventListener('scroll', this.onScroll, true);
    };
    /**
     * @private
     * @return {?}
     */
    PoDropdownComponent.prototype.removeListeners = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.clickoutListener) {
            this.clickoutListener();
        }
        if (this.resizeListener) {
            this.resizeListener();
        }
        window.removeEventListener('scroll', this.onScroll, true);
    };
    /**
     * @private
     * @return {?}
     */
    PoDropdownComponent.prototype.showDropdown = /**
     * @private
     * @return {?}
     */
    function () {
        this.icon = 'po-icon-arrow-up';
        this.initializeListeners();
        this.popupRef.open();
        this.open = true;
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    PoDropdownComponent.prototype.wasClickedOnDropdown = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var clickedOnDropdown = this.checkClickArea(event);
        if (!clickedOnDropdown) {
            this.hideDropdown();
        }
    };
    PoDropdownComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-dropdown',
                    template: "<div #dropdownRef\n  class=\"po-dropdown\"\n  tabindex=\"{{ disabled ? -1 : 0 }}\"\n  (click)=\"toggleDropdown()\"\n  (keydown)=\"onKeyDown($event)\">\n\n  <div class=\"po-dropdown-button\" [ngClass]=\"{'po-dropdown-button-disabled' : disabled}\">\n    {{ label }}\n    <div class=\"po-dropdown-icon po-icon {{icon}}\"></div>\n  </div>\n\n  <po-popup #popupRef\n    p-hide-arrow\n    p-is-corner-align\n    p-position=\"bottom-left\"\n    [p-actions]=\"actions\"\n    [p-custom-positions]=\"['bottom-left', 'top-left']\"\n    [p-target]=\"dropdownRef\">\n  </po-popup>\n\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    PoDropdownComponent.ctorParameters = function () { return [
        { type: Renderer2 }
    ]; };
    PoDropdownComponent.propDecorators = {
        dropdownRef: [{ type: ViewChild, args: ['dropdownRef', { read: ElementRef, static: true },] }],
        popupRef: [{ type: ViewChild, args: ['popupRef', { static: false },] }]
    };
    return PoDropdownComponent;
}(PoDropdownBaseComponent));
export { PoDropdownComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoDropdownComponent.prototype.clickoutListener;
    /**
     * @type {?}
     * @private
     */
    PoDropdownComponent.prototype.resizeListener;
    /** @type {?} */
    PoDropdownComponent.prototype.dropdownRef;
    /** @type {?} */
    PoDropdownComponent.prototype.popupRef;
    /**
     * @type {?}
     * @private
     */
    PoDropdownComponent.prototype.onScroll;
    /**
     * @type {?}
     * @private
     */
    PoDropdownComponent.prototype.renderer;
}
