// @portinari/portinari-ui/esm5/lib/components/po-accordion/po-accordion-item-header/po-accordion-item-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
var PoAccordionItemHeaderComponent = /** @class */ (function () {
    function PoAccordionItemHeaderComponent() {
        this.expanded = false;
        this.toggle = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PoAccordionItemHeaderComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.expanded = !this.expanded;
        this.toggle.emit(this.expanded);
    };
    PoAccordionItemHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-accordion-item-header',
                    template: "<header class=\"po-accordion-item-header\">\n  <button class=\"po-accordion-item-header-button po-clickable\" (click)=\"onClick()\">\n    <div class=\"po-text-ellipsis po-accordion-item-header-title\">{{ label }}</div>\n    <span #icon class=\"po-icon po-accordion-item-header-icon po-icon-arrow-down\">\n    </span>\n  </button>\n</header>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PoAccordionItemHeaderComponent.propDecorators = {
        expanded: [{ type: Input, args: ['p-expanded',] }],
        label: [{ type: Input, args: ['p-label',] }],
        toggle: [{ type: Output, args: ['p-toggle',] }]
    };
    return PoAccordionItemHeaderComponent;
}());
export { PoAccordionItemHeaderComponent };
if (false) {
    /** @type {?} */
    PoAccordionItemHeaderComponent.prototype.expanded;
    /** @type {?} */
    PoAccordionItemHeaderComponent.prototype.label;
    /** @type {?} */
    PoAccordionItemHeaderComponent.prototype.toggle;
}
