// @portinari/portinari-ui/esm5/lib/components/po-menu/po-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ContentChild, ElementRef, Renderer2, ViewContainerRef } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { map } from 'rxjs/operators';
import { getFormattedLink, isMobile, openExternalLink } from '../../utils/util';
import { PoMenuBaseComponent } from './po-menu-base.component';
import { PoMenuHeaderTemplateDirective } from './po-menu-header-template/po-menu-header-template.directive';
import { PoMenuItemsService } from './services/po-menu-items.service';
import { PoMenuService } from './services/po-menu.service';
/** @type {?} */
var poMenuDebounceTime = 400;
/** @type {?} */
var poMenuMinLength = 3;
/** @type {?} */
var poMenuRootLevel = 1;
/**
 * \@docsExtends PoMenuBaseComponent
 *
 * \@description
 *
 * Aparece completo em telas com largura maior que 1200px, caso contrário o menu é escondido e chamado por meio de um botão.
 *
 * O menu também pode ser colapsado. Essa opção é habilitada quando todos os itens de primeiro nível possuírem ícones e textos curtos.
 * Se colapsado, somente os itens de primeiro nível serão exibidos e, caso o item selecionado possua sub-níveis,
 * então o menu alternará novamente para o estado aberto.
 *
 * Existe a possibilidade de customizar a logomarca, que é exibida na parte superior do componente.
 *
 * E para adicionar um conteúdo personalizado entre a logomarca e o campo de filtro,
 * basta adicionar este conteúdo com a diretiva [**p-menu-header-template**](/documentation/po-menu-header-template).
 *
 * Caso utilizar o filtro de menus, é possível realizar buscas em serviço, apenas informando a URL do serviço ou a instância de
 * um serviço customizado implementando a interface `PoMenuFilter`.
 *
 * Para o menu funcionar corretamente é necessário importar o `RouterModule` e `Routes` do módulo principal de
 *  sua aplicação:
 *
 * ````
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     RouterModule,
 *     Routes,
 *     ...
 *     PoModule,
 *     ...
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 *
 * Além disso é necessário criar um módulo configurando as rotas da aplicação.
 *
 * ```
 * import { NgModule } from '\@angular/core';
 *
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * import { HelloWorldComponent } from './hello-world/hello-world.component';
 *
 * const routes: Routes = [
 *   {path: 'hello-world', component: HelloWorldComponent}
 * ];
 *
 * \@NgModule({
 *   imports: [RouterModule.forRoot(routes, {useHash: true})],
 *   exports: [RouterModule]
 * })
 * export class AppRoutingModule {}
 * ```
 *
 * \@example
 *
 * <example name="po-menu-basic" title="Portinari Menu Basic">
 *   <file name="sample-po-menu-basic/sample-po-menu-basic.component.html"> </file>
 *   <file name="sample-po-menu-basic/sample-po-menu-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-menu-labs" title="Portinari Menu Labs">
 *   <file name="sample-po-menu-labs/sample-po-menu-labs.component.html"> </file>
 *   <file name="sample-po-menu-labs/sample-po-menu-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-menu-human-resources" title="Portinari Menu - Human Resources">
 *   <file name="sample-po-menu-human-resources/sample-po-menu-human-resources.component.html"> </file>
 *   <file name="sample-po-menu-human-resources/sample-po-menu-human-resources.component.ts"> </file>
 *   <file name="sample-po-menu-human-resources/sample-po-menu-human-resources.service.ts"> </file>
 * </example>
 */
var PoMenuComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoMenuComponent, _super);
    function PoMenuComponent(changeDetector, viewRef, element, renderer, router, menuItemsService, menuService) {
        var _this = _super.call(this, menuService) || this;
        _this.changeDetector = changeDetector;
        _this.element = element;
        _this.renderer = renderer;
        _this.router = router;
        _this.menuItemsService = menuItemsService;
        _this.filterLoading = false;
        _this.mobileOpened = false;
        _this.noData = false;
        _this.filteringItems = false;
        _this.menuInitialized = false;
        _this.menuPrevious = null;
        _this.parentRef = viewRef['_view']['component'];
        return _this;
    }
    Object.defineProperty(PoMenuComponent.prototype, "isActiveItemMenuSubMenu", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return this.activeMenuItem['level'] > this.groupedMenuItem['level'];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuComponent.prototype, "enableCollapse", {
        get: /**
         * @return {?}
         */
        function () {
            return this.isCollapsed && !this.collapsedMobile;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuComponent.prototype, "enableCollapseButton", {
        get: /**
         * @return {?}
         */
        function () {
            return this.allowCollapseMenu && !this.collapsed && !this.mobileOpened;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuComponent.prototype, "hasFooter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.allowCollapseMenu && !this.mobileOpened && this.enableCollapseButton || (this.collapsed && !this.collapsedMobile);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuComponent.prototype, "isCollapsed", {
        get: /**
         * @return {?}
         */
        function () {
            return this.allowCollapseMenu && this.collapsed;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        if (this.filteringItems && this.filter) {
            return;
        }
        /** @type {?} */
        var menuCurrent = JSON.stringify(this.menus);
        if (this.menuPrevious !== menuCurrent || !this.menuInitialized) {
            this.updateMenu();
            this.validateCollapseClass();
        }
    };
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.itemSubscription.unsubscribe();
        this.routeSubscription.unsubscribe();
        if (this.resizeListener) {
            this.resizeListener();
        }
    };
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.subscribeToMenuItem();
        this.subscribeToRoute();
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var urlRouter = _this.checkingRouterChildrenFragments();
            _this.checkActiveMenuByUrl(urlRouter);
        }));
    };
    /**
     * @param {?} urlPath
     * @param {?} menus
     * @return {?}
     */
    PoMenuComponent.prototype.activateMenuByUrl = /**
     * @param {?} urlPath
     * @param {?} menus
     * @return {?}
     */
    function (urlPath, menus) {
        var _this = this;
        if (menus) {
            return menus.some((/**
             * @param {?} menu
             * @return {?}
             */
            function (menu) {
                /** @type {?} */
                var formattedLink = getFormattedLink(menu.link);
                if (formattedLink === urlPath) {
                    _this.linkActive = formattedLink;
                    _this.activateMenuItem(menu);
                    return true;
                }
                else {
                    return _this.activateMenuByUrl(urlPath, menu.subItems);
                }
            }));
        }
    };
    /**
     * @param {?} urlPath
     * @return {?}
     */
    PoMenuComponent.prototype.checkActiveMenuByUrl = /**
     * @param {?} urlPath
     * @return {?}
     */
    function (urlPath) {
        if (!this.linkActive || this.linkActive !== urlPath) {
            this.activateMenuByUrl(urlPath, this.menus);
        }
    };
    /**
     * <a id="colapseMethod"></a>
     *
     * *Método para colapsar (retrair) o menu.
     */
    /**
     * <a id="colapseMethod"></a>
     *
     * *Método para colapsar (retrair) o menu.
     * @return {?}
     */
    PoMenuComponent.prototype.collapse = /**
     * <a id="colapseMethod"></a>
     *
     * *Método para colapsar (retrair) o menu.
     * @return {?}
     */
    function () {
        this.validateToggleMenu(true);
    };
    /**
     * @param {?} filter
     * @return {?}
     */
    PoMenuComponent.prototype.debounceFilter = /**
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        var _this = this;
        clearTimeout(this.timeoutFilter);
        this.timeoutFilter = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.filterProcess(filter);
        }), poMenuDebounceTime);
    };
    /**
     * <a id="expandMethod"></a>
     *
     * *Método para expandir (aumentar) o menu.
     */
    /**
     * <a id="expandMethod"></a>
     *
     * *Método para expandir (aumentar) o menu.
     * @return {?}
     */
    PoMenuComponent.prototype.expand = /**
     * <a id="expandMethod"></a>
     *
     * *Método para expandir (aumentar) o menu.
     * @return {?}
     */
    function () {
        this.validateToggleMenu(false);
    };
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.subscribeToMenuItem = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.itemSubscription = this.menuItemsService.receiveFromChildMenuClicked().subscribe((/**
         * @param {?} menu
         * @return {?}
         */
        function (menu) {
            _this.clickMenuItem(menu);
        }));
    };
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.subscribeToRoute = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.routeSubscription = this.router.events.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val instanceof NavigationEnd || val instanceof NavigationCancel) {
                /** @type {?} */
                var urlRouter = _this.checkingRouterChildrenFragments();
                _this.checkActiveMenuByUrl(urlRouter);
            }
        }));
    };
    /**
     * <a id="toggleMethod"></a>
     * *Método que colapsa e expande o menu alternadamente.
     *
     * > *Os métodos apenas vão colapsar/expandir o menu se:
     *  - Todos os itens de menu tiverem valor nas propriedades `icon` e `shortLabel`.
     */
    /**
     * <a id="toggleMethod"></a>
     * *Método que colapsa e expande o menu alternadamente.
     *
     * > *Os métodos apenas vão colapsar/expandir o menu se:
     *  - Todos os itens de menu tiverem valor nas propriedades `icon` e `shortLabel`.
     * @return {?}
     */
    PoMenuComponent.prototype.toggle = /**
     * <a id="toggleMethod"></a>
     * *Método que colapsa e expande o menu alternadamente.
     *
     * > *Os métodos apenas vão colapsar/expandir o menu se:
     *  - Todos os itens de menu tiverem valor nas propriedades `icon` e `shortLabel`.
     * @return {?}
     */
    function () {
        this.validateToggleMenu(!this.collapsed);
    };
    /**
     * @return {?}
     */
    PoMenuComponent.prototype.toggleMenuMobile = /**
     * @return {?}
     */
    function () {
        this.mobileOpened = !this.mobileOpened;
        this.collapsedMobile = this.collapsed && this.mobileOpened;
        this.validateCollapseClass(this.collapsedMobile);
        if (isMobile()) {
            return;
        }
        if (this.mobileOpened) {
            this.createResizeListener();
        }
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.activateCollapseSubMenuItem = /**
     * @private
     * @return {?}
     */
    function () {
        this.clearGroupMenuIfFirstLevel(this.activeMenuItem);
        if (!this.collapsed && this.activeMenuItem['level'] > poMenuRootLevel && this.isActiveItemMenuSubMenu) {
            this.openParentMenu(this.activeMenuItem);
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuComponent.prototype.activateMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.activeMenuItem = menu;
        this.linkActive = menu.link;
        if (this.activeMenuItem['level'] > poMenuRootLevel) {
            this.openParentMenu(this.activeMenuItem);
        }
        else {
            this.groupedMenuItem = null;
        }
        this.menuItemsService.sendToChildMenuClicked({ active: this.activeMenuItem, grouped: this.groupedMenuItem, activatedByRoute: true });
    };
    /**
     * @private
     * @param {?} menus
     * @return {?}
     */
    PoMenuComponent.prototype.areSubMenus = /**
     * @private
     * @param {?} menus
     * @return {?}
     */
    function (menus) {
        return menus.every((/**
         * @param {?} menu
         * @return {?}
         */
        function (menu) { return menu['level'] > poMenuRootLevel; }));
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.checkingRouterChildrenFragments = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var childrenPrimary = this.router.parseUrl(this.router.url).root.children['primary'];
        return childrenPrimary ? "/" + childrenPrimary.segments.map((/**
         * @param {?} it
         * @return {?}
         */
        function (it) { return it.path; })).join('/') : '';
    };
    /**
     * @private
     * @param {?} activeMenuItem
     * @return {?}
     */
    PoMenuComponent.prototype.clearGroupMenuIfFirstLevel = /**
     * @private
     * @param {?} activeMenuItem
     * @return {?}
     */
    function (activeMenuItem) {
        if (activeMenuItem['level'] === poMenuRootLevel) {
            this.groupedMenuItem = undefined;
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuComponent.prototype.clickMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        if (menu.action) {
            this.executeMenuAction(menu);
        }
        if (menu['type'] === 'externalLink') {
            openExternalLink(menu.link);
        }
        else if (menu['type'] === 'internalLink') {
            this.activateMenuItem(menu);
        }
        else if (menu['type'] === 'subItems') {
            if (this.filteringItems) {
                this.filteringItems = false;
            }
            this.groupMenuItem(menu);
        }
        if (menu['type'] !== 'subItems') {
            this.mobileOpened = false;
        }
    };
    /**
     * @private
     * @param {?=} menuItem
     * @return {?}
     */
    PoMenuComponent.prototype.convertToMenuItemFiltered = /**
     * @private
     * @param {?=} menuItem
     * @return {?}
     */
    function (menuItem) {
        if (menuItem === void 0) { menuItem = { label: '', link: '' }; }
        var label = menuItem.label, link = menuItem.link;
        /** @type {?} */
        var menuItemFiltered = { label: label, link: link };
        this.setMenuItemProperties(menuItemFiltered);
        return menuItemFiltered;
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.createResizeListener = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @return {?}
         */
        function () {
            _this.toggleResize();
            _this.resizeListener();
        }));
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuComponent.prototype.executeMenuAction = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        if (typeof menu.action === 'string') {
            this.parentRef[menu.action](menu);
        }
        else {
            menu.action.call(this.parentRef, menu);
        }
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    PoMenuComponent.prototype.filterItems = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var trimFilter, _a, _b;
            return tslib_1.__generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        trimFilter = filter && filter.trim();
                        if (!trimFilter) return [3 /*break*/, 4];
                        this.filteredItems = [];
                        _a = this;
                        if (!this.filterService) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.filterOnService(trimFilter)];
                    case 1:
                        _b = _c.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        _b = this.filterLocalItems(trimFilter);
                        _c.label = 3;
                    case 3:
                        _a.filteredItems = _b;
                        this.filteringItems = true;
                        return [3 /*break*/, 5];
                    case 4:
                        this.filteredItems = tslib_1.__spread(this.menus);
                        this.filteringItems = false;
                        _c.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    PoMenuComponent.prototype.filterLocalItems = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        /** @type {?} */
        var filteredItems = [];
        this.findItems(this.menus, filter.toLowerCase(), filteredItems);
        return filteredItems;
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    PoMenuComponent.prototype.filterProcess = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        var _this = this;
        this.filterLoading = true;
        this.filterItems(filter).then((/**
         * @return {?}
         */
        function () {
            _this.filterLoading = false;
            _this.showNoData();
            _this.changeDetector.detectChanges();
            _this.menuItemsService.sendToChildMenuClicked({ active: _this.activeMenuItem, grouped: _this.groupedMenuItem });
        })).catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this.filterLoading = false;
            Promise.reject(error);
        }));
    };
    /**
     * @private
     * @param {?=} search
     * @return {?}
     */
    PoMenuComponent.prototype.filterOnService = /**
     * @private
     * @param {?=} search
     * @return {?}
     */
    function (search) {
        if (search === void 0) { search = ''; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(search.length >= poMenuMinLength)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.filterService.getFilteredData(search, this.params)
                                .pipe(map((/**
                             * @param {?} menuItemsFiltered
                             * @return {?}
                             */
                            function (menuItemsFiltered) { return menuItemsFiltered.map((/**
                             * @param {?} menuItem
                             * @return {?}
                             */
                            function (menuItem) { return _this.convertToMenuItemFiltered(menuItem); })); })))
                                .toPromise()];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2: return [2 /*return*/, this.filteredItems];
                }
            });
        });
    };
    /**
     * @private
     * @param {?} menus
     * @param {?} filter
     * @param {?} filteredItems
     * @return {?}
     */
    PoMenuComponent.prototype.findItems = /**
     * @private
     * @param {?} menus
     * @param {?} filter
     * @param {?} filteredItems
     * @return {?}
     */
    function (menus, filter, filteredItems) {
        var _this = this;
        menus.forEach((/**
         * @param {?} menu
         * @return {?}
         */
        function (menu) {
            if ((menu.label.toLowerCase().includes(filter) && !menu.subItems) ||
                (menu.subItems && _this.findItems(menu.subItems, filter, filteredItems))) {
                filteredItems.push(menu);
            }
        }));
    };
    /**
     * @private
     * @param {?} menus
     * @param {?} menuItem
     * @return {?}
     */
    PoMenuComponent.prototype.findParent = /**
     * @private
     * @param {?} menus
     * @param {?} menuItem
     * @return {?}
     */
    function (menus, menuItem) {
        /** @type {?} */
        var getParent = (/**
         * @param {?} menuItems
         * @param {?} id
         * @return {?}
         */
        function (menuItems, id) {
            if (menuItems) {
                for (var index = 0; index < menuItems.length; index++) {
                    /** @type {?} */
                    var menu = menuItems[index];
                    if (menu.subItems && menu.subItems.find((/**
                     * @param {?} subItem
                     * @return {?}
                     */
                    function (subItem) { return subItem['id'] === id; }))) {
                        return menu;
                    }
                    /** @type {?} */
                    var found = getParent(menu.subItems, id);
                    if (found) {
                        return found;
                    }
                }
            }
        });
        return getParent(menus, menuItem['id']);
    };
    /**
     * @private
     * @param {?} menus
     * @param {?} menu
     * @return {?}
     */
    PoMenuComponent.prototype.findRootParent = /**
     * @private
     * @param {?} menus
     * @param {?} menu
     * @return {?}
     */
    function (menus, menu) {
        /** @type {?} */
        var findParent = this.findParent;
        /** @type {?} */
        var getRootParent = (/**
         * @param {?} menuItems
         * @param {?} menuItem
         * @return {?}
         */
        function (menuItems, menuItem) {
            /** @type {?} */
            var parent = findParent(menuItems, menuItem);
            if (parent['level'] !== poMenuRootLevel) {
                parent = getRootParent(menuItems, parent);
            }
            return parent;
        });
        return getRootParent(menus, menu);
    };
    /**
     * @private
     * @param {?} menus
     * @param {?} activeMenuItem
     * @param {?} groupedMenuItem
     * @return {?}
     */
    PoMenuComponent.prototype.getActiveMenuParent = /**
     * @private
     * @param {?} menus
     * @param {?} activeMenuItem
     * @param {?} groupedMenuItem
     * @return {?}
     */
    function (menus, activeMenuItem, groupedMenuItem) {
        if (this.areSubMenus([groupedMenuItem, activeMenuItem])) {
            return this.findRootParent(menus, activeMenuItem);
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuComponent.prototype.groupMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        if (this.collapsed) {
            this.toggleMenuCollapse();
        }
        menu['isOpened'] = !menu['isOpened'];
        this.groupedMenuItem = menu;
        if (this.activeMenuItem && menu['isOpened']
            && this.isActiveItemMenuSubMenu
            && this.isRootMenuEqualGroupedMenu(this.menus, this.activeMenuItem, menu)) {
            this.activateMenuItem(this.activeMenuItem);
        }
        this.menuItemsService.sendToChildMenuClicked({ active: this.activeMenuItem, grouped: this.groupedMenuItem });
    };
    /**
     * @private
     * @param {?} menus
     * @param {?} activeMenuItem
     * @param {?} groupedMenuItem
     * @return {?}
     */
    PoMenuComponent.prototype.isRootMenuEqualGroupedMenu = /**
     * @private
     * @param {?} menus
     * @param {?} activeMenuItem
     * @param {?} groupedMenuItem
     * @return {?}
     */
    function (menus, activeMenuItem, groupedMenuItem) {
        /** @type {?} */
        var activeMenuRootParent = this.findRootParent(menus, activeMenuItem);
        return activeMenuRootParent['id'] === groupedMenuItem['id'];
    };
    /**
     * @private
     * @param {?} childMenu
     * @return {?}
     */
    PoMenuComponent.prototype.openParentMenu = /**
     * @private
     * @param {?} childMenu
     * @return {?}
     */
    function (childMenu) {
        /** @type {?} */
        var parent = this.findParent(this.menus, childMenu);
        parent['isOpened'] = true;
        this.groupedMenuItem = parent;
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.showNoData = /**
     * @private
     * @return {?}
     */
    function () {
        this.noData = this.filteredItems.length === 0;
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.toggleGroupedMenuItem = /**
     * @private
     * @return {?}
     */
    function () {
        this.groupedMenuItem['isOpened'] = !this.collapsed && this.allowCollapseMenu;
    };
    /**
     * @private
     * @param {?=} collapsed
     * @return {?}
     */
    PoMenuComponent.prototype.toggleMenuCollapse = /**
     * @private
     * @param {?=} collapsed
     * @return {?}
     */
    function (collapsed) {
        if (collapsed === void 0) { collapsed = false; }
        this.collapsed = collapsed;
        if (this.groupedMenuItem && this.activeMenuItem) {
            this.groupedMenuItem = this.getActiveMenuParent(this.menus, this.activeMenuItem, this.groupedMenuItem) || this.groupedMenuItem;
            this.toggleGroupedMenuItem();
        }
        if (this.activeMenuItem) {
            this.activateCollapseSubMenuItem();
            this.menuItemsService.sendToChildMenuClicked({ active: this.activeMenuItem, grouped: this.groupedMenuItem, activatedByRoute: true });
        }
        this.updateMenu();
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.toggleResize = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.mobileOpened) {
            this.mobileOpened = false;
            this.collapsedMobile = false;
            this.validateCollapseClass(this.collapsedMobile);
        }
    };
    /**
     * @private
     * @param {?} collapsed
     * @return {?}
     */
    PoMenuComponent.prototype.validateToggleMenu = /**
     * @private
     * @param {?} collapsed
     * @return {?}
     */
    function (collapsed) {
        if (!this.allowCollapseMenu) {
            return;
        }
        this.toggleMenuCollapse(collapsed);
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuComponent.prototype.updateMenu = /**
     * @private
     * @return {?}
     */
    function () {
        this.menuInitialized = true;
        this.setMenuExtraProperties();
        this.filteredItems = tslib_1.__spread(this.menus);
        this.menuPrevious = JSON.stringify(this.menus);
        this.validateMenus(this.menus);
    };
    /**
     * @protected
     * @param {?=} collapsedMobile
     * @return {?}
     */
    PoMenuComponent.prototype.validateCollapseClass = /**
     * @protected
     * @param {?=} collapsedMobile
     * @return {?}
     */
    function (collapsedMobile) {
        /** @type {?} */
        var wrapper = this.element.nativeElement.parentNode;
        this.renderer[this.isCollapsed && !collapsedMobile ? 'addClass' : 'removeClass'](wrapper, 'po-collapsed-menu');
    };
    PoMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-menu',
                    template: "<div *ngIf=\"mobileOpened\" class=\"po-menu-overlay\" (click)=\"toggleMenuMobile()\"></div>\n<div class=\"po-menu-mobile po-clickable\" (click)=\"toggleMenuMobile()\">\n  <span class=\"po-icon po-icon-menu\"></span>\n</div>\n\n<div class=\"po-menu\" [ngClass]=\"{ 'po-menu-animation' : mobileOpened }\">\n\n  <div class=\"po-menu-header\">\n    <div *ngIf=\"logo || (shortLogo && enableCollapse)\" class=\"po-menu-header-container-logo\">\n      <a href=\"./\">\n        <img\n          [ngClass]=\"enableCollapse ? 'po-menu-short-logo' : 'po-menu-logo'\"\n          [src]=\"enableCollapse ? shortLogo || logo : logo\">\n      </a>\n    </div>\n\n    <div *ngIf=\"!enableCollapse && menuHeaderTemplate\" class=\"po-menu-header-template\">\n      <ng-container *ngTemplateOutlet=\"menuHeaderTemplate.templateRef\"></ng-container>\n    </div>\n\n    <po-menu-filter *ngIf=\"filter && !enableCollapse\"\n      [p-loading]=\"filterLoading\"\n      (p-filter)=\"debounceFilter($event)\">\n    </po-menu-filter>\n  </div>\n\n  <nav class=\"po-menu-body\">\n\n    <!-- Inner e outer para esconder scroll -->\n    <div class=\"po-menu-outer\">\n      <div class=\"po-menu-inner\">\n\n        <div *ngIf=\"noData\" class=\"po-menu-item-wrapper\">\n          <div class=\"po-menu-item-first\">\n            <div class=\"po-menu-icon-container po-menu-item-no-data\">\n              <span class=\"po-icon po-icon-info po-menu-icon-item po-lg-2\"></span>\n              <div class=\"po-lg-10 po-menu-icon-label\"> {{literals.itemNotFound}} </div>\n            </div>\n          </div>\n        </div>\n\n        <div *ngFor=\"let menu of filteredItems, let menuIndex = index;\" class=\"po-menu-item-wrapper\">\n          <po-menu-item\n            [class.po-menu-item-first]=\"menuIndex === 0\"\n            [p-action]=\"menu.action\"\n            [p-badge-alert]=\"menu.badgeAlert\"\n            [p-badge-color]=\"menu.badge ? menu.badge.color : undefined\"\n            [p-badge-value]=\"menu.badge ? menu.badge.value : undefined\"\n            [p-collapsed-menu]=\"enableCollapse\"\n            [p-icon]=\"allowIcons ? menu.icon : null\"\n            [p-id]=\"menu.id\"\n            [p-label]=\"menu.label\"\n            [p-level]=\"menu.level\"\n            [p-link]=\"menu.link\"\n            [p-short-label]=\"menu.shortLabel\"\n            [p-sub-items]=\"menu.subItems\"\n            [p-type]=\"menu.type\">\n          </po-menu-item>\n        </div>\n      </div>\n    </div>\n  </nav>\n\n  <div *ngIf=\"hasFooter\" class=\"po-menu-footer\">\n    <a\n      class=\"po-menu-collapse-button-icon po-clickable\"\n      (click)=\"toggle()\">\n\n      <span class=\"po-icon\"\n        [class.po-icon-menu-close]=\"enableCollapseButton\"\n        [class.po-icon-menu-open]=\"enableCollapse\">\n      </span>\n    </a>\n  </div>\n</div>\n",
                    providers: [PoMenuItemsService, PoMenuService]
                }] }
    ];
    /** @nocollapse */
    PoMenuComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: Renderer2 },
        { type: Router },
        { type: PoMenuItemsService },
        { type: PoMenuService }
    ]; };
    PoMenuComponent.propDecorators = {
        menuHeaderTemplate: [{ type: ContentChild, args: [PoMenuHeaderTemplateDirective, { static: true },] }]
    };
    return PoMenuComponent;
}(PoMenuBaseComponent));
export { PoMenuComponent };
if (false) {
    /** @type {?} */
    PoMenuComponent.prototype.menuHeaderTemplate;
    /** @type {?} */
    PoMenuComponent.prototype.activeMenuItem;
    /** @type {?} */
    PoMenuComponent.prototype.collapsedMobile;
    /** @type {?} */
    PoMenuComponent.prototype.filterLoading;
    /** @type {?} */
    PoMenuComponent.prototype.groupedMenuItem;
    /** @type {?} */
    PoMenuComponent.prototype.linkActive;
    /** @type {?} */
    PoMenuComponent.prototype.mobileOpened;
    /** @type {?} */
    PoMenuComponent.prototype.noData;
    /** @type {?} */
    PoMenuComponent.prototype.parentRef;
    /** @type {?} */
    PoMenuComponent.prototype.timeoutFilter;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.filteringItems;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.menuInitialized;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.menuPrevious;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.resizeListener;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.itemSubscription;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.routeSubscription;
    /** @type {?} */
    PoMenuComponent.prototype.changeDetector;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.element;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PoMenuComponent.prototype.menuItemsService;
}
