// @portinari/portinari-ui/esm5/lib/components/po-grid/po-grid-cell-action/po-grid-cell-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var PoGridCellActionComponent = /** @class */ (function () {
    function PoGridCellActionComponent() {
        this.position = '';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    PoGridCellActionComponent.prototype.onKeyDownContent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // console.log('onKeyDownContent: ', event);
        // ENTER
        if (event.keyCode === 13) {
            event.preventDefault();
            // this.openActions(this.value);
            return;
        }
    };
    PoGridCellActionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-grid-cell-action',
                    template: "<div>\n  <div #contentElement [id]=\"position\"\n    class=\"po-grid-cell-action-content\"\n    tabindex=\"0\"\n    (keydown)=\"onKeyDownContent($event)\">\n    <span>{{ value || '&nbsp;' }}</span>\n  </div>\n</div>\n\n",
                    styles: ["\n  .po-grid-cell-action-content {\n    border-bottom: 1px solid #eceeee;\n    border-left: none;\n    border-right: 1px solid #eceeee;\n    border-top: none;\n    font-family: NunitoSans;\n    font-size: 14px;\n    height: 44px;\n    line-height: 1.71;\n    margin: 0;\n    max-height: 44px;\n    max-width: 56px;\n    min-height: 44px;\n    min-width: 56px;\n    padding: 10px 16px;\n    text-align: center;\n    width: 56px;\n  }\n\n  .po-grid-cell-action-content:focus {\n    border: none;\n    outline: 1px solid #0c9abe;\n  }\n  "]
                }] }
    ];
    /** @nocollapse */
    PoGridCellActionComponent.ctorParameters = function () { return []; };
    PoGridCellActionComponent.propDecorators = {
        position: [{ type: Input, args: ['p-position',] }],
        value: [{ type: Input, args: ['p-value',] }]
    };
    return PoGridCellActionComponent;
}());
export { PoGridCellActionComponent };
if (false) {
    /** @type {?} */
    PoGridCellActionComponent.prototype.position;
    /** @type {?} */
    PoGridCellActionComponent.prototype.value;
}
