// @portinari/portinari-ui/esm5/lib/components/po-progress/po-progress.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { PoProgressBaseComponent } from './po-progress-base.component';
import { PoProgressStatus } from './enums/po-progress-status.enum';
/**
 * \@docsExtends PoProgressBaseComponent
 *
 * \@example
 * <example name="po-progress-basic" title="Portinari Progress Basic">
 *   <file name="sample-po-progress-basic/sample-po-progress-basic.component.html"> </file>
 *   <file name="sample-po-progress-basic/sample-po-progress-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-progress-labs" title="Portinari Progress Labs">
 *   <file name="sample-po-progress-labs/sample-po-progress-labs.component.html"> </file>
 *   <file name="sample-po-progress-labs/sample-po-progress-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-progress-publication" title="Portinari Progress - Publication">
 *   <file name="sample-po-progress-publication/sample-po-progress-publication.component.html"> </file>
 *   <file name="sample-po-progress-publication/sample-po-progress-publication.component.ts"> </file>
 * </example>
 */
var PoProgressComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoProgressComponent, _super);
    function PoProgressComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PoProgressComponent.prototype, "isAllowCancel", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.cancel.observers.length && this.status !== PoProgressStatus.Success;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoProgressComponent.prototype, "isAllowProgressInfo", {
        get: /**
         * @return {?}
         */
        function () {
            return !!(this.info || this.infoIcon || this.isAllowCancel || this.isAllowRetry);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoProgressComponent.prototype, "isAllowRetry", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.retry.observers.length && this.status === PoProgressStatus.Error;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoProgressComponent.prototype, "statusClass", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.status === PoProgressStatus.Success) {
                return 'po-progress-success';
            }
            if (this.status === PoProgressStatus.Error) {
                return 'po-progress-error';
            }
            return 'po-progress-default';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoProgressComponent.prototype.emitCancellation = /**
     * @return {?}
     */
    function () {
        this.cancel.emit(this.status);
    };
    /**
     * @return {?}
     */
    PoProgressComponent.prototype.emitRetry = /**
     * @return {?}
     */
    function () {
        this.retry.emit();
    };
    PoProgressComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-progress',
                    template: "<div class=\"po-progress\" [ngClass]=\"statusClass\">\n\n  <label *ngIf=\"text\" class=\"po-progress-description-mobile po-progress-description-text\">\n    {{ text }}\n  </label>\n\n  <po-progress-bar\n    class=\"po-progress-bar\"\n    [p-indeterminate]=\"indeterminate\"\n    [p-value]=\"value\">\n  </po-progress-bar>\n\n  <div *ngIf=\"text\" class=\"po-progress-description\">\n    <label class=\"po-progress-description-text\">\n      {{ text }}\n    </label>\n  </div>\n\n  <div *ngIf=\"isAllowProgressInfo\" class=\"po-progress-info\">\n    <span *ngIf=\"infoIcon\" class=\"po-progress-info-icon po-icon {{ infoIcon }}\"></span>\n    <span *ngIf=\"info\" class=\"po-progress-info-text\">{{ info }}</span>\n\n    <button *ngIf=\"isAllowRetry\"\n      class=\"po-progress-info-icon-action po-icon po-icon-refresh po-clickable\"\n      (click)=\"emitRetry()\">\n    </button>\n\n    <button *ngIf=\"isAllowCancel\"\n      class=\"po-progress-info-icon-action po-icon po-icon-close po-clickable\"\n      (click)=\"emitCancellation()\">\n    </button>\n\n  </div>\n\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    return PoProgressComponent;
}(PoProgressBaseComponent));
export { PoProgressComponent };
