// @portinari/portinari-ui/esm2015/lib/components/po-toolbar/po-toolbar-profile/po-toolbar-profile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoControlPositionService } from '../../../services/po-control-position/po-control-position.service';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoToolbarProfileComponent
 *
 * \@description
 *
 * Componente `po-toolbar-profile`.
 */
export class PoToolbarProfileComponent {
    /**
     * @return {?}
     */
    get profileAvatar() {
        return this.profile ? this.profile.avatar : undefined;
    }
}
PoToolbarProfileComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-toolbar-profile',
                template: "<div #profileElement class=\"po-toolbar-profile po-clickable\" (click)=\"popup.toggle()\">\n  <po-avatar p-size=\"xs\" [p-src]=\"profileAvatar\"></po-avatar>\n</div>\n\n<po-popup #popup\n  [p-actions]=\"profileActions\"\n  [p-target]=\"profileElement\">\n\n  <div p-popup-header-template\n    *ngIf=\"profile?.title\"\n    class=\"po-toolbar-profile-item-header\">\n\n    <po-avatar\n      class=\"po-toolbar-profile-item-avatar\"\n      p-size=\"sm\"\n      [p-src]=\"profileAvatar\">\n    </po-avatar>\n\n    <div>\n      <div class=\"po-toolbar-profile-item-header-title\">\n        {{ profile.title }}\n      </div>\n\n      <div *ngIf=\"profile.subtitle\" class=\"po-toolbar-profile-item-header-subtitle\">\n        {{ profile.subtitle }}\n      </div>\n    </div>\n  </div>\n\n</po-popup>\n",
                providers: [PoControlPositionService]
            }] }
];
PoToolbarProfileComponent.propDecorators = {
    profile: [{ type: Input, args: ['p-profile',] }],
    profileActions: [{ type: Input, args: ['p-profile-actions',] }]
};
if (false) {
    /**
     * Objeto que implementa a interface `PoToolbarProfile`.
     * @type {?}
     */
    PoToolbarProfileComponent.prototype.profile;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    PoToolbarProfileComponent.prototype.profileActions;
}
