// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-content/po-page-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2 } from '@angular/core';
import { PoPageContentBaseComponent } from './po-page-content-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends PoPageContentBaseComponent
 */
export class PoPageContentComponent extends PoPageContentBaseComponent {
    /**
     * @param {?} renderer
     */
    constructor(renderer) {
        super();
        this.renderer = renderer;
        this.contentOpacity = 0;
        this.height = '90%';
        this.overflowY = 'none';
        this.initializeListeners();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.recalculateHeaderSize();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeListeners();
    }
    /**
     * @return {?}
     */
    recalculateHeaderSize() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const pageHeaderElement = document.querySelector('div.po-page-header');
            this.setHeightContent(pageHeaderElement);
            this.contentOpacity = 1;
        }));
    }
    /**
     * @param {?} poPageHeader
     * @return {?}
     */
    setHeightContent(poPageHeader) {
        /** @type {?} */
        const bodyHeight = document.body.clientHeight;
        /** @type {?} */
        const pageHeaderHeight = poPageHeader ? poPageHeader.offsetTop + poPageHeader.offsetHeight : 0;
        /** @type {?} */
        const newHeight = bodyHeight - pageHeaderHeight;
        this.height = `${newHeight}px`;
    }
    /**
     * @private
     * @return {?}
     */
    initializeListeners() {
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @return {?}
         */
        () => {
            this.recalculateHeaderSize();
        }));
    }
    /**
     * @private
     * @return {?}
     */
    removeListeners() {
        this.resizeListener();
    }
}
PoPageContentComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-content',
                template: "<div class=\"po-page-content\"\n  [style.height]=\"height\"\n  [style.opacity]=\"contentOpacity\"\n  [style.overflow-y]=\"overflowY\">\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
/** @nocollapse */
PoPageContentComponent.ctorParameters = () => [
    { type: Renderer2 }
];
if (false) {
    /** @type {?} */
    PoPageContentComponent.prototype.contentOpacity;
    /** @type {?} */
    PoPageContentComponent.prototype.height;
    /** @type {?} */
    PoPageContentComponent.prototype.overflowY;
    /** @type {?} */
    PoPageContentComponent.prototype.renderer;
}
