// @portinari/portinari-ui/esm2015/lib/components/po-tabs/po-tab-button/po-tab-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean } from '../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por manipular os botões de aba.
 */
export class PoTabButtonComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        // Função sera emitida quando a tab ficar ativada
        this.activated = new EventEmitter();
        // Função sera emitida quando a tab ficar desabilitada ou escondida
        this.changeState = new EventEmitter();
        // Método recebido do usuário para ser disparado quando clicar na aba
        this.click = new EventEmitter();
    }
    // Ativa o botão
    /**
     * @param {?} value
     * @return {?}
     */
    set active(value) {
        this._active = value;
        this.emitActivated();
    }
    /**
     * @return {?}
     */
    get active() {
        return this._active;
    }
    // Oculta o botão
    /**
     * @param {?} value
     * @return {?}
     */
    set hide(value) {
        this._hide = convertToBoolean(value);
        this.setDisplayOnHide();
    }
    /**
     * @return {?}
     */
    get hide() {
        return this._hide;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if ((changes.hide && changes.hide.currentValue) || (changes.disabled && changes.disabled.currentValue)) {
            this.changeState.emit(this);
        }
    }
    /**
     * @return {?}
     */
    onClick() {
        if (!this.disabled) {
            this.click.emit(this.id);
        }
    }
    /**
     * @private
     * @return {?}
     */
    emitActivated() {
        if (this.active) {
            this.activated.emit(this);
        }
    }
    /**
     * @private
     * @return {?}
     */
    setDisplayOnHide() {
        this.elementRef.nativeElement.style.display = this.hide ? 'none' : '';
    }
}
PoTabButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-tab-button',
                template: "<div\n  class=\"po-tab-button-md\"\n  [attr.tabindex]=\"disabled ? null : 0\"\n  [class.po-tab-button-active]=\"active\"\n  [class.po-tab-button-disabled]=\"disabled\"\n  [class.po-tab-button-sm]=\"small\"\n  (click)=\"onClick()\"\n  (keyup.enter)=\"onClick()\">\n\n  <span class=\"po-tab-button-label po-text-ellipsis\">\n    {{label}}\n  </span>\n</div>\n"
            }] }
];
/** @nocollapse */
PoTabButtonComponent.ctorParameters = () => [
    { type: ElementRef }
];
PoTabButtonComponent.propDecorators = {
    active: [{ type: Input, args: ['p-active',] }],
    disabled: [{ type: Input, args: ['p-disabled',] }],
    hide: [{ type: Input, args: ['p-hide',] }],
    id: [{ type: Input, args: ['p-id',] }],
    label: [{ type: Input, args: ['p-label',] }],
    small: [{ type: Input, args: ['p-small',] }],
    activated: [{ type: Output, args: ['p-activated',] }],
    changeState: [{ type: Output, args: ['p-change-state',] }],
    click: [{ type: Output, args: ['p-click',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTabButtonComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    PoTabButtonComponent.prototype._hide;
    /** @type {?} */
    PoTabButtonComponent.prototype.disabled;
    /** @type {?} */
    PoTabButtonComponent.prototype.id;
    /** @type {?} */
    PoTabButtonComponent.prototype.label;
    /** @type {?} */
    PoTabButtonComponent.prototype.small;
    /** @type {?} */
    PoTabButtonComponent.prototype.activated;
    /** @type {?} */
    PoTabButtonComponent.prototype.changeState;
    /** @type {?} */
    PoTabButtonComponent.prototype.click;
    /**
     * @type {?}
     * @private
     */
    PoTabButtonComponent.prototype.elementRef;
}
