// @portinari/portinari-ui/esm2015/lib/components/po-field/po-rich-text/po-rich-text-body/po-rich-text-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { isIE } from './../../../../utils/util';
import { PoKeyCodeEnum } from './../../../../enums/po-key-code.enum';
/** @type {?} */
const poRichTextBodyCommands = [
    'bold', 'italic', 'underline', 'justifyleft', 'justifycenter', 'justifyright', 'justifyfull', 'insertUnorderedList', 'Createlink'
];
export class PoRichTextBodyComponent {
    constructor() {
        this.change = new EventEmitter();
        this.commands = new EventEmitter();
        this.shortcutCommand = new EventEmitter();
        this.value = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.bodyElement.nativeElement.designMode = 'on';
        // timeout necessário para setar o valor vindo do writeValue do componente principal.
        setTimeout((/**
         * @return {?}
         */
        () => this.updateValueWithModelValue()));
    }
    /**
     * @param {?} command
     * @return {?}
     */
    executeCommand(command) {
        this.bodyElement.nativeElement.focus();
        if (typeof (command) === 'object') {
            if (command.command === 'InsertHTML') {
                const { command: linkCommand, value: { urlLink }, value: { urlLinkText } } = command;
                this.handleCommandLink(linkCommand, urlLink, urlLinkText);
            }
            else {
                document.execCommand(command.command, false, command.value);
            }
        }
        else {
            document.execCommand(command, false, null);
        }
        this.updateModel();
        this.value.emit(this.modelValue);
    }
    /**
     * @return {?}
     */
    onBlur() {
        if (this.modelValue !== this.valueBeforeChange) {
            clearTimeout(this.timeoutChange);
            this.timeoutChange = setTimeout((/**
             * @return {?}
             */
            () => {
                this.change.emit(this.modelValue);
            }), 200);
        }
    }
    /**
     * @return {?}
     */
    focus() {
        this.bodyElement.nativeElement.focus();
    }
    /**
     * @return {?}
     */
    onClick() {
        this.emitSelectionCommands();
    }
    /**
     * @return {?}
     */
    onFocus() {
        this.valueBeforeChange = this.modelValue;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeyDown(event) {
        /** @type {?} */
        const keyL = event.keyCode === PoKeyCodeEnum.keyL;
        if (keyL && event.ctrlKey || keyL && event.metaKey) {
            event.preventDefault();
            this.shortcutCommand.emit();
        }
    }
    /**
     * @return {?}
     */
    onKeyUp() {
        // Tratamento necessário para eliminar a tag <br> criada no firefox quando o body for limpo.
        /** @type {?} */
        const bodyElement = this.bodyElement.nativeElement;
        if (!bodyElement.innerText.trim() && bodyElement.childNodes.length === 1 && bodyElement.querySelector('br')) {
            bodyElement.querySelector('br').remove();
        }
        this.updateModel();
        this.emitSelectionCommands();
    }
    /**
     * @return {?}
     */
    update() {
        setTimeout((/**
         * @return {?}
         */
        () => this.updateModel()));
        setTimeout((/**
         * @return {?}
         */
        () => this.onKeyUp()));
    }
    /**
     * @private
     * @return {?}
     */
    cursorPositionedInALink() {
        /** @type {?} */
        const link = document.getSelection();
        return link.focusNode.parentElement.tagName === 'A';
    }
    /**
     * @private
     * @return {?}
     */
    emitSelectionCommands() {
        /** @type {?} */
        const commands = poRichTextBodyCommands.filter((/**
         * @param {?} command
         * @return {?}
         */
        command => document.queryCommandState(command)));
        /** @type {?} */
        const rgbColor = document.queryCommandValue('ForeColor');
        /** @type {?} */
        const hexColor = this.rgbToHex(rgbColor);
        if (this.cursorPositionedInALink()) {
            commands.push('Createlink');
        }
        this.commands.emit({ commands, hexColor });
    }
    /**
     * @private
     * @param {?} linkCommand
     * @param {?} urlLink
     * @param {?} urlLinkText
     * @return {?}
     */
    handleCommandLink(linkCommand, urlLink, urlLinkText) {
        if (isIE()) {
            this.insertHtmlLinkElement(urlLink, urlLinkText);
        }
        else {
            // necessário '&nbsp;' no fim pois o Firefox mantém o cursor dentro da tag;
            /** @type {?} */
            const linkValue = `<a class="po-rich-text-link" href="${urlLink}" target="_blank">${urlLinkText || urlLink}</a>`;
            document.execCommand(linkCommand, false, linkValue);
        }
    }
    // tratamento específico para IE pois não suporta o comando 'insertHTML'.
    /**
     * @private
     * @param {?} urlLink
     * @param {?} urlLinkText
     * @return {?}
     */
    insertHtmlLinkElement(urlLink, urlLinkText) {
        /** @type {?} */
        const selection = document.getSelection();
        /** @type {?} */
        const selectionRange = selection.getRangeAt(0);
        /** @type {?} */
        const elementLink = document.createElement('a');
        /** @type {?} */
        const elementlinkText = document.createTextNode(urlLinkText);
        elementLink.appendChild(elementlinkText);
        elementLink.href = urlLink;
        elementLink.setAttribute('target', '_blank');
        elementLink.classList.add('po-rich-text-link');
        selectionRange.deleteContents();
        selectionRange.insertNode(elementLink);
    }
    /**
     * @private
     * @param {?} rgb
     * @return {?}
     */
    rgbToHex(rgb) {
        // Tratamento necessário para converter o código rgb para hexadecimal.
        /** @type {?} */
        const sep = rgb.indexOf(',') > -1 ? ',' : ' ';
        rgb = rgb.substr(4).split(')')[0].split(sep);
        /** @type {?} */
        let r = (+rgb[0]).toString(16);
        /** @type {?} */
        let g = (+rgb[1]).toString(16);
        /** @type {?} */
        let b = (+rgb[2]).toString(16);
        if (r.length === 1) {
            r = '0' + r;
        }
        if (g.length === 1) {
            g = '0' + g;
        }
        if (b.length === 1) {
            b = '0' + b;
        }
        return '#' + r + g + b;
    }
    /**
     * @private
     * @return {?}
     */
    updateModel() {
        this.modelValue = this.bodyElement.nativeElement.innerHTML;
        this.value.emit(this.modelValue);
    }
    /**
     * @private
     * @return {?}
     */
    updateValueWithModelValue() {
        if (this.modelValue) {
            this.bodyElement.nativeElement.insertAdjacentHTML('afterbegin', this.modelValue);
        }
    }
}
PoRichTextBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-rich-text-body',
                template: "<div #bodyElement\n  class=\"po-rich-text-body\"\n  tabindex=\"0\"\n  [attr.contenteditable]=\"!readonly\"\n  [attr.data-placeholder]=\"placeholder\"\n  [style.height.px]=\"height\"\n  (blur)=\"onBlur()\"\n  (click)=\"onClick()\"\n  (cut)=\"update()\"\n  (focus)=\"onFocus()\"\n  (keydown)=\"onKeyDown($event)\"\n  (keyup)=\"onKeyUp()\"\n  (paste)=\"update()\">\n</div>\n"
            }] }
];
PoRichTextBodyComponent.propDecorators = {
    bodyElement: [{ type: ViewChild, args: ['bodyElement', { static: true },] }],
    height: [{ type: Input, args: ['p-height',] }],
    modelValue: [{ type: Input, args: ['p-model-value',] }],
    placeholder: [{ type: Input, args: ['p-placeholder',] }],
    readonly: [{ type: Input, args: ['p-readonly',] }],
    change: [{ type: Output, args: ['p-change',] }],
    commands: [{ type: Output, args: ['p-commands',] }],
    shortcutCommand: [{ type: Output, args: ['p-shortcut-command',] }],
    value: [{ type: Output, args: ['p-value',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoRichTextBodyComponent.prototype.timeoutChange;
    /**
     * @type {?}
     * @private
     */
    PoRichTextBodyComponent.prototype.valueBeforeChange;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.bodyElement;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.height;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.modelValue;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.placeholder;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.readonly;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.change;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.commands;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.shortcutCommand;
    /** @type {?} */
    PoRichTextBodyComponent.prototype.value;
}
