// @portinari/portinari-ui/esm2015/lib/components/po-toolbar/po-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { PoToolbarBaseComponent } from './po-toolbar-base.component';
/**
 * \@docsExtends PoToolbarBaseComponent
 *
 * \@example
 *
 * <example name="po-toolbar-basic" title="Portinari Toolbar Basic">
 *  <file name="sample-po-toolbar-basic/sample-po-toolbar-basic.component.html"> </file>
 *  <file name="sample-po-toolbar-basic/sample-po-toolbar-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-toolbar-labs" title="Portinari Toolbar Labs">
 *  <file name="sample-po-toolbar-labs/sample-po-toolbar-labs.component.html"> </file>
 *  <file name="sample-po-toolbar-labs/sample-po-toolbar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-toolbar-logged" title="Portinari Toolbar - Logged">
 *  <file name="sample-po-toolbar-logged/sample-po-toolbar-logged.component.html"> </file>
 *  <file name="sample-po-toolbar-logged/sample-po-toolbar-logged.component.ts"> </file>
 * </example>
 */
export class PoToolbarComponent extends PoToolbarBaseComponent {
    /**
     * @param {?} titleService
     * @param {?} viewRef
     */
    constructor(titleService, viewRef) {
        super();
        this.titleService = titleService;
        this.parentRef = viewRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setTitle(this.title);
    }
    /**
     * @private
     * @param {?} newTitle
     * @return {?}
     */
    setTitle(newTitle) {
        this.titleService.setTitle(newTitle);
    }
}
PoToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-toolbar',
                template: "<div class=\"po-toolbar\">\n  <div class=\"po-toolbar-title\"> {{title}} </div>\n\n  <div class=\"po-toolbar-group-icon\">\n    <po-toolbar-actions *ngIf=\"actions && actions.length\"\n      [p-actions]=\"actions\"\n      [p-actions-icon]=\"actionsIcon\">\n    </po-toolbar-actions>\n\n    <po-toolbar-notification *ngIf=\"showNotification && notificationActions\"\n      [p-notification-actions]=\"notificationActions\"\n      [p-notification-number]=\"notificationNumber\">\n    </po-toolbar-notification>\n\n    <po-toolbar-profile *ngIf=\"isShowProfile\"\n      [p-profile]=\"profile\"\n      [p-profile-actions]=\"profileActions\">\n    </po-toolbar-profile>\n  </div>\n</div>\n"
            }] }
];
/** @nocollapse */
PoToolbarComponent.ctorParameters = () => [
    { type: Title },
    { type: ViewContainerRef }
];
if (false) {
    /** @type {?} */
    PoToolbarComponent.prototype.parentRef;
    /**
     * @type {?}
     * @private
     */
    PoToolbarComponent.prototype.titleService;
}
