// @portinari/portinari-ui/esm5/lib/components/po-tabs/po-tab/po-tab-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean, uuid } from '../../../utils/util';
/**
 * \@description
 *
 * O componente `po-tab` renderiza uma aba na qual envolve um conteúdo HTML.
 *
 * Com este componente é possível atribuir um rótulo para auxiliar na identificação do conteúdo, ativar para que o mesmo seja exibido,
 * desabilitar para impossibilitar o acesso, bem como ocultar para indisponibilizar a aba.
 *
 * > Para controlar a navegação entre diversas abas, utilizar o componente [`po-tabs`](/documentation/po-tabs).
 * @abstract
 */
var PoTabBaseComponent = /** @class */ (function () {
    function PoTabBaseComponent() {
        this._active = false;
        this._disabled = false;
        this._hide = false;
        // ID da aba
        this.id = uuid();
        /**
         * Método disparado ao clicar na aba.
         */
        this.click = new EventEmitter();
    }
    Object.defineProperty(PoTabBaseComponent.prototype, "active", {
        get: /**
         * @return {?}
         */
        function () {
            return this._active;
        },
        /**
         * @optional
         *
         * @description
         *
         * Ativa a aba exibindo seu conteúdo.
         *
         * > Sugere-se utilizar na aba de conteúdo inicial.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Ativa a aba exibindo seu conteúdo.
         *
         * > Sugere-se utilizar na aba de conteúdo inicial.
         *
         * \@default `false`
         * @param {?} active
         * @return {?}
         */
        function (active) {
            this._active = convertToBoolean(active);
            this.setDisplayOnActive();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTabBaseComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita a aba.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Desabilita a aba.
         *
         * \@default `false`
         * @param {?} disabled
         * @return {?}
         */
        function (disabled) {
            this._disabled = convertToBoolean(disabled);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTabBaseComponent.prototype, "hide", {
        get: /**
         * @return {?}
         */
        function () {
            return this._hide;
        },
        /**
         * @optional
         *
         * @description
         *
         * Oculta a aba.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Oculta a aba.
         *
         * \@default `false`
         * @param {?} hide
         * @return {?}
         */
        function (hide) {
            this._hide = convertToBoolean(hide);
        },
        enumerable: true,
        configurable: true
    });
    PoTabBaseComponent.propDecorators = {
        active: [{ type: Input, args: ['p-active',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        hide: [{ type: Input, args: ['p-hide',] }],
        label: [{ type: Input, args: ['p-label',] }],
        click: [{ type: Output, args: ['p-click',] }]
    };
    return PoTabBaseComponent;
}());
export { PoTabBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._active;
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    PoTabBaseComponent.prototype._hide;
    /** @type {?} */
    PoTabBaseComponent.prototype.id;
    /**
     * Rótulo da aba.
     * @type {?}
     */
    PoTabBaseComponent.prototype.label;
    /**
     * Método disparado ao clicar na aba.
     * @type {?}
     */
    PoTabBaseComponent.prototype.click;
    /**
     * @abstract
     * @protected
     * @return {?}
     */
    PoTabBaseComponent.prototype.setDisplayOnActive = function () { };
}
