// @portinari/portinari-ui/esm2015/lib/components/po-disclaimer/po-disclaimer-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
/** @type {?} */
const PO_DISCLAIMER_TYPES = ['default', 'danger'];
/** @type {?} */
const PO_DISCLAIMER_DEFAULT_TYPE = 'default';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * O componente po-disclaimer é responsável por representar tags.
 * Seu uso é recomendado em buscas e em campos onde é necessário representar objetos selecionados,
 * como por exemplo, no po-multi-select.
 *
 */
export class PoDisclaimerBaseComponent {
    constructor() {
        this._type = 'default';
        this._hideClose = false;
        this.showDisclaimer = true;
        /**
         * Evento disparado ao fechar o disclaimer.
         * Para este evento será passado como parâmetro um objeto com value, label e property.
         */
        this.closeAction = new EventEmitter();
    }
    /**
     * \@description
     *
     * Esta propriedade esconde o botão para fechamento do po-disclaimer, ao utilizar esta propriedade
     * sem passar valor a mesma é setada como false, onde o botão de fechamento está visível.
     *
     * \@default false
     * @param {?} value
     * @return {?}
     */
    set hideClose(value) {
        this._hideClose = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get hideClose() {
        return this._hideClose;
    }
    /**
     * \@description
     *
     * Tipo do po-disclaimer. Pode ser 'default' ou 'danger'.
     *
     * \@default default
     * \@optional
     * @param {?} type
     * @return {?}
     */
    set type(type) {
        this._type = PO_DISCLAIMER_TYPES.includes(type) ? type : PO_DISCLAIMER_DEFAULT_TYPE;
    }
    /**
     * @return {?}
     */
    get type() {
        return this._type;
    }
    /**
     * @return {?}
     */
    close() {
        this.showDisclaimer = false;
        this.closeAction.emit({ value: this.value, label: this.label, property: this.property });
    }
    /**
     * @return {?}
     */
    getLabel() {
        return this.label ? this.label : this.value;
    }
}
PoDisclaimerBaseComponent.propDecorators = {
    label: [{ type: Input, args: ['p-label',] }],
    value: [{ type: Input, args: ['p-value',] }],
    property: [{ type: Input, args: ['p-property',] }],
    hideClose: [{ type: Input, args: ['p-hide-close',] }],
    closeAction: [{ type: Output, args: ['p-close-action',] }],
    type: [{ type: Input, args: ['p-type',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoDisclaimerBaseComponent.prototype._type;
    /**
     * @type {?}
     * @private
     */
    PoDisclaimerBaseComponent.prototype._hideClose;
    /** @type {?} */
    PoDisclaimerBaseComponent.prototype.showDisclaimer;
    /**
     * Label que aparecerá dentro do po-disclaimer.
     * Quando não for definido um label será apresentada a propriedade p-value.
     * @type {?}
     */
    PoDisclaimerBaseComponent.prototype.label;
    /**
     * Valor do po-disclaimer.
     * @type {?}
     */
    PoDisclaimerBaseComponent.prototype.value;
    /**
     * Nome da propriedade vinculada à este po-disclaimer.
     * @type {?}
     */
    PoDisclaimerBaseComponent.prototype.property;
    /**
     * Evento disparado ao fechar o disclaimer.
     * Para este evento será passado como parâmetro um objeto com value, label e property.
     * @type {?}
     */
    PoDisclaimerBaseComponent.prototype.closeAction;
}
