// @portinari/portinari-ui/esm2015/lib/components/po-slide/po-slide.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ElementRef, HostListener, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { animate, AnimationBuilder, keyframes, style } from '@angular/animations';
import { PoSlideBaseComponent } from './po-slide-base.component';
import { PoSlideContentTemplateDirective } from './directives/po-slide-content-template.directive';
import { PoSlideItemComponent } from './po-slide-item/po-slide-item.component';
/** @type {?} */
const poSlideDefaultHeight = 336;
/** @type {?} */
const poSlideIntervalMin = 1000;
/** @type {?} */
const poSlideMinHeight = 192;
/** @type {?} */
const poSlideTiming = '250ms ease';
/**
 * \@docsExtends PoSlideBaseComponent
 *
 * \@example
 * <example name="po-slide-basic" title="Portinari Slide Basic">
 *   <file name="sample-po-slide-basic/sample-po-slide-basic.component.html"> </file>
 *   <file name="sample-po-slide-basic/sample-po-slide-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-slide-labs" title="Portinari Slide Labs">
 *   <file name="sample-po-slide-labs/sample-po-slide-labs.component.html"> </file>
 *   <file name="sample-po-slide-labs/sample-po-slide-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-slide-useful-articles" title="Portinari Slide - Useful articles">
 *   <file name="sample-po-slide-useful-articles/sample-po-slide-useful-articles.component.html"> </file>
 *   <file name="sample-po-slide-useful-articles/sample-po-slide-useful-articles.component.ts"> </file>
 * </example>
 *
 * <example name="po-slide-landscapes" title="Portinari Slide - Landscapes">
 *   <file name="sample-po-slide-landscapes/sample-po-slide-landscapes.component.html"> </file>
 *   <file name="sample-po-slide-landscapes/sample-po-slide-landscapes.component.ts"> </file>
 * </example>
 */
export class PoSlideComponent extends PoSlideBaseComponent {
    /**
     * @param {?} builder
     */
    constructor(builder) {
        super();
        this.builder = builder;
        this.isLoaded = false;
        this.currentSlideIndex = 0;
        this.slideItems = [];
    }
    /**
     * @private
     * @return {?}
     */
    get hasElements() {
        return !!this.slide.nativeElement.offsetWidth && !!this.itemsElements && !!this.itemsElements.length;
    }
    /**
     * @private
     * @return {?}
     */
    get isImageSlide() {
        return !this.slideContentTemplate;
    }
    /**
     * @private
     * @return {?}
     */
    get offset() {
        return this.currentSlideIndex * this.slideItemWidth;
    }
    /**
     * @return {?}
     */
    get hasSlides() {
        return !!this.slides && !!this.slides.length;
    }
    /**
     * @return {?}
     */
    onResize() {
        if (this.slide) {
            this.setSlideItemWidth();
            this.goToItem(this.currentSlideIndex);
        }
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        if (!this.isLoaded && this.hasElements) {
            this.setSlideItemWidth();
            this.isLoaded = true;
            if (this.hasSlides) {
                this.startSlide();
            }
        }
    }
    /**
     * @param {?} index
     * @return {?}
     */
    goToItem(index) {
        if (this.interval > poSlideIntervalMin) {
            this.startInterval();
        }
        this.currentSlideIndex = index;
        this.animate(this.offset);
    }
    /**
     * @return {?}
     */
    nextControl() {
        if (this.interval > poSlideIntervalMin) {
            this.startInterval();
        }
        this.next();
    }
    /**
     * @return {?}
     */
    next() {
        if (this.currentSlideIndex + 1 === this.slideItems.length) {
            this.currentSlideIndex = 0;
            this.animate(0);
            return;
        }
        this.currentSlideIndex = (this.currentSlideIndex + 1) % this.slideItems.length;
        this.animate(this.offset);
    }
    /**
     * @return {?}
     */
    previous() {
        if (this.currentSlideIndex === 0) {
            this.currentSlideIndex = this.slideItems.length - 1;
            this.animate(this.offset);
            return;
        }
        this.currentSlideIndex = ((this.currentSlideIndex - 1) + this.slideItems.length) % this.slideItems.length;
        this.animate(this.offset);
    }
    /**
     * @return {?}
     */
    previousControl() {
        if (this.interval > poSlideIntervalMin) {
            this.startInterval();
        }
        this.previous();
    }
    /**
     * @param {?} height
     * @return {?}
     */
    setSlideHeight(height) {
        this.setHeight(height);
    }
    /**
     * @private
     * @param {?} offset
     * @return {?}
     */
    animate(offset) {
        if (this.hasElements) {
            /** @type {?} */
            const animation = this.buildTransitionAnimation(offset);
            this.player = animation.create(this.slide.nativeElement);
            this.player.play();
        }
    }
    /**
     * @private
     * @param {?} offset
     * @return {?}
     */
    buildTransitionAnimation(offset) {
        return this.builder.build([
            animate(poSlideTiming, keyframes([style({ transform: `translateX(-${offset}px)` })]))
        ]);
    }
    /**
     * @private
     * @param {?} slides
     * @return {?}
     */
    createArrayForTemplate(slides) {
        this.slideItems = [...slides];
    }
    /**
     * @private
     * @param {?} slides
     * @return {?}
     */
    createArrayFromSlides(slides) {
        /** @type {?} */
        const isStringArray = slides.every((/**
         * @param {?} item
         * @return {?}
         */
        item => typeof item === 'string'));
        if (isStringArray) {
            slides.forEach((/**
             * @param {?} element
             * @return {?}
             */
            element => this.slideItems.push({ image: `${element}` })));
        }
        else {
            this.slideItems = [...(/** @type {?} */ (slides))];
        }
    }
    /**
     * @private
     * @param {?} height
     * @return {?}
     */
    setDefaultHeight(height) {
        if (height && height <= poSlideMinHeight || !height && this.isImageSlide) {
            this.slide.nativeElement.style.height = `${poSlideDefaultHeight}px`;
            this.imageHeight = poSlideDefaultHeight;
        }
        else {
            this.imageHeight = undefined;
        }
    }
    /**
     * @private
     * @param {?} height
     * @return {?}
     */
    setHeight(height) {
        if (height && height > poSlideMinHeight) {
            this.slide.nativeElement.style.height = `${height}px`;
            this.imageHeight = height;
        }
        else {
            this.setDefaultHeight(height);
        }
    }
    /**
     * @private
     * @return {?}
     */
    setSlideItemWidth() {
        if (this.hasElements) {
            this.slideItemWidth = this.itemsElements.first.itemElement.nativeElement.getBoundingClientRect().width;
        }
    }
    /**
     * @protected
     * @return {?}
     */
    cancelInterval() {
        clearInterval(this.setInterval);
    }
    /**
     * @protected
     * @param {?} slides
     * @return {?}
     */
    setSlideItems(slides) {
        if (this.hasSlides) {
            this.slideContentTemplate ? this.createArrayForTemplate(slides) : this.createArrayFromSlides(slides);
        }
        else {
            this.slideItems = [];
            this.cancelInterval();
        }
    }
    /**
     * @protected
     * @return {?}
     */
    startSlide() {
        this.setSlideHeight(this.height);
        if (this.interval > poSlideIntervalMin) {
            this.startInterval();
        }
        this.currentSlideIndex = 0;
    }
    /**
     * @protected
     * @return {?}
     */
    startInterval() {
        if (this.setInterval) {
            this.cancelInterval();
        }
        this.setInterval = this.hasSlides && this.hasElements ? setInterval((/**
         * @return {?}
         */
        () => { this.next(); }), this.interval) : undefined;
    }
}
PoSlideComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-slide',
                template: "<div class=\"po-slide-wrapper\">\n  <div class=\"po-slide-outer\">\n    <div class=\"po-slide-inner\" #slide>\n      <ng-container *ngFor=\"let item of slideItems; let index = index;\">\n        <po-slide-item\n          [p-action]=\"item.action\"\n          [p-data]=\"item\"\n          [p-image]=\"item.image\"\n          [p-image-height]=\"imageHeight\"\n          [p-template]=\"slideContentTemplate\"\n          [p-alt]=\"item.alt\"\n          [p-link]=\"item.link\">\n        </po-slide-item>\n      </ng-container>\n    </div>\n\n    <po-slide-control *ngIf=\"hasSlides && slides.length > 1\" \n      p-control=\"previous\"\n      (p-click)=\"previousControl()\">\n    </po-slide-control>\n\n    <po-slide-control *ngIf=\"hasSlides && slides.length > 1\"\n      p-control=\"next\"\n      (p-click)=\"nextControl()\">\n    </po-slide-control>\n\n  </div>\n\n  <po-slide-circles\n    *ngIf=\"hasSlides && slides.length > 1\"\n    [p-current-slide-index]=\"currentSlideIndex\"\n    [p-items]=\"slideItems\"\n    (p-click)=\"goToItem($event)\">\n  </po-slide-circles>\n</div>\n"
            }] }
];
/** @nocollapse */
PoSlideComponent.ctorParameters = () => [
    { type: AnimationBuilder }
];
PoSlideComponent.propDecorators = {
    slideContentTemplate: [{ type: ContentChild, args: [PoSlideContentTemplateDirective, { static: true },] }],
    slide: [{ type: ViewChild, args: ['slide', { static: true },] }],
    itemsElements: [{ type: ViewChildren, args: [PoSlideItemComponent,] }],
    onResize: [{ type: HostListener, args: ['window:resize',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.isLoaded;
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.player;
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.setInterval;
    /** @type {?} */
    PoSlideComponent.prototype.currentSlideIndex;
    /** @type {?} */
    PoSlideComponent.prototype.imageHeight;
    /** @type {?} */
    PoSlideComponent.prototype.slideItems;
    /** @type {?} */
    PoSlideComponent.prototype.slideItemWidth;
    /** @type {?} */
    PoSlideComponent.prototype.slideContentTemplate;
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.slide;
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.itemsElements;
    /**
     * @type {?}
     * @private
     */
    PoSlideComponent.prototype.builder;
}
