// @portinari/portinari-ui/esm2015/lib/components/po-field/po-lookup/services/po-lookup-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço padrão utilizado para filtrar os dados do componente po-lookup.
 */
export class PoLookupFilterService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * @param {?} filteredItemsParams
     * @return {?}
     */
    getFilteredItems(filteredItemsParams) {
        const { filterParams } = filteredItemsParams, restFilteredItemsParams = tslib_1.__rest(filteredItemsParams, ["filterParams"]);
        /** @type {?} */
        const validatedFilterParams = this.validateParams(filterParams);
        /** @type {?} */
        const params = Object.assign({}, restFilteredItemsParams, validatedFilterParams);
        return this.httpClient.get(this.url, { params });
    }
    /**
     * @param {?} value
     * @param {?=} filterParams
     * @return {?}
     */
    getObjectByValue(value, filterParams) {
        /** @type {?} */
        const validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(`${this.url}/${value}`, { params: validatedFilterParams });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    setUrl(url) {
        this.url = url;
    }
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    validateParams(params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    }
}
PoLookupFilterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PoLookupFilterService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoLookupFilterService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    PoLookupFilterService.prototype.httpClient;
}
