// @portinari/portinari-ui/esm5/lib/components/po-menu/po-menu-item/po-menu-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { convertToInt, convertToBoolean } from '../../../utils/util';
import { PoMenuItemsService } from '../services/po-menu-items.service';
// valor para que caibam 3 linhas de `label`
/** @type {?} */
var poMenuItemSubItemSize = 98;
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que implementa cada item do po-menu.
 */
var PoMenuItemComponent = /** @class */ (function () {
    function PoMenuItemComponent(menuItemsService) {
        this.menuItemsService = menuItemsService;
        this._isSelected = false;
        this._isSubItem = false;
        this.maxHeight = 0;
    }
    Object.defineProperty(PoMenuItemComponent.prototype, "badgeValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._badgeValue;
        },
        // Valor do badge.
        set: 
        // Valor do badge.
        /**
         * @param {?} badgeValue
         * @return {?}
         */
        function (badgeValue) {
            this._badgeValue = convertToInt(badgeValue);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuItemComponent.prototype, "isSelected", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isSelected;
        },
        // Indica se o item está selecionado.
        set: 
        // Indica se o item está selecionado.
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isSelected = convertToBoolean(value);
            this.isSelectedSubItem = this.isSelected && this.isSubItem;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuItemComponent.prototype, "isSubItem", {
        get: /**
         * @return {?}
         */
        function () {
            return this._isSubItem;
        },
        // Indica se o item é um sub item
        set: 
        // Indica se o item é um sub item
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._isSubItem = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuItemComponent.prototype, "subItems", {
        get: /**
         * @return {?}
         */
        function () {
            return this._subItems;
        },
        // Lista de sub-items.
        set: 
        // Lista de sub-items.
        /**
         * @param {?} subitems
         * @return {?}
         */
        function (subitems) {
            this._subItems = subitems;
            if (this.isOpened) {
                this.calcMenuSubItemsMaxHeight();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoMenuItemComponent.prototype, "canShowBadge", {
        get: /**
         * @return {?}
         */
        function () {
            return this.type !== 'subItems' && (this.badgeValue || this.badgeValue === 0) && this.badgeValue >= 0;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoMenuItemComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.itemSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    PoMenuItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // subscribe to menu component messages
        this.itemSubscription = this.menuItemsService.receiveFromParentMenuClicked().subscribe((/**
         * @param {?} menu
         * @return {?}
         */
        function (menu) {
            _this.processMenuItem(menu);
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PoMenuItemComponent.prototype.clickMenuItem = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!(event.ctrlKey || event.metaKey)) {
            event.preventDefault();
            // Emmit to parent
            this.menuItemsService.sendToParentMenuClicked({
                link: this.link,
                action: this.action,
                id: this.id,
                icon: this.icon,
                label: this.label,
                level: this.level,
                subItems: this.subItems,
                isSelected: this.isSelected,
                isOpened: this.isOpened,
                shortLabel: this.shortLabel,
                type: this.type
            });
        }
    };
    /**
     * @private
     * @param {?} menuActive
     * @param {?} menuOpened
     * @param {?} hasSubItemOpened
     * @param {?} activatedByRoute
     * @return {?}
     */
    PoMenuItemComponent.prototype.accordionAnimation = /**
     * @private
     * @param {?} menuActive
     * @param {?} menuOpened
     * @param {?} hasSubItemOpened
     * @param {?} activatedByRoute
     * @return {?}
     */
    function (menuActive, menuOpened, hasSubItemOpened, activatedByRoute) {
        if (this.id === menuOpened['id']) {
            this.maxHeight = this.subItems.length * poMenuItemSubItemSize;
        }
        if (hasSubItemOpened) {
            this.maxHeight = menuOpened['isOpened'] ?
                (this.maxHeight + menuOpened.subItems.length * poMenuItemSubItemSize) :
                (this.maxHeight - menuOpened.subItems.length * poMenuItemSubItemSize);
            if (activatedByRoute) {
                this.maxHeight = this.getMinimumHeight(0, this, menuActive);
            }
        }
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuItemComponent.prototype.activateMenu = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.isSelected = menu && this.id === menu.id;
    };
    /**
     * @private
     * @return {?}
     */
    PoMenuItemComponent.prototype.calcMenuSubItemsMaxHeight = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var subItems = Array.from(_this.menuSubItems.nativeElement.querySelectorAll('.po-menu-item'));
            subItems.forEach((/**
             * @param {?} menuItem
             * @return {?}
             */
            function (menuItem) { return _this.maxHeight += menuItem.offsetHeight; }));
        }));
    };
    /**
     * @private
     * @param {?} minimumHeight
     * @param {?} menuItem
     * @param {?} menuActive
     * @return {?}
     */
    PoMenuItemComponent.prototype.getMinimumHeight = /**
     * @private
     * @param {?} minimumHeight
     * @param {?} menuItem
     * @param {?} menuActive
     * @return {?}
     */
    function (minimumHeight, menuItem, menuActive) {
        minimumHeight += poMenuItemSubItemSize;
        if (menuItem.subItems && this.hasSubItem(menuItem.subItems, menuActive['id'])) {
            for (var index = 0; index < menuItem.subItems.length; index++) {
                minimumHeight = this.getMinimumHeight(minimumHeight, menuItem.subItems[index], menuActive);
            }
        }
        return minimumHeight;
    };
    /**
     * @private
     * @param {?} menuActive
     * @param {?} menuOpened
     * @param {?=} activatedByRoute
     * @return {?}
     */
    PoMenuItemComponent.prototype.groupedMenu = /**
     * @private
     * @param {?} menuActive
     * @param {?} menuOpened
     * @param {?=} activatedByRoute
     * @return {?}
     */
    function (menuActive, menuOpened, activatedByRoute) {
        if (activatedByRoute === void 0) { activatedByRoute = false; }
        /** @type {?} */
        var hasSubItemOpened = (menuOpened && this.id !== menuOpened['id']) ? this.hasSubItem(this.subItems, menuOpened['id']) : false;
        this.isOpened = this.isMenuOpened(menuOpened, hasSubItemOpened);
        this.isSelected = (menuActive && !this.isOpened) ? this.hasSubItem(this.subItems, menuActive['id']) : false;
        if (!this.isOpened) {
            this.maxHeight = 0;
            return;
        }
        this.accordionAnimation(menuActive, menuOpened, hasSubItemOpened, activatedByRoute);
    };
    /**
     * @private
     * @param {?} subItems
     * @param {?} id
     * @return {?}
     */
    PoMenuItemComponent.prototype.hasSubItem = /**
     * @private
     * @param {?} subItems
     * @param {?} id
     * @return {?}
     */
    function (subItems, id) {
        var _this = this;
        if (subItems) {
            return subItems.some((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                return item['id'] === id ? true : _this.hasSubItem(item.subItems, id);
            }));
        }
    };
    /**
     * @private
     * @param {?} menuOpened
     * @param {?} hasSubItemOpened
     * @return {?}
     */
    PoMenuItemComponent.prototype.isMenuOpened = /**
     * @private
     * @param {?} menuOpened
     * @param {?} hasSubItemOpened
     * @return {?}
     */
    function (menuOpened, hasSubItemOpened) {
        if (menuOpened) {
            return (this.id === menuOpened['id']) ? menuOpened['isOpened'] : hasSubItemOpened;
        }
        return false;
    };
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    PoMenuItemComponent.prototype.processMenuItem = /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        if (this.type === 'internalLink') {
            this.activateMenu(menu.active);
            return;
        }
        if (this.type === 'subItems') {
            this.groupedMenu(menu.active, menu.grouped, menu.activatedByRoute);
            return;
        }
    };
    PoMenuItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-menu-item',
                    template: "<!-- menu com link interno -->\n<a *ngIf=\"type === 'internalLink'\" class=\"po-menu-item-link\" [routerLink]=\"link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n<!-- menu com link externo -->\n<a *ngIf=\"type === 'externalLink'\" class=\"po-menu-item-link\" [href]=\"link\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n<!-- menu sem link -->\n<a *ngIf=\"type === 'noLink'\" class=\"po-menu-item-link\" href=\"javascript:;\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate\"></ng-container>\n</a>\n<!-- menu com sub itens -->\n<div *ngIf=\"type === 'subItems'\" class=\"po-menu-item-link po-clickable\">\n  <ng-container *ngTemplateOutlet=\"menuItemTemplate;\"></ng-container>\n  <div #menuSubItems\n    class=\"po-menu-sub-items\"\n    [hidden]=\"collapsedMenu\"\n    [style.maxHeight.px]=\"maxHeight\">\n    <div *ngFor=\"let subItem of subItems\">\n      <po-menu-item\n        p-is-sub-item\n        [p-action]=\"subItem.action\"\n        [p-badge-alert]=\"subItem.badgeAlert\"\n        [p-badge-color]=\"subItem.badge ? subItem.badge.color : undefined\"\n        [p-badge-value]=\"subItem.badge ? subItem.badge.value : undefined\"\n        [p-id]=\"subItem.id\"\n        [p-label]=\"subItem.label\"\n        [p-level]=\"subItem.level\"\n        [p-link]=\"subItem.link\"\n        [p-sub-items]=\"subItem.subItems\"\n        [p-type]=\"subItem.type\">\n      </po-menu-item>\n    </div>\n  </div>\n</div>\n\n<ng-template #menuItemTemplate>\n  <div class=\"po-menu-item\"\n    [class.po-menu-icon-container]=\"level === 1 && icon\"\n    [class.po-menu-item-selected]=\"isSelected\"\n    [class.po-menu-item-level-two]=\"level === 2\"\n    [class.po-menu-item-level-three]=\"level === 3\"\n    [class.po-menu-item-level-four]=\"level === 4\"\n    [class.po-menu-item-grouper-up]=\"type === 'subItems' && isOpened\"\n    [class.po-menu-item-grouper-down]=\"type === 'subItems' && !isOpened\"\n    [class.po-menu-sub-item-selected]=\"isSelectedSubItem\"\n    (click)=\"clickMenuItem($event);\">\n    <po-badge *ngIf=\"canShowBadge\"\n      [ngClass]=\"!collapsedMenu ? 'po-menu-badge-align' : 'po-menu-badge-align-collapsed'\"\n      [p-color]=\"badgeColor\"\n      [p-value]=\"badgeValue\">\n    </po-badge>\n    <span *ngIf=\"level === 1 && icon\" class=\"po-icon {{icon}} po-menu-icon-item\"></span>\n    <div *ngIf=\"badgeAlert\"\n      class=\"po-color-07\"\n      [ngClass]=\"!collapsedMenu ? 'po-menu-badge-alert' : 'po-menu-badge-alert-collapsed'\">\n    </div>\n    <span *ngIf=\"type === 'subItems' && !collapsedMenu\"\n      class=\"po-icon po-menu-group-icon\"\n      [class.po-icon-arrow-up]=\"isOpened\"\n      [class.po-icon-arrow-down]=\"!isOpened\">\n    </span>\n    <div [class.po-menu-icon-label]=\"level === 1 && icon\">\n      {{ label }}\n    </div>\n    <div *ngIf=\"collapsedMenu\" class=\"po-menu-short-label\">{{ shortLabel }}</div>\n  </div>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    PoMenuItemComponent.ctorParameters = function () { return [
        { type: PoMenuItemsService }
    ]; };
    PoMenuItemComponent.propDecorators = {
        action: [{ type: Input, args: ['p-action',] }],
        badgeAlert: [{ type: Input, args: ['p-badge-alert',] }],
        badgeColor: [{ type: Input, args: ['p-badge-color',] }],
        badgeValue: [{ type: Input, args: ['p-badge-value',] }],
        collapsedMenu: [{ type: Input, args: ['p-collapsed-menu',] }],
        icon: [{ type: Input, args: ['p-icon',] }],
        id: [{ type: Input, args: ['p-id',] }],
        isOpened: [{ type: Input, args: ['p-is-opened',] }],
        isSelected: [{ type: Input, args: ['p-is-selected',] }],
        isSubItem: [{ type: Input, args: ['p-is-sub-item',] }],
        label: [{ type: Input, args: ['p-label',] }],
        level: [{ type: Input, args: ['p-level',] }],
        link: [{ type: Input, args: ['p-link',] }],
        shortLabel: [{ type: Input, args: ['p-short-label',] }],
        subItems: [{ type: Input, args: ['p-sub-items',] }],
        type: [{ type: Input, args: ['p-type',] }],
        menuSubItems: [{ type: ViewChild, args: ['menuSubItems', { static: false },] }]
    };
    return PoMenuItemComponent;
}());
export { PoMenuItemComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype._badgeValue;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype._isSelected;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype._isSubItem;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype._subItems;
    /** @type {?} */
    PoMenuItemComponent.prototype.isSelectedSubItem;
    /** @type {?} */
    PoMenuItemComponent.prototype.maxHeight;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype.itemSubscription;
    /** @type {?} */
    PoMenuItemComponent.prototype.action;
    /** @type {?} */
    PoMenuItemComponent.prototype.badgeAlert;
    /** @type {?} */
    PoMenuItemComponent.prototype.badgeColor;
    /** @type {?} */
    PoMenuItemComponent.prototype.collapsedMenu;
    /** @type {?} */
    PoMenuItemComponent.prototype.icon;
    /** @type {?} */
    PoMenuItemComponent.prototype.id;
    /** @type {?} */
    PoMenuItemComponent.prototype.isOpened;
    /** @type {?} */
    PoMenuItemComponent.prototype.label;
    /** @type {?} */
    PoMenuItemComponent.prototype.level;
    /** @type {?} */
    PoMenuItemComponent.prototype.link;
    /** @type {?} */
    PoMenuItemComponent.prototype.shortLabel;
    /** @type {?} */
    PoMenuItemComponent.prototype.type;
    /** @type {?} */
    PoMenuItemComponent.prototype.menuSubItems;
    /**
     * @type {?}
     * @private
     */
    PoMenuItemComponent.prototype.menuItemsService;
}
