// @portinari/portinari-ui/esm5/lib/components/po-chart/po-chart-types/po-chart-dynamic-type.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
/** @type {?} */
var Padding = 24;
/**
 * @abstract
 */
var PoChartDynamicTypeComponent = /** @class */ (function () {
    function PoChartDynamicTypeComponent() {
        this.innerRadius = 0;
        this.onSerieClick = new Subject();
        this.onSerieHover = new Subject();
        this.series = [];
    }
    /**
     * @param {?} chartWrapperElement
     * @param {?} chartHeaderElement
     * @param {?} chartLegendElement
     * @return {?}
     */
    PoChartDynamicTypeComponent.prototype.calculateSVGContainerDimensions = /**
     * @param {?} chartWrapperElement
     * @param {?} chartHeaderElement
     * @param {?} chartLegendElement
     * @return {?}
     */
    function (chartWrapperElement, chartHeaderElement, chartLegendElement) {
        /** @type {?} */
        var svgContainerHeightCalc = this.height - chartHeaderElement - chartLegendElement - (Padding * 2);
        this.svgHeight = svgContainerHeightCalc <= 0 ? 0 : svgContainerHeightCalc;
        this.centerX = chartWrapperElement / 2;
    };
    /**
     * @return {?}
     */
    PoChartDynamicTypeComponent.prototype.calculateTotalValue = /**
     * @return {?}
     */
    function () {
        this.totalValue = this.series.reduce((/**
         * @param {?} previousValue
         * @param {?} serie
         * @return {?}
         */
        function (previousValue, serie) { return previousValue + serie.value; }), 0);
    };
    PoChartDynamicTypeComponent.propDecorators = {
        chartBody: [{ type: ViewChild, args: ['chartBody', { static: true },] }],
        svgContainer: [{ type: ViewChild, args: ['svgContainer', { static: true },] }]
    };
    return PoChartDynamicTypeComponent;
}());
export { PoChartDynamicTypeComponent };
if (false) {
    /**
     * @type {?}
     * @protected
     */
    PoChartDynamicTypeComponent.prototype.windowResizeListener;
    /**
     * @type {?}
     * @protected
     */
    PoChartDynamicTypeComponent.prototype.windowScrollListener;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.centerX;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartElementCategory;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartElementValue;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartHeader;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartLegend;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartWrapper;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.colors;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.height;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.innerRadius;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.onSerieClick;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.onSerieHover;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.series;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.svgElement;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.svgHeight;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.target;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.tooltipElement;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.tooltipText;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.totalValue;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.type;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.chartBody;
    /** @type {?} */
    PoChartDynamicTypeComponent.prototype.svgContainer;
}
