// @portinari/portinari-ui/esm2015/lib/components/po-grid/po-grid-head/po-grid-head.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ChangeDetectorRef } from '@angular/core';
export class PoGridHeadComponent {
    /**
     * @param {?} changeDetectorRef
     */
    constructor(changeDetectorRef) {
        this.changeDetectorRef = changeDetectorRef;
        this.align = 'left';
        this.boolean = false;
        this.position = '';
        this.tabIndex = -1;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set width(value) {
        this._width = value;
    }
    /**
     * @return {?}
     */
    get width() {
        return this._width ? `${this._width}px` : '100%';
    }
}
PoGridHeadComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-grid-head',
                template: "<div>\n  <div [id]=\"position\"\n    class=\"po-grid-head-content po-grid-head-{{ align }}\"\n    tabindex=\"{{ title ? 0 : -1 }}\">\n    <span>{{ title || '&nbsp;' }}</span>\n  </div>\n</div>\n\n",
                styles: [`
  .po-grid-head-content {
    border-bottom: 2px solid #ceeaf2;
    border-left: none;
    border-right: 1px solid #eceeee;
    border-top: none;
    color: #0c9abe;
    font-family: NunitoSans-bold;
    font-size: 14px;
    height: 44px;
    line-height: 1.71;
    margin: 0;
    max-height: 44px;
    min-height: 44px;
    overflow: hidden;
    padding: 10px 16px;
    text-align: left;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .po-grid-head-center {
    text-align: center;
  }

  .po-grid-head-right {
    text-align: right;
  }

  .po-grid-head-content:focus {
    border: 1px solid #0c9abe;
    outline: none;
  }
  `]
            }] }
];
/** @nocollapse */
PoGridHeadComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PoGridHeadComponent.propDecorators = {
    align: [{ type: Input, args: ['p-align',] }],
    boolean: [{ type: Input, args: ['p-freeze',] }],
    position: [{ type: Input, args: ['p-position',] }],
    tabIndex: [{ type: Input, args: ['p-tab-index',] }],
    width: [{ type: Input, args: ['p-width',] }],
    title: [{ type: Input, args: ['p-title',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoGridHeadComponent.prototype._width;
    /** @type {?} */
    PoGridHeadComponent.prototype.align;
    /** @type {?} */
    PoGridHeadComponent.prototype.boolean;
    /** @type {?} */
    PoGridHeadComponent.prototype.position;
    /** @type {?} */
    PoGridHeadComponent.prototype.tabIndex;
    /** @type {?} */
    PoGridHeadComponent.prototype.title;
    /**
     * @type {?}
     * @private
     */
    PoGridHeadComponent.prototype.changeDetectorRef;
}
