// @portinari/portinari-ui/esm2015/lib/components/po-table/po-table-column-manager/po-table-column-manager.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
import { browserLanguage, capitalizeFirstLetter, convertToInt, poLocaleDefault } from '../../../utils/util';
import { PoPopoverComponent } from '../../po-popover/po-popover.component';
/** @type {?} */
const PoTableColumnManagerMaxColumnsDefault = 99999;
/** @type {?} */
export const poTableColumnManagerLiteralsDefault = {
    en: {
        columnsManager: 'Columns manager',
        restoreDefault: 'Restore default'
    },
    es: {
        columnsManager: 'Gerente de columna',
        restoreDefault: 'Restaurar por defecto'
    },
    pt: {
        columnsManager: 'Gerenciador de colunas',
        restoreDefault: 'Restaurar padrão'
    },
    ru: {
        columnsManager: 'менеджер колонок',
        restoreDefault: 'сброс настроек'
    }
};
export class PoTableColumnManagerComponent {
    /**
     * @param {?} renderer
     */
    constructor(renderer) {
        this.renderer = renderer;
        this._maxColumns = PoTableColumnManagerMaxColumnsDefault;
        this.columnsOptions = [];
        this.literals = Object.assign({}, poTableColumnManagerLiteralsDefault[poLocaleDefault], poTableColumnManagerLiteralsDefault[browserLanguage()]);
        this.visibleColumns = [];
        this.defaultColumns = [];
        this.columns = [];
        this.visibleColumnsChange = new EventEmitter();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set maxColumns(value) {
        this._maxColumns = convertToInt(value, PoTableColumnManagerMaxColumnsDefault);
    }
    /**
     * @return {?}
     */
    get maxColumns() {
        return this._maxColumns;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.updateColumnsOptions(this.columns);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        const { columns, maxColumns, target } = changes;
        if (target && target.firstChange) {
            this.initializeListeners();
        }
        if (columns) {
            this.onChangeColumns(columns);
        }
        if (maxColumns) {
            this.updateColumnsOptions(this.columns);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeListeners();
    }
    /**
     * @param {?} checkedColumns
     * @return {?}
     */
    onChangeVisibleColumns(checkedColumns) {
        this.disableColumnsOptions(this.columnsOptions);
        /** @type {?} */
        const visibleTableColumns = this.getVisibleTableColumns(checkedColumns);
        this.visibleColumnsChange.emit(visibleTableColumns);
    }
    /**
     * @return {?}
     */
    restore() {
        this.updateColumnsOptions(this.defaultColumns);
    }
    // desabilitará as colunas, que não estiverem selecionadas, após exeder o numero maximo de colunas.
    /**
     * @private
     * @param {?=} columns
     * @return {?}
     */
    disableColumnsOptions(columns = []) {
        // necessario timeout para que seja possivel atualizar os columnsOptions apos a mudança do model
        setTimeout((/**
         * @return {?}
         */
        () => {
            this.columnsOptions = columns.map((/**
             * @param {?} column
             * @return {?}
             */
            column => (Object.assign({}, column, { disabled: this.isDisableColumn(column.value) }))));
        }));
    }
    /**
     * @private
     * @param {?} column
     * @return {?}
     */
    getColumnTitleLabel(column) {
        return column.label || capitalizeFirstLetter(column.property);
    }
    /**
     * Retorna um Array de column.property das colunas que são visiveis.
     * @private
     * @param {?} columns
     * @return {?}
     */
    getVisibleColumns(columns) {
        /** @type {?} */
        const visibleColumns = [];
        columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        column => {
            if (column.visible !== false && visibleColumns.length < this.maxColumns && column.type !== 'detail') {
                visibleColumns.push(column.property);
            }
        }));
        return visibleColumns;
    }
    /**
     * Retorna um Array PoTableColumn a partir das colunas visiveis no gerenciador de colunas.
     * @private
     * @param {?} visibleColumns
     * @return {?}
     */
    getVisibleTableColumns(visibleColumns) {
        return this.columns.map((/**
         * @param {?} column
         * @return {?}
         */
        column => (Object.assign({}, column, { visible: visibleColumns.includes(column.property) || column.type === 'detail' }))));
    }
    /**
     * @private
     * @return {?}
     */
    initializeListeners() {
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @return {?}
         */
        () => {
            if (this.popover) {
                this.popover.close();
            }
        }));
    }
    /**
     * @private
     * @param {?} property
     * @return {?}
     */
    isDisableColumn(property) {
        return this.visibleColumns.length >= this.maxColumns ? !this.visibleColumns.includes(property) : false;
    }
    /**
     * @private
     * @param {?=} columns
     * @return {?}
     */
    mapTableColumnsToCheckboxOptions(columns = []) {
        /** @type {?} */
        const columnsOptions = [];
        columns.forEach((/**
         * @param {?} column
         * @return {?}
         */
        column => {
            if (column.type !== 'detail') {
                columnsOptions.push({
                    value: column.property,
                    label: this.getColumnTitleLabel(column),
                    disabled: this.isDisableColumn(column.property)
                });
            }
        }));
        return columnsOptions;
    }
    /**
     * @private
     * @param {?} columns
     * @return {?}
     */
    onChangeColumns(columns) {
        const { firstChange, currentValue = [], previousValue = [] } = columns;
        // atualizara o defaultColumns, quando for a primeira vez ou quando o defaultColumns for diferente do currentValue
        if (firstChange || (this.defaultColumns.length !== currentValue.length)) {
            this.defaultColumns = currentValue;
        }
        // verifica se o valor anterior é diferente do atual para atualizar as columnsOptions apenas quando for necessario
        if (previousValue.length !== currentValue.length) {
            this.updateColumnsOptions(currentValue);
        }
    }
    /**
     * @private
     * @return {?}
     */
    removeListeners() {
        if (this.resizeListener) {
            this.resizeListener();
        }
    }
    /**
     * @private
     * @param {?} columns
     * @return {?}
     */
    updateColumnsOptions(columns) {
        this.visibleColumns = this.getVisibleColumns(columns);
        this.columnsOptions = this.mapTableColumnsToCheckboxOptions(columns);
        this.onChangeVisibleColumns(this.visibleColumns);
    }
}
PoTableColumnManagerComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-table-column-manager',
                template: "<po-popover #popover *ngIf=\"target\" [p-target]=\"target\" p-position=\"bottom-left\">\n\n  <div class=\"po-table-column-manager-header\">\n    <div class=\"po-table-column-manager-header-title\">{{ literals.columnsManager }}</div>\n\n    <div class=\"po-table-column-manager-header-close\">\n      <button class=\"po-table-column-manager-header-close-button po-clickable po-icon po-icon-close\" (click)=\"popover.close()\">\n      </button>\n    </div>\n  </div>\n\n  <div class=\"po-table-column-manager-body\">\n    <po-checkbox-group\n      name=\"visibleColumns\"\n      [(ngModel)]=\"visibleColumns\"\n      p-columns=\"1\"\n      [p-options]=\"columnsOptions\"\n      (p-change)=\"onChangeVisibleColumns($event)\">\n    </po-checkbox-group>\n  </div>\n\n  <div class=\"po-table-column-manager-footer\">\n    <po-button\n      class=\"po-table-column-manager-footer-restore\"\n      p-small\n      p-type=\"link\"\n      [p-label]=\"literals.restoreDefault\"\n      (p-click)=\"restore()\">\n    </po-button>\n  </div>\n</po-popover>\n"
            }] }
];
/** @nocollapse */
PoTableColumnManagerComponent.ctorParameters = () => [
    { type: Renderer2 }
];
PoTableColumnManagerComponent.propDecorators = {
    columns: [{ type: Input, args: ['p-columns',] }],
    maxColumns: [{ type: Input, args: ['p-max-columns',] }],
    target: [{ type: Input, args: ['p-target',] }],
    visibleColumnsChange: [{ type: Output, args: ['p-visible-columns-change',] }],
    popover: [{ type: ViewChild, args: [PoPopoverComponent, { static: false },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoTableColumnManagerComponent.prototype._maxColumns;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.columnsOptions;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.literals;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.visibleColumns;
    /**
     * @type {?}
     * @private
     */
    PoTableColumnManagerComponent.prototype.defaultColumns;
    /**
     * @type {?}
     * @private
     */
    PoTableColumnManagerComponent.prototype.resizeListener;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.columns;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.target;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.visibleColumnsChange;
    /** @type {?} */
    PoTableColumnManagerComponent.prototype.popover;
    /**
     * @type {?}
     * @private
     */
    PoTableColumnManagerComponent.prototype.renderer;
}
