// @portinari/portinari-ui/esm5/lib/components/po-tag/po-tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { PoKeyCodeEnum } from '../../enums/po-key-code.enum';
import { PoTagBaseComponent } from './po-tag-base.component';
import { PoTagIcon } from './enums/po-tag-icon.enum';
import { PoTagType } from './enums/po-tag-type.enum';
/** @type {?} */
var poTagTypeDefault = 'po-tag-' + PoTagType.Info;
/**
 * \@docsExtends PoTagBaseComponent
 *
 * \@example
 *
 * <example name="po-tag-basic" title="Portinari Tag Basic">
 *  <file name="sample-po-tag-basic/sample-po-tag-basic.component.html"> </file>
 *  <file name="sample-po-tag-basic/sample-po-tag-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-tag-labs" title="Portinari Tag Labs">
 *  <file name="sample-po-tag-labs/sample-po-tag-labs.component.html"> </file>
 *  <file name="sample-po-tag-labs/sample-po-tag-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-tag-bank-account" title="Portinari Tag - Bank Account">
 *  <file name="sample-po-tag-bank-account/sample-po-tag-bank-account.component.html"> </file>
 *  <file name="sample-po-tag-bank-account/sample-po-tag-bank-account.component.ts"> </file>
 * </example>
 */
var PoTagComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoTagComponent, _super);
    function PoTagComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    PoTagComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.isClickable = this.click.observers.length > 0;
    };
    Object.defineProperty(PoTagComponent.prototype, "iconFromType", {
        get: /**
         * @return {?}
         */
        function () {
            switch (this.type) {
                case PoTagType.Danger: return PoTagIcon.Danger;
                case PoTagType.Info: return PoTagIcon.Info;
                case PoTagType.Success: return PoTagIcon.Success;
                case PoTagType.Warning: return PoTagIcon.Warning;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "iconTypeString", {
        get: /**
         * @return {?}
         */
        function () {
            return typeof this.icon === 'string';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "tagColor", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.type) {
                return this.inverse ? "po-tag-" + this.type + "-inverse" : "po-tag-" + this.type;
            }
            if (this.color && !this.type) {
                return this.inverse ? "po-text-" + this.color : "po-" + this.color;
            }
            if (!this.type && !this.color) {
                return this.inverse ? poTagTypeDefault + "-inverse" : poTagTypeDefault;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoTagComponent.prototype, "tagOrientation", {
        get: /**
         * @return {?}
         */
        function () {
            return this.orientation === this.poTagOrientation.Horizontal;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PoTagComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var submittedTagItem = { value: this.value, type: this.type };
        this.click.emit(submittedTagItem);
    };
    /**
     * @param {?} event
     * @param {?} keyCode
     * @return {?}
     */
    PoTagComponent.prototype.onKeyPressed = /**
     * @param {?} event
     * @param {?} keyCode
     * @return {?}
     */
    function (event, keyCode) {
        /** @type {?} */
        var isValidKey = event.keyCode === PoKeyCodeEnum[keyCode];
        if (isValidKey) {
            event.preventDefault();
            event.stopPropagation();
            this.onClick();
        }
    };
    PoTagComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-tag',
                    template: "<div class=\"po-tag-container\" [class.po-tag-container-horizontal]=\"tagOrientation\">\n  <div *ngIf=\"label\" class=\"po-tag-title po-text-nowrap\">\n    <span class=\"po-tag-label\">{{ tagOrientation ? label + ':' : label }}</span>\n  </div>\n\n  <div class=\"po-tag-sub-container\">\n    <div class=\"po-tag\"\n      [class.po-clickable]=\"isClickable\"\n      [class.po-tag-inverse]=\"inverse\"\n      [ngClass]=\"tagColor\"\n      tabindex=\"0\"\n      (click)=\"onClick()\"\n      (keydown)=\"onKeyPressed($event, 'enter')\"\n      (keyup)=\"onKeyPressed($event, 'space')\">\n\n      <span *ngIf=\"icon\" class=\"po-icon\" [ngClass]=\"!type && iconTypeString ? icon : iconFromType\"></span>\n      <span class=\"po-tag-value\">{{value}}</span>\n    </div>\n  </div>\n</div>\n"
                }] }
    ];
    return PoTagComponent;
}(PoTagBaseComponent));
export { PoTagComponent };
if (false) {
    /** @type {?} */
    PoTagComponent.prototype.isClickable;
}
