// @portinari/portinari-ui/esm2015/lib/components/po-stepper/po-stepper-circle/po-stepper-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { PoStepperStatus } from '../enums/po-stepper-status.enum';
/** @type {?} */
const poMediumStepSize = 32;
/** @type {?} */
const poLargeStepSize = 48;
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
export class PoStepperCircleComponent {
    /**
     * @return {?}
     */
    get isActive() {
        return this.status === PoStepperStatus.Active;
    }
    /**
     * @return {?}
     */
    get isDefault() {
        return this.status === PoStepperStatus.Default;
    }
    /**
     * @return {?}
     */
    get isDisabled() {
        return this.status === PoStepperStatus.Disabled;
    }
    /**
     * @return {?}
     */
    get isDone() {
        return this.status === PoStepperStatus.Done;
    }
    /**
     * @return {?}
     */
    get isError() {
        return this.status === PoStepperStatus.Error;
    }
    /**
     * @return {?}
     */
    get isLargeStep() {
        return this.size >= poLargeStepSize;
    }
    /**
     * @return {?}
     */
    get isMediumStep() {
        return this.size >= poMediumStepSize && !this.isLargeStep;
    }
}
PoStepperCircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-stepper-circle',
                template: "<div class=\"po-stepper-circle\"\n  [class.po-stepper-circle-with-icon]=\"icons\"\n  [style.height.px]=\"size\"\n  [style.width.px]=\"size\"\n  [tabindex]=\"isDisabled ? -1 : 0\">\n\n  <span class=\"po-stepper-circle-content\"\n    [class.po-icon]=\"icons\"\n    [class.po-icon-exclamation]=\"icons && isError\"\n    [class.po-icon-info]=\"icons && (isActive || isDefault || isDisabled)\"\n    [class.po-icon-ok]=\"icons && isDone\"\n    [class.po-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.po-stepper-circle-content-md]=\"isMediumStep\">\n\n    {{ !icons ? content: '' }}\n  </span>\n</div>\n"
            }] }
];
PoStepperCircleComponent.propDecorators = {
    content: [{ type: Input, args: ['p-content',] }],
    icons: [{ type: Input, args: ['p-icons',] }],
    size: [{ type: Input, args: ['p-size',] }],
    status: [{ type: Input, args: ['p-status',] }]
};
if (false) {
    /** @type {?} */
    PoStepperCircleComponent.prototype.content;
    /** @type {?} */
    PoStepperCircleComponent.prototype.icons;
    /** @type {?} */
    PoStepperCircleComponent.prototype.size;
    /** @type {?} */
    PoStepperCircleComponent.prototype.status;
}
