// @portinari/portinari-ui/esm2015/lib/components/po-page/po-page-edit/po-page-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { PoPageEditBaseComponent } from './po-page-edit-base.component';
import { callAction, hasAction } from '../po-page-util/po-page-util';
/**
 * \@docsExtends PoPageEditBaseComponent
 *
 * \@example
 *
 * <example name="po-page-edit-basic" title="Portinari Page Edit Basic">
 *  <file name="sample-po-page-edit-basic/sample-po-page-edit-basic.component.html"> </file>
 *  <file name="sample-po-page-edit-basic/sample-po-page-edit-basic.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-edit-labs" title="Portinari Page Edit Labs">
 *  <file name="sample-po-page-edit-labs/sample-po-page-edit-labs.component.html"> </file>
 *  <file name="sample-po-page-edit-labs/sample-po-page-edit-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-page-edit-user" title="Portinari Page Edit - User">
 *  <file name="sample-po-page-edit-user/sample-po-page-edit-user.component.html"> </file>
 *  <file name="sample-po-page-edit-user/sample-po-page-edit-user.component.ts"> </file>
 * </example>
 */
export class PoPageEditComponent extends PoPageEditBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
        this.hasAction = hasAction;
        this.callAction = callAction;
        this.parentContext = viewRef['_view']['component'];
    }
    /**
     * @param {?} icon
     * @return {?}
     */
    getIcon(icon) {
        if (icon === 'cancel') {
            return this.isPrimaryAction('cancel') ? 'po-icon-close' : '';
        }
        if (icon === 'saveNew') {
            return this.isPrimaryAction('saveNew') ? 'po-icon-ok' : '';
        }
        return '';
    }
    /**
     * @param {?} type
     * @return {?}
     */
    getType(type) {
        /** @type {?} */
        const isCancelPrimaryAction = type === 'cancel' && this.isPrimaryAction('cancel');
        /** @type {?} */
        const isSaveNewPrimaryAction = type === 'saveNew' && this.isPrimaryAction('saveNew');
        return isCancelPrimaryAction || isSaveNewPrimaryAction ? 'primary' : 'default';
    }
    /**
     * @return {?}
     */
    hasAnyAction() {
        return hasAction('cancel', this.parentContext) || hasAction('saveNew', this.parentContext) || hasAction('save', this.parentContext);
    }
    /**
     * @return {?}
     */
    hasPageHeader() {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    }
    /**
     * @private
     * @param {?} action
     * @return {?}
     */
    isPrimaryAction(action) {
        /** @type {?} */
        const hasSaveAction = !hasAction('save', this.parentContext);
        if (action === 'saveNew') {
            return hasSaveAction;
        }
        if (action === 'cancel') {
            return !hasAction('saveNew', this.parentContext) && hasSaveAction;
        }
        return false;
    }
}
PoPageEditComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-page-edit',
                template: "<po-page>\n  <!-- HEADER -->\n  <po-page-header\n    *ngIf=\"hasPageHeader()\"\n    [p-breadcrumb]=\"breadcrumb\"\n    [p-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div *ngIf=\"hasAnyAction()\" class=\"po-page-header-actions\">\n      <po-button\n        *ngIf=\"hasAction('cancel', parentContext)\"\n        [p-icon]=\"getIcon('cancel')\"\n        [p-label]=\"literals.cancel\"\n        [p-type]=\"getType('cancel')\"\n        (p-click)=\"callAction('cancel', parentContext)\">\n      </po-button>\n\n      <po-button\n        *ngIf=\"hasAction('saveNew', parentContext)\"\n        [p-disabled]=\"disableSubmit\"\n        [p-icon]=\"getIcon('saveNew')\"\n        [p-label]=\"literals.saveNew\"\n        [p-type]=\"getType('saveNew')\"\n        (p-click)=\"callAction('saveNew', parentContext)\">\n      </po-button>\n\n      <po-button\n        *ngIf=\"hasAction('save', parentContext)\"\n        p-icon=\"po-icon-ok\"\n        p-type=\"primary\"\n        [p-disabled]=\"disableSubmit\"\n        [p-label]=\"literals.save\"\n        (p-click)=\"callAction('save', parentContext)\">\n      </po-button>\n    </div>\n\n  </po-page-header>\n\n  <!-- CONTENT -->\n  <po-page-content>\n    <ng-content>\n    </ng-content>\n  </po-page-content>\n\n</po-page>\n"
            }] }
];
/** @nocollapse */
PoPageEditComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
if (false) {
    /** @type {?} */
    PoPageEditComponent.prototype.hasAction;
    /** @type {?} */
    PoPageEditComponent.prototype.callAction;
    /** @type {?} */
    PoPageEditComponent.prototype.parentContext;
}
