// @portinari/portinari-ui/esm5/lib/components/po-table/po-table-column-link/po-table-column-link.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { isExternalLink } from '../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir link nas colunas.
 */
var PoTableColumnLinkComponent = /** @class */ (function () {
    function PoTableColumnLinkComponent() {
    }
    Object.defineProperty(PoTableColumnLinkComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.disabled) {
                return 'disabled';
            }
            if (this.action) {
                return 'action';
            }
            if (isExternalLink(this.link)) {
                return 'externalLink';
            }
            return 'internalLink';
        },
        enumerable: true,
        configurable: true
    });
    PoTableColumnLinkComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-table-column-link',
                    template: "<ng-container [ngSwitch]=\"type\">\n  <p *ngSwitchCase=\"'disabled'\" class=\"po-table-link-disabled\">{{ value }}</p>\n  <a *ngSwitchCase=\"'action'\" class=\"po-table-link\" (click)=\"action(value, row)\">{{ value }}</a>\n  <a *ngSwitchCase=\"'externalLink'\" class=\"po-table-link\" [href]=\"link\" target=\"_blank\">{{ value }}</a>\n  <a *ngSwitchCase=\"'internalLink'\" class=\"po-table-link\" [routerLink]=\"link\">{{ value }}</a>\n</ng-container>\n"
                }] }
    ];
    PoTableColumnLinkComponent.propDecorators = {
        action: [{ type: Input, args: ['p-action',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        link: [{ type: Input, args: ['p-link',] }],
        row: [{ type: Input, args: ['p-row',] }],
        value: [{ type: Input, args: ['p-value',] }]
    };
    return PoTableColumnLinkComponent;
}());
export { PoTableColumnLinkComponent };
if (false) {
    /** @type {?} */
    PoTableColumnLinkComponent.prototype.action;
    /** @type {?} */
    PoTableColumnLinkComponent.prototype.disabled;
    /** @type {?} */
    PoTableColumnLinkComponent.prototype.link;
    /** @type {?} */
    PoTableColumnLinkComponent.prototype.row;
    /** @type {?} */
    PoTableColumnLinkComponent.prototype.value;
}
