// @portinari/portinari-ui/esm5/lib/components/po-accordion/po-accordion.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ContentChildren, QueryList } from '@angular/core';
import { PoAccordionBaseComponent } from './po-accordion-base.component';
import { PoAccordionItemComponent } from './po-accordion-item/po-accordion-item.component';
import { PoAccordionService } from './services/po-accordion.service';
/**
 * \@docsExtends PoAccordionBaseComponent
 *
 * \@example
 *
 * <example name="po-accordion-basic" title="Portinari Accordion Basic" >
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.html"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-po-accordion-basic/sample-po-accordion-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-labs" title="Portinari Accordion Labs" >
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.html"> </file>
 *  <file name="sample-po-accordion-labs/sample-po-accordion-labs.component.ts"> </file>
 * </example>
 *
 * <example name="po-accordion-faq" title="Portinari Accordion - FAQs" >
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.html"> </file>
 *  <file name="sample-po-accordion-faq/sample-po-accordion-faq.component.ts"> </file>
 * </example>
 */
var PoAccordionComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PoAccordionComponent, _super);
    function PoAccordionComponent(accordionService) {
        var _this = _super.call(this) || this;
        _this.accordionService = accordionService;
        _this.receiveFromChildAccordionSubscription();
        return _this;
    }
    /**
     * @return {?}
     */
    PoAccordionComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.accordionServiceSubscription.unsubscribe();
    };
    /**
     * @param {?} event
     * @param {?} poAccordionItem
     * @return {?}
     */
    PoAccordionComponent.prototype.headerToggle = /**
     * @param {?} event
     * @param {?} poAccordionItem
     * @return {?}
     */
    function (event, poAccordionItem) {
        poAccordionItem.expanded = event;
        this.toggle(poAccordionItem);
    };
    /**
     * @private
     * @return {?}
     */
    PoAccordionComponent.prototype.receiveFromChildAccordionSubscription = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.accordionServiceSubscription = this.accordionService.receiveFromChildAccordionClicked()
            .subscribe((/**
         * @param {?} poAccordionItem
         * @return {?}
         */
        function (poAccordionItem) { return _this.toggle(poAccordionItem); }));
    };
    /**
     * @private
     * @param {?} poAccordionItem
     * @return {?}
     */
    PoAccordionComponent.prototype.toggle = /**
     * @private
     * @param {?} poAccordionItem
     * @return {?}
     */
    function (poAccordionItem) {
        /** @type {?} */
        var isCurrentAccordionCollapsed = !poAccordionItem.expanded;
        if (isCurrentAccordionCollapsed) {
            this.expandedActiveAccordionItem = null;
            return;
        }
        if (this.expandedActiveAccordionItem) {
            this.expandedActiveAccordionItem.collapse();
        }
        this.expandedActiveAccordionItem = poAccordionItem;
    };
    PoAccordionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-accordion',
                    template: "<div class=\"po-accordion po-container po-container-no-padding\">\n\n  <section *ngFor=\"let poAccordionItem of poAccordionItems\"\n    class=\"po-accordion-item\"\n    [class.po-accordion-item-active]=\"poAccordionItem.expanded\">\n\n    <po-accordion-item-header\n      [p-expanded]=\"poAccordionItem.expanded\"\n      [p-label]=\"poAccordionItem.label\"\n      (p-toggle)=\"headerToggle($event, poAccordionItem)\">\n    </po-accordion-item-header>\n\n    <po-accordion-item-body [p-expanded]=\"poAccordionItem.expanded\">\n      <ng-container *ngTemplateOutlet=\"poAccordionItem.templateRef\"></ng-container>\n    </po-accordion-item-body>\n  </section>\n\n</div>\n",
                    providers: [PoAccordionService]
                }] }
    ];
    /** @nocollapse */
    PoAccordionComponent.ctorParameters = function () { return [
        { type: PoAccordionService }
    ]; };
    PoAccordionComponent.propDecorators = {
        poAccordionItems: [{ type: ContentChildren, args: [PoAccordionItemComponent,] }]
    };
    return PoAccordionComponent;
}(PoAccordionBaseComponent));
export { PoAccordionComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.accordionServiceSubscription;
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.expandedActiveAccordionItem;
    /** @type {?} */
    PoAccordionComponent.prototype.poAccordionItems;
    /**
     * @type {?}
     * @private
     */
    PoAccordionComponent.prototype.accordionService;
}
