// @portinari/portinari-ui/esm5/lib/components/po-accordion/po-accordion-item/po-accordion-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { PoAccordionService } from '../services/po-accordion.service';
/**
 * \@description
 *
 * Componente utilizado para renderizar os itens do `po-accordion`.
 *
 * O componente `po-accordion` já faz o controle de abertura e fechamento dos itens automaticamente,
 * mas caso houver a necessidade de abrir algum dos `po-accordion-item` via Typescript, pode ser feita da seguinte forma:
 *
 * ```
 * <po-accordion>
 *   <po-accordion-item p-label="Portinari Accordion 1" #item1>
 *      Accordion 1
 *   </po-accordion-item>
 *
 *   <po-accordion-item p-label="Portinari Accordion 2">
 *      Accordion 2
 *   </po-accordion-item>
 * </po-accordion>
 * ```
 *
 * e no typescript pode-se utilizar o `\@ViewChild`:
 *
 * ```
 * \@ViewChild(PoAccordionItemComponent, { static: true }) item1: PoAccordionItemComponent;
 *
 *  ngAfterContentInit() {
 *    // ou utilizar o método collapse()
 *    this.item1.expand();
 *  }
 * ```
 */
var PoAccordionItemComponent = /** @class */ (function () {
    function PoAccordionItemComponent(accordionService) {
        this.accordionService = accordionService;
    }
    /**
     * Método para colapsar o `po-accordion-item`.
     */
    /**
     * Método para colapsar o `po-accordion-item`.
     * @return {?}
     */
    PoAccordionItemComponent.prototype.collapse = /**
     * Método para colapsar o `po-accordion-item`.
     * @return {?}
     */
    function () {
        this.expanded = false;
        this.accordionService.sendToParentAccordionItemClicked(this);
    };
    /**
     * Método para expandir o `po-accordion-item`.
     */
    /**
     * Método para expandir o `po-accordion-item`.
     * @return {?}
     */
    PoAccordionItemComponent.prototype.expand = /**
     * Método para expandir o `po-accordion-item`.
     * @return {?}
     */
    function () {
        this.expanded = true;
        this.accordionService.sendToParentAccordionItemClicked(this);
    };
    PoAccordionItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'po-accordion-item',
                    template: "<ng-template>\n  <ng-content></ng-content>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    PoAccordionItemComponent.ctorParameters = function () { return [
        { type: PoAccordionService }
    ]; };
    PoAccordionItemComponent.propDecorators = {
        label: [{ type: Input, args: ['p-label',] }],
        templateRef: [{ type: ViewChild, args: [TemplateRef, { static: true },] }]
    };
    return PoAccordionItemComponent;
}());
export { PoAccordionItemComponent };
if (false) {
    /** @type {?} */
    PoAccordionItemComponent.prototype.expanded;
    /**
     * Título do item.
     * @type {?}
     */
    PoAccordionItemComponent.prototype.label;
    /** @type {?} */
    PoAccordionItemComponent.prototype.templateRef;
    /**
     * @type {?}
     * @private
     */
    PoAccordionItemComponent.prototype.accordionService;
}
