// @portinari/portinari-ui/esm5/lib/components/po-button/po-button-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { convertToBoolean } from '../../utils/util';
/** @type {?} */
var PO_BUTTON_TYPES = ['default', 'primary', 'danger', 'link'];
/** @type {?} */
var PO_BUTTON_TYPE_DEFAULT = 'default';
/**
 * \@description
 *
 * O `po-button` permite que o usuário execute ações predefinidas pelo desenvolvedor.
 *
 * Através dos tipos, é possível identificar a importância de cada ação, sendo ela primária (`primary`) ou até mesmo uma
 * ação irreversível (`danger`), como a exclusão de um registro.
 *
 * #### Boas práticas
 *
 * - Evite `labels` extensos que quebram o layout do `po-button`, use `labels` diretos, curtos e intuitivos.
 * - Utilize apenas um `po-button` configurado como `primary` por página.
 * - Para ações irreversíveis use sempre o tipo `danger`.
 */
var PoButtonBaseComponent = /** @class */ (function () {
    function PoButtonBaseComponent() {
        this._disabled = false;
        this._loading = false;
        this._small = false;
        this._type = 'default';
        /**
         * Ação que será executada quando o usuário clicar sobre o `po-button`.
         */
        this.click = new EventEmitter();
    }
    Object.defineProperty(PoButtonBaseComponent.prototype, "loading", {
        get: /**
         * @return {?}
         */
        function () {
            return this._loading;
        },
        /**
         * @optional
         *
         * @description
         *
         * Exibe um ícone de carregamento à esquerda do _label_ do botão.
         *
         * > Quando esta propriedade estiver habilitada, desabilitará o botão.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Exibe um ícone de carregamento à esquerda do _label_ do botão.
         *
         * > Quando esta propriedade estiver habilitada, desabilitará o botão.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._loading = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoButtonBaseComponent.prototype, "small", {
        get: /**
         * @return {?}
         */
        function () {
            return this._small;
        },
        /**
         * @optional
         *
         * @description
         *
         * Deixa o botão menor.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Deixa o botão menor.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._small = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoButtonBaseComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this._type;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define o estilo do `po-button`.
         *
         * Valore válidos:
         *  - `default`: estilo padrão do `po-button`.
         *  - `primary`: deixa o `po-button` com destaque, deve ser usado para ações primárias.
         *  - `danger`: deve ser usado para ações que o usuário precisa ter cuidado ao executa-lá.
         *  - `link`: o `po-button` recebe o estilo de um link.
         *
         * @default `default`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define o estilo do `po-button`.
         *
         * Valore válidos:
         *  - `default`: estilo padrão do `po-button`.
         *  - `primary`: deixa o `po-button` com destaque, deve ser usado para ações primárias.
         *  - `danger`: deve ser usado para ações que o usuário precisa ter cuidado ao executa-lá.
         *  - `link`: o `po-button` recebe o estilo de um link.
         *
         * \@default `default`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._type = PO_BUTTON_TYPES.includes(value) ? value : PO_BUTTON_TYPE_DEFAULT;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PoButtonBaseComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita o `po-button` e não permite que o usuário interaja com o mesmo.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Desabilita o `po-button` e não permite que o usuário interaja com o mesmo.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = (/** @type {?} */ (value)) === '' ? true : convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    PoButtonBaseComponent.propDecorators = {
        label: [{ type: Input, args: ['p-label',] }],
        loading: [{ type: Input, args: ['p-loading',] }],
        icon: [{ type: Input, args: ['p-icon',] }],
        small: [{ type: Input, args: ['p-small',] }],
        type: [{ type: Input, args: ['p-type',] }],
        disabled: [{ type: Input, args: ['p-disabled',] }],
        click: [{ type: Output, args: ['p-click',] }]
    };
    return PoButtonBaseComponent;
}());
export { PoButtonBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._loading;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._small;
    /**
     * @type {?}
     * @private
     */
    PoButtonBaseComponent.prototype._type;
    /**
     * \@optional
     *
     * \@description
     *
     * Label do botão.
     * @type {?}
     */
    PoButtonBaseComponent.prototype.label;
    /**
     * Ícone exibido ao lado esquerdo do label do botão.
     *
     * É possível usar qualquer uma dos ícones da [Biblioteca de ícones](/guides/icons).
     * @type {?}
     */
    PoButtonBaseComponent.prototype.icon;
    /**
     * Ação que será executada quando o usuário clicar sobre o `po-button`.
     * @type {?}
     */
    PoButtonBaseComponent.prototype.click;
}
