// @portinari/portinari-ui/esm2015/lib/components/po-grid/po-grid-cell-action/po-grid-cell-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class PoGridCellActionComponent {
    constructor() {
        this.position = '';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeyDownContent(event) {
        // console.log('onKeyDownContent: ', event);
        // ENTER
        if (event.keyCode === 13) {
            event.preventDefault();
            // this.openActions(this.value);
            return;
        }
    }
}
PoGridCellActionComponent.decorators = [
    { type: Component, args: [{
                selector: 'po-grid-cell-action',
                template: "<div>\n  <div #contentElement [id]=\"position\"\n    class=\"po-grid-cell-action-content\"\n    tabindex=\"0\"\n    (keydown)=\"onKeyDownContent($event)\">\n    <span>{{ value || '&nbsp;' }}</span>\n  </div>\n</div>\n\n",
                styles: [`
  .po-grid-cell-action-content {
    border-bottom: 1px solid #eceeee;
    border-left: none;
    border-right: 1px solid #eceeee;
    border-top: none;
    font-family: NunitoSans;
    font-size: 14px;
    height: 44px;
    line-height: 1.71;
    margin: 0;
    max-height: 44px;
    max-width: 56px;
    min-height: 44px;
    min-width: 56px;
    padding: 10px 16px;
    text-align: center;
    width: 56px;
  }

  .po-grid-cell-action-content:focus {
    border: none;
    outline: 1px solid #0c9abe;
  }
  `]
            }] }
];
/** @nocollapse */
PoGridCellActionComponent.ctorParameters = () => [];
PoGridCellActionComponent.propDecorators = {
    position: [{ type: Input, args: ['p-position',] }],
    value: [{ type: Input, args: ['p-value',] }]
};
if (false) {
    /** @type {?} */
    PoGridCellActionComponent.prototype.position;
    /** @type {?} */
    PoGridCellActionComponent.prototype.value;
}
